---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%", 
  message=FALSE, 
  warning=FALSE
)
```

# UMI4Cats <img src="man/figures/logo.png" width="121px" height="140px" align="right" style="padding-left:10px;background-color:white;" />

<!-- badges: start -->
```{r badges, echo = FALSE, results='asis'}
cat(
	# badge_devel(color="blue"),
  # badger::badge_dependencies("UMI4Cats"), 
	badger::badge_lifecycle(stage="maturing")
)

```
<!-- badges: end -->

The goal of UMI4Cats is to provide and easy-to-use package to analyze UMI-4C contact data.

## Installation

```
devtools::install_gitlab("pasquali-lab/umi4cats")
```
Now you can load the package using `library(UMI4Cats)`. 

## Basic usage

```{r load}
library(UMI4Cats)
``` 

```{r process, eval=FALSE}
## 0) Download example data
path <- downloadUMI4CexampleData()

## 1) Generate Digested genome ----------------------------
# The selected RE in this case is DpnII (|GATC), so the cs5p is "" and cs3p is GATC
hg19_dpnii <- digestGenome(cut_pos = 0,
                           res_enz = "GATC",
                           name_RE = "DpnII",
                           ref_gen = BSgenome.Hsapiens.UCSC.hg19::BSgenome.Hsapiens.UCSC.hg19, 
                           out_path = "digested_genome/")

## 2) Process UMI-4C fastq files --------------------------
raw_dir <- file.path(path, "CIITA", "raw")

contactsUMI4C(fastq_dir = raw_dir,
              wk_dir = "CIITA",
              bait_seq = "GGACAAGCTCCCTGCAACTCA",
              bait_pad = "GGACTTGCA",
              res_enz = "GATC",
              cut_pos = 0,
              digested_genome = hg19_dpnii,
              bowtie_index = file.path(path, "ref_genome", "ucsc.hg19.chr16"),
              ref_gen = BSgenome.Hsapiens.UCSC.hg19::BSgenome.Hsapiens.UCSC.hg19,
             threads = 5)
```

```{r analysis-plot, eval=TRUE}
## 3) Get filtering and alignment stats -------------------
statsUMI4C(wk_dir = system.file("extdata", "CIITA",
                               package="UMI4Cats"))

## 4) Analyze UMI-4C results ------------------------------
# Load sample processed file paths
files <- list.files(system.file("extdata", "CIITA", "count", 
                                package="UMI4Cats"),
                    pattern="*_counts.tsv",
                    full.names=TRUE)

# Create colData including all relevant information
colData <- data.frame(sampleID = gsub("_counts.tsv.gz", "", basename(files)),
                      file = files,
                      stringsAsFactors=FALSE)

library(tidyr)
colData <- colData %>% 
  separate(sampleID, 
           into=c("condition", "replicate", "viewpoint"),
           remove=FALSE)

# Load UMI-4C data and generate UMI4C object
umi <- makeUMI4C(colData=colData,
                 viewpoint_name="CIITA")

## 5) Perform differential test ---------------------------
umi <- fisherUMI4C(umi,
                   filter_low = 20)

## 6) Plot results ----------------------------------------
plotUMI4C(umi, 
          ylim=c(0,15),
          xlim=c(10.75e6, 11.25e6)
          )
```
