---
title: "Analyzing UMI-4C data with UMI4Cats"
author: "Mireia Ramos-Rodríguez and Marc Subirana-Granés"
package: UMI4Cats
bibliography: bibliography.bib
output: 
  BiocStyle::html_document:
    fig_width: 7
    fig_height: 5
vignette: >
  %\VignetteIndexEntry{Analyzing UMI-4C data with UMI4Cats}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
    collapse = TRUE,
    comment = "#>",
    eval = TRUE,
    warning = FALSE,
    message = FALSE,
    fig.align = "center",
    out.width = "60%"
)
```

```{r logo, echo=FALSE, eval=TRUE, out.width='10%'}
knitr::include_graphics("../man/figures/UMI4Cats.png", dpi = 800)
```

# Introduction

Hello stranger! If you are here, that means you've successfully completed the UMI-4C protocol and got some sequencing results! This vignette has the objective to guide you through performing a simple analysis of your brand-new UMI-4C contact data. Let's dive in!

```{r load}
library(UMI4Cats)
```

## Overview of the package

```{r umi4cats-scheme, echo=FALSE, eval=TRUE, fig.cap="Overview of the different functions included in the UMI4Cats package to analyze UMI-4C data."}
knitr::include_graphics("figures/scheme.png", dpi = 400)
```

## About the experimental design

One of the strengths of the UMI-4C assay [@Schwartzman2016] is that of reducing the PCR duplication bias allowing a more accurate quantification of chromatin interactions. For this reason, UMI-4C is mostly used when trying to compare changes in chromatin interactions between two conditions, cell types or developmental stages.

Taking into account this main application, UMI4Cats has been developed to facilitate the differential analysis between conditions at a given viewpoint of interest. When analyzing your data with this package, you should take into account the following points: 

- Each analysis (and `UMI4C` object) should correspond to the **same viewpoint**. If you are analyzing different viewpoints in 
the same or different loci, you need to analyze them separately. 

- The UMI4Cats package is mostly oriented to the performance of differential analysis. For this reason, in the analysis step, it requires to **annotate each sample "replicate" and "condition"**. However, if you only have one replicate or condition you will be able to perform the analysis and visualization of the data but not the differential analysis.

- When performing the differential analysis, UMI4Cats is only able to deal with a **condition with 2 different levels**. If you have more than two conditions, you should produce different UMI4C objects and perform pairwise comparisons.

## About the example datasets

The datasets used in this vignette (from @Ramos-Rodriguez2019) are available for download if you want to reproduce the contents of this vignette.

Briefly, the datasets correspond to human pancreatic islets exposed (`cyt`) or not (`ctrl`) to pro-inflammatory cytokines for 48 hours. In this example we are using the UMI-4C data generated from two different biological replicates (HI24 and HI32) using the promoter of the *CIITA* gene as viewpoint.

# Quick start

In this section we summarize a complete analysis using the examples provided in this package. 

```{r processing-quick-start, eval=FALSE}
## 0) Download example data -------------------------------
path <- downloadUMI4CexampleData()

## 1) Generate Digested genome ----------------------------
# The selected RE in this case is DpnII (|GATC), so the cut_pos is 0, and the res_enz "GATC".
hg19_dpnii <- digestGenome(
    cut_pos = 0,
    res_enz = "GATC",
    name_RE = "DpnII",
    ref_gen = BSgenome.Hsapiens.UCSC.hg19::BSgenome.Hsapiens.UCSC.hg19,
    out_path = "digested_genome/"
)

## 2) Process UMI-4C fastq files --------------------------
raw_dir <- file.path(path, "CIITA", "fastq")

contactsUMI4C(
    fastq_dir = raw_dir,
    wk_dir = "CIITA",
    bait_seq = "GGACAAGCTCCCTGCAACTCA",
    bait_pad = "GGACTTGCA",
    res_enz = "GATC",
    cut_pos = 0,
    digested_genome = hg19_dpnii,
    bowtie_index = file.path(path, "ref_genome", "ucsc.hg19.chr16"),
    ref_gen = BSgenome.Hsapiens.UCSC.hg19::BSgenome.Hsapiens.UCSC.hg19,
    threads = 5
)
``` 

```{r analysis-quick-start}
## 3) Get filtering and alignment stats -------------------
statsUMI4C(wk_dir = system.file("extdata", "CIITA",
    package = "UMI4Cats"
))

## 4) Analyze the results ---------------------------------
# Load sample processed file paths
files <- list.files(system.file("extdata", "CIITA", "count",
    package = "UMI4Cats"
),
pattern = "*_counts.tsv.gz",
full.names = TRUE
)

# Create colData including all relevant information
colData <- data.frame(
    sampleID = gsub("_counts.tsv.gz", "", basename(files)),
    file = files,
    stringsAsFactors = FALSE
)

library(tidyr)
colData <- colData %>%
    separate(sampleID,
        into = c("condition", "replicate", "viewpoint"),
        remove = FALSE
    )

# Load UMI-4C data and generate UMI4C object
umi <- makeUMI4C(
    colData = colData,
    viewpoint_name = "CIITA"
)

## 5) Perform differential test ---------------------------
umi <- fisherUMI4C(umi,
    filter_low = 30
)

## 6)  Plot results ---------------------------------------
plotUMI4C(umi,
    ylim = c(0, 10) # ,
    # xlim=c(11e6, 11.5e6)
)
```


# Preparing necessary files

## Demultiplexing fastQ files containing multiple baits

One of the many advantages of using the UMI-4C protocol is that it allows multiplexing of different baits using the same sample. 

To facilitate the analysis, UMI4Cats  provides a function for demultiplexing the paired-end FastQ files returned by the sequencing facility: `demultiplexFastq`.

This function requieres as input the name of the R1 file -- it will automatically detect the R2 --, the barcode sequences and a path for the outputs. The barcode sequences and names to be used for each output sequences need to be provided as a `data.frame` with column names `sample` and `barcode`. 

```{r demultiplex, eval=FALSE}
path <- downloadUMI4CexampleData(use_sample = TRUE)

fastq <- file.path(path, "CIITA", "fastq", "sub_ctrl_hi24_CIITA_R1.fastq.gz")

barcodes <- data.frame(
    sample = c("CIITA"),
    barcode = c("GGACAAGCTCCCTGCAACTCA")
)

demultiplexFastq(
    barcodes = barcodes,
    fastq = fastq,
    out_path = path
)
```

## Reference genome digestion
For the processing of the UMI-4C fastQ files it is necessary to construct a digested genome using the same restriction enzyme used in the UMI-4C experiments.

The `UMI4Cats` package includes the `digestGenome()` function to make this process as easy as possible. The function uses a `BSgenome` object^[More information on `BSgenome` package and objects can be found [here](https://bioconductor.org/packages/release/bioc/html/BSgenome.html)] as reference genome which is digested by a given restriction enzyme cutting sequence (`res_enz`).

Besides the restriction sequence, it is also necessary to provide, as a zero-based numeric integer, the position at which the restriction enzyme cuts (`cut_pos`).

In the following table you can see three examples of the different cutting sequences for *DpnII*, *Csp6I* and *HindIII*.

Restriction enzyme | Restriction seq | `res_enz` | `cut_pos`
-------------------|-----------------|-----------|---------
DpnII              | :`GATC`         | GATC      | 0        
Csp6I              | `G`:`TAC`       | GTAC      | 1    
HindIII            | `A`:`AGCTT`     | AAGCTT    | 1       

For this example, we are using the hg19 `BSGenome` object and we are going to digest it using the *DpnII* enzyme. 

```{r digest}
library(BSgenome.Hsapiens.UCSC.hg19)
refgen <- BSgenome.Hsapiens.UCSC.hg19

hg19_dpnii <- digestGenome(
    res_enz = "GATC",
    cut_pos = 0,
    name_RE = "dpnII",
    ref_gen = refgen,
    sel_chr = "chr16",
    out_path = "digested_genome/"
)

hg19_dpnii
```

The digested genome will be saved in the folder provided in `out_path` as RData objects divided by chromosome. The path of the digested genome files is outputed by the function, so you can save it as a variable (in this case `hg19_dpnii`) and use it for downstream analyses.

# Processing UMI-4C FASTQ files

This step converts FastQ files of paired-end reads to UMI counts in the fragments resulted from the *in silico* digestion of the genome. The function for doing this analysis is `contactsUMI4C()` and should be ran in samples generated with the same viewpoint and restriction enzyme. 

The function will consider all FastQ files in the same folder `fastq_dir` to be part of the same experiment (viewpoint + restriction enzyme). However, if you want to specify a subset of samples to perform the analysis you can do so by using the `file_pattern` argument. This way, only the files matching the specified pattern will be use  as input for the analyses.

The different pairs for each sample should be named `_R1` or `_R2` and the function only accepts the following fastq suffixes: `.fastq`, `.fq`, `.fq.gz` or `.fastq.gz`.

For each analysis, the user needs to define 3 different sequences:

- **Bait/viewpoint sequence** (`bait_seq`). This is the downstream primer sequence (DS primer) that matches the sequence of the queried bait.
- **Padding sequence** (`bait_pad`). The padding sequence corresponds to the nucleotides between the DS primer end and the restriction enzyme site.
- **Restriction enzyme sequence** (`res_enz`). This sequence is the restriction sequence recognized by the selected restriction enzyme.

```{r read-scheme, echo=FALSE, eval=TRUE, fig.cap="Schematic of a UMI-4C read detailing the different elements that need to be used as input for processing the data."}
knitr::include_graphics("figures/read_scheme.png", dpi = 350)
```

Additionally, it is necessary to define the restiction enzyme cutting postion (`cut_pos`) as previously did for generating the digested genome. The function will also need the path for the digested genome (`digested_genome`) generated with the function `digestGenome()`.

As `contactsUMI4C()` performs the alignment using [Bowtie2](http://bowtie-bio.sourceforge.net/bowtie2/index.shtml), you will need to provide 
the reference genome indexes generated with bowtie2^[See the [getting started section](http://bowtie-bio.sourceforge.net/bowtie2/manual.shtml#getting-started-with-bowtie-2-lambda-phage-example) on the Bowtie2 page for more information on how to generate the index for the reference genome.]. Remember to make sure that your reference genome and digested genome correspond to the same build (in this example, hg19). 

```{r processing, message=TRUE}
## Use reduced example to make vignette faster
raw_dir <- system.file("extdata", "CIITA", "fastq",
    package = "UMI4Cats"
    )

index_path <- paste0(downloadUMI4CexampleData(only_index=TRUE),
                     "/ucsc.hg19.chr16")

## Run main function to process UMI-4C contacts
contactsUMI4C(
    fastq_dir = raw_dir,
    wk_dir = "CIITA",
    file_pattern = "ctrl_hi24_CIITA", # Select only one sample to reduce time
    bait_seq = "GGACAAGCTCCCTGCAACTCA",
    bait_pad = "GGACTTGCA",
    res_enz = "GATC",
    cut_pos = 0,
    digested_genome = hg19_dpnii,
    bowtie_index = index_path,
    ref_gen = BSgenome.Hsapiens.UCSC.hg19::BSgenome.Hsapiens.UCSC.hg19,
    sel_seqname = "chr16",
    threads = 2
)
```

```{r include=FALSE}
unlink(c(index_path, hg19_dpnii, "CIITA"), recursive = TRUE)
```


Internally, `contactsUMI4C()` will then run sequentially the following processes:

1. FastQ files preparation (`prepUMI4C`). In this preprocessing step, only reads containing the `bait_seq` + `bait_pad` + `res_enz` will be selected. Also, reads with mean Phread quality scores < 20 will be filtered out.
2. Split reads at restriction sites (`splitUMI4C`). Using the `res_enz` sequence, all reads from R1 and R2 will be split to mimic the fragments that were generated experimentally.
3. Align split reads to the reference genome (`alignmentUMI4C`). 
4. Collapse reads using the UMI (`counterUMI4C`). This step is done to count molecular events thus reducing artifacts due to PCR duplicates. The function will return all fragment ends 10Mb around the viewpoint.

**Note on memory usage**: For the preparation and splitting, the FastQ file is loaded into memory. If you are having problems with the memory size available in your computer, you can change the number of lines to be loaded using the `numb_reads` parameter. See `?contactsUMI4C` for more information.

Finally, the output (UMI-4C counts) is a compressed tsv file per sample that will be saved in `wk_dir/count`. These files will be used as input for the analyses performed in the following section.

## Quality control measures
Once the processing step has been ran, the statistics of the UMI-4C filtering, alignment and final number of UMIs can be generated from the logs generated by the `contactsUMI4C()` function. 

By using these logs, the function `statsUMI4C()` will produce a summary plot and a summary table with all statistics (in `wk_dir/logs/stats_summary.txt`). 

```{r stats}
statsUMI4C(wk_dir = system.file("extdata", "CIITA",
    package = "UMI4Cats"
))

stats <- read.delim(system.file("extdata", "CIITA", "logs", "stats_summary.txt",
    package = "UMI4Cats"
))

knitr::kable(stats)
```

The quality control measures summarized both in the plot and the table are:

- **Specific reads**. Corresponds to the number of reads that contained the full viewpoint sequence (`bait_seq` + `bait_pad` + `res_enz`).
- **Filtered reads**. A quality filtering to ensure that filtered (or kept) reads have a mean Phred quality score `>= 20`.
- **Mapping stats**. Indicates how many split reads are mapped or unmapped to the reference genome.
- **UMIs**. Shows the final number of molecular contacts detected.

# Analyzing UMI-4C data

After processing the fastQ reads and obtaining tables summarizing the contacts and the number of UMIs supporting each interaction with the viewpoint, the next step is that of analyzing these data, detect differential contacts and visualize the genomic interactions.

## Build the `UMI4C` object

The first step of the UMI-4C data analysis consists in loading the tables generated by the function `contactsUMI4C()` and use them to construct a `UMI4C` object. All these steps are performed automatically by the `makeUMI4C()` function. 

The `makeUMI4C` will need as input a data frame (`colData`) containing all relevant experiment information that will be needed for analyzing the data later on. The mandatory columns that this data frame should have are:

1. `sampleID`: Unique identifier for the sample.
2. `replicate`: Replicate identifier or number.
3. `condition`: Condition for performing the differential analysis. For example: "control" and "treatment", two different cell types, etc. The condition column should only have **two** different values. If more condition variables are provided, the differential analysis will fail.
4. `file`: Complete path and filename where the tsv files generated by `contactsUMI4C` are saved.

You can also include other additional columns to `colData`.

By default, the function will combine the samples using the **condition** column from `colData`, but you can change this behavior using the `grouping` argument in `makeUMI4C()`. For example, if you set `grouping="sampleID"` it will plot the individual profiles or if you set `grouping="replicate"` your samples will be summarized according to the replicates you provided in `colData`. By default, the reference group will be the one with less number of UMIs. You can override this behaviour by setting the `ref_umi4c` argument, which should be one of the groups present in the column used as `grouping`. 

Additionally, the `makeUMI4C` function also contains other arguments that can be used if you want to tweak the default parameters of the analysis. See `?makeUMI4C` to have a complete list and description of all the arguments.

```{r make-umi4c}
# Load sample processed file paths
files <- list.files(system.file("extdata", "CIITA", "count",
    package = "UMI4Cats"
),
pattern = "*_counts.tsv.gz",
full.names = TRUE
)

# Create colData including all relevant information
colData <- data.frame(
    sampleID = gsub("_counts.tsv.gz", "", basename(files)),
    file = files,
    stringsAsFactors = FALSE
)

library(tidyr)
colData <- colData %>%
    separate(sampleID,
        into = c("condition", "replicate", "viewpoint"),
        remove = FALSE
    )

# Load UMI-4C data and generate UMI4C object
umi <- makeUMI4C(
    colData = colData,
    viewpoint_name = "CIITA",
    grouping = "condition",
    ref_umi4c = "ctrl"
)

umi
```

The `makeUMI4C` function will perform the following steps to generate the `UMI4C` object:

1. **Remove fragment ends around the bait**, as they are generally biased because of their proximity to the viewpoint. The value of the region that will be excluded from the analysis can be specified using the `bait_exclusion` argument. The default is 3kb.
2. **Focus the scope** of the analysis in a specific genomic region around the bait, by default this is a 2Mb window centerd on the viewpoint. The default value can be changed using the `bait_expansion` argument.
3. Sum the UMIs of the different samples belonging to the same group (defined by the `grouping` variable).
4. **Obtain the normalization matrices** that will be used to scale the groups to the reference, by default the group with less UMIs. If you want to avoid this normalization step, you can set `normalized` to `FALSE`.
5. Calculate the **domainograms** for each group.
6. Calculate the **adaptive trend** for each group.

### Accessing information from the `UMI4C` object

The usual accessor functions from the `SummarizedExperiment-class`^[See more about the SummarizedExperiment class [here](https://bioconductor.org/packages/release/bioc/html/SummarizedExperiment.html)] also work with the UMI-4C class (for example: `assay`, `colData`, etc.). Other additional accessors have been created to retrieve different information:

- `dgram()`. Get a list of the domaingorams for each group.
- `bait()`. Retrieve a GRanges object with the bait position.
- `trend()`. Obtain a data.frame in long format with the adaptive smoothing trend.
- `results()`. Retrieve results from the differential analysis. This only works if a differential analysis has been performed on the UMI4C object.

```{r methods-umi4c}
head(assay(umi)) # Retrieve raw UMIs

colData(umi) # Retrieve column information

rowRanges(umi) # Retrieve fragment coordinates

dgram(umi) # Retrieve domainograms

bait(umi) # Retrieve bait coordinates

head(trend(umi)) # Retrieve adaptive smoothing trend
```

## Performing differential analysis

Once the `UMI4C` object is generated, you can perform a differential analysis using the `fisherUMI4C()` function. 

First of all, queried regions will be filtered according to the median number of UMIs present in the `filter_low` parameter. You can reduce this number or disable filtering using `filter_low = FALSE`.

Then, a contingency table for each region where the differential test should be performed will be created, where the group stored in `metadata(umi)$ref_umi4c` will be used as references. The values on the contingency table correspond to the following:


Group     | Query region | Region     
----------|--------------|-----------
Reference | $n1$         | $N1 - n1$
Condition | $n2$         | $N2 - n2$

Where $N1$ and $N2$ correspond to the total number of UMIs in the whole analyzed region (`metadata(umi)$region`) and $n1$ and $n2$ correspond to the total number of UMIs in the query region that is to be tested.

After all the Fisher's Exact Tests are performed, p-values are adjusted using the FDR method. Query regions with adjusted p-values > 0.05 will be considered significantly different. Check `?fisherUMI4()` for more information and additional arguments you can provide to the function.

### Differential analysis in a specific set of regions

Many times, UMI-4C is used to test whether a certain set of promoter-enhancer contacts differ between conditions or cell types. For this cases, the `fisherUMI4c()` with the argument `query_regions` set to those regions of interest can be used. Your query regions should be provided as a `GRanges` object. If your set of query regions is narrow (~1kb), you can resize them to the desired width using the `resize` argument.


```{r dif-query}
library(GenomicRanges)

# Provide your own query regions as GRanges objects
enhancers <- GRanges(c(
    "chr16:10925006-10928900",
    "chr16:11102721-11103700"
))

# Perform differential test
umi_dif1 <- fisherUMI4C(umi,
    query_regions = enhancers,
    filter_low = 20,
    resize = 5e3
)
```


### Differential analysis in the binned region

In case you do not have a candidate set of enhancer regions, the `fisherUMI4C()` function will bin the experiment region in bin of `window_size` size (default is 5kb). This will then be the query regions where the differential test will be performed, after filtering for number of UMIs with the `filter_low` argument.


```{r dif-bins}
# Perform differential test
umi_dif2 <- fisherUMI4C(umi,
    filter_low = 20,
    window_size = 5e3
)
```


### Retrieve differential analysis results

Results from this test can be retrieved using the `results()` on the `UMI4C` object returned by the test function. 

```{r fisher-umi4c}
results(umi_dif1, ordered = TRUE, counts = TRUE, format = "data.frame")

results(umi_dif2, ordered = TRUE, counts = TRUE)
```
 
The parameter `counts` indicates wheter raw counts used for the test should be outputed. `umis_ref` corresponds to the number of raw UMIs from the sample/group used as reference (accessible through `metadata(umi_dif)$ref_umi4c`).

# Visualizing UMI-4C contact data

Once the `UMI4C` object is created, you can plot it using the `plotUMI4C` function. 

The gene annotations will be extracted from the `TxDb.Hsapiens.UCSC.hg19.knownGene` package by default. Make sure that the annotations coincide with your reference genome. You can check the package [`GenomicFeatures`](https://bioconductor.org/packages/release/bioc/html/GenomicFeatures.html) for more information on available `TxDb` objects. 

```{r plot-umi4c}
plotUMI4C(umi,
    TxDb = TxDb.Hsapiens.UCSC.hg19.knownGene::TxDb.Hsapiens.UCSC.hg19.knownGene
)
```

The domainogram plotting is controlled by the `dgram_plot` argument. If you set it to `FALSE`, the domainogram will not be plotted.

```{r plot-nodgram}
plotUMI4C(umi, dgram_plot = FALSE)
```

In case you are interested in plotting the profiles of the different samples contained in your experiment, you will need to create another `UMI4C` object setting `grouping=sampleID` and then you will be able to plot this information:

```{r plot-reps}
umi_reps <- makeUMI4C(
    colData = colData,
    viewpoint_name = "CIITA",
    grouping = "sampleID"
)

plotUMI4C(umi_reps)
```

If the `UMI4C` object contains information on the differential contacts, this data will be shown in the plot as well.

```{r plot-dif}
plotUMI4C(umi_dif1)
plotUMI4C(umi_dif2)
```

There are several different arguments you can provide to `plotUMI4C` to modify the output plot. You can check them at `?plotUMI4C`.

The `plotUMI4C` function is a wrapper for separate functions that plot the different elements included in the figure. You can use each of the functions separately if you are interesting in combining them differently or with other ggplot2 objects. Check each function documentation at `?plotTrend`, `?plotGenes`, `?plotDomainogram` and `?plotDifferential`.

# References

