<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>UMI4C Contacts Processing — contactsUMI4C • UMI4Cats</title>

<!-- favicons -->
<link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png" />
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png" />
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png" />
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png" />

<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="UMI4C Contacts Processing — contactsUMI4C" />
<meta property="og:description" content="Using demultiplexed FastQ files as input, performs all necessary steps to end
 up with a tsv file summarizing the
restriction enzyme fragments and the number of UMIs supporting that specific
contact with the viewpoint (bait) of interest." />
<meta property="og:image" content="/logo.png" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">UMI4Cats</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.99.15</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../articles/UMI4Cats.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/Pasquali-lab/UMI4Cats/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>UMI4C Contacts Processing</h1>
    <small class="dont-index">Source: <a href='https://github.com/Pasquali-lab/UMI4Cats/blob/master/R/contactsUMI4C.R'><code>R/contactsUMI4C.R</code></a></small>
    <div class="hidden name"><code>contactsUMI4C.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Using demultiplexed FastQ files as input, performs all necessary steps to end
 up with a tsv file summarizing the
restriction enzyme fragments and the number of UMIs supporting that specific
contact with the viewpoint (bait) of interest.</p>
    </div>

    <pre class="usage"><span class='fu'>contactsUMI4C</span><span class='op'>(</span>
  <span class='va'>fastq_dir</span>,
  <span class='va'>wk_dir</span>,
  file_pattern <span class='op'>=</span> <span class='cn'>NULL</span>,
  <span class='va'>bait_seq</span>,
  <span class='va'>bait_pad</span>,
  <span class='va'>res_enz</span>,
  <span class='va'>cut_pos</span>,
  <span class='va'>digested_genome</span>,
  <span class='va'>bowtie_index</span>,
  threads <span class='op'>=</span> <span class='fl'>1</span>,
  numb_reads <span class='op'>=</span> <span class='fl'>1e+11</span>,
  rm_tmp <span class='op'>=</span> <span class='cn'>TRUE</span>,
  min_flen <span class='op'>=</span> <span class='fl'>20</span>,
  filter_bp <span class='op'>=</span> <span class='fl'>1e+07</span>,
  <span class='va'>ref_gen</span>,
  sel_seqname <span class='op'>=</span> <span class='cn'>NULL</span>
<span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>fastq_dir</th>
      <td><p>Path of the directory containing the FastQ files (compressed
or uncompressed).</p></td>
    </tr>
    <tr>
      <th>wk_dir</th>
      <td><p>Working directory where to save the outputs generated by the
UMI-4c analysis.</p></td>
    </tr>
    <tr>
      <th>file_pattern</th>
      <td><p>Character that can be used to filter the files you want
to analyze in the <code>fastq_dir</code>.</p></td>
    </tr>
    <tr>
      <th>bait_seq</th>
      <td><p>Character containing the bait primer sequence.</p></td>
    </tr>
    <tr>
      <th>bait_pad</th>
      <td><p>Character containing the pad sequence (sequence between the
bait primer and the restriction enzyme sequence).</p></td>
    </tr>
    <tr>
      <th>res_enz</th>
      <td><p>Character containing the restriction enzyme sequence.</p></td>
    </tr>
    <tr>
      <th>cut_pos</th>
      <td><p>Numeric indicating the nucleotide position where restriction
enzyme cuts (zero-based) (for example, for DpnII is 0).</p></td>
    </tr>
    <tr>
      <th>digested_genome</th>
      <td><p>Path for the digested genome file generated using the
<code><a href='digestGenome.html'>digestGenome</a></code> function.</p></td>
    </tr>
    <tr>
      <th>bowtie_index</th>
      <td><p>Path and prefix of the bowtie index to use for the
alignment.</p></td>
    </tr>
    <tr>
      <th>threads</th>
      <td><p>Number of threads to use in the analysis. Default=1.</p></td>
    </tr>
    <tr>
      <th>numb_reads</th>
      <td><p>Number of lines from the FastQ file to load in each loop.
If having memory size problems, change it to a smaller number. Default=10e10.</p></td>
    </tr>
    <tr>
      <th>rm_tmp</th>
      <td><p>Logical indicating whether to remove temporary files (sam and
intermediate bams). TRUE or FALSE. Default=TRUE.</p></td>
    </tr>
    <tr>
      <th>min_flen</th>
      <td><p>Minimal fragment length to use for selecting the fragments.
Default=20</p></td>
    </tr>
    <tr>
      <th>filter_bp</th>
      <td><p>Integer indicating the bp upstream and downstream of the
viewpoint to select for further analysis. Default=10e6</p></td>
    </tr>
    <tr>
      <th>ref_gen</th>
      <td><p>A BSgenome object of the reference genome.</p></td>
    </tr>
    <tr>
      <th>sel_seqname</th>
      <td><p>A character with the chromosome name to focus the
search for the viewpoint sequence.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>This function is a combination of calls to other functions that
perform the necessary steps for processing
UMI-4C data.</p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='kw'>if</span> <span class='op'>(</span><span class='cn'>FALSE</span><span class='op'>)</span> <span class='op'>{</span>
<span class='va'>path</span> <span class='op'>&lt;-</span> <span class='fu'><a href='downloadUMI4CexampleData.html'>downloadUMI4CexampleData</a></span><span class='op'>(</span><span class='op'>)</span>

<span class='va'>hg19_dpnii</span> <span class='op'>&lt;-</span> <span class='fu'><a href='digestGenome.html'>digestGenome</a></span><span class='op'>(</span>
    cut_pos <span class='op'>=</span> <span class='fl'>0</span>,
    res_enz <span class='op'>=</span> <span class='st'>"GATC"</span>,
    name_RE <span class='op'>=</span> <span class='st'>"DpnII"</span>,
    ref_gen <span class='op'>=</span> <span class='fu'>BSgenome.Hsapiens.UCSC.hg19</span><span class='fu'>::</span><span class='va'><a href='https://rdrr.io/pkg/BSgenome.Hsapiens.UCSC.hg19/man/package.html'>BSgenome.Hsapiens.UCSC.hg19</a></span>,
    out_path <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/file.path.html'>file.path</a></span><span class='op'>(</span><span class='va'>path</span>, <span class='st'>"digested_genome"</span><span class='op'>)</span>
<span class='op'>)</span>


<span class='va'>raw_dir</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/file.path.html'>file.path</a></span><span class='op'>(</span><span class='va'>path</span>, <span class='st'>"CIITA"</span>, <span class='st'>"fastq"</span><span class='op'>)</span>

<span class='fu'>contactsUMI4C</span><span class='op'>(</span>
    fastq_dir <span class='op'>=</span> <span class='va'>raw_dir</span>,
    wk_dir <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/file.path.html'>file.path</a></span><span class='op'>(</span><span class='va'>path</span>, <span class='st'>"CIITA"</span><span class='op'>)</span>,
    bait_seq <span class='op'>=</span> <span class='st'>"GGACAAGCTCCCTGCAACTCA"</span>,
    bait_pad <span class='op'>=</span> <span class='st'>"GGACTTGCA"</span>,
    res_enz <span class='op'>=</span> <span class='st'>"GATC"</span>,
    cut_pos <span class='op'>=</span> <span class='fl'>0</span>,
    digested_genome <span class='op'>=</span> <span class='va'>hg19_dpnii</span>,
    bowtie_index <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/file.path.html'>file.path</a></span><span class='op'>(</span><span class='va'>path</span>, <span class='st'>"ref_genome"</span>, <span class='st'>"ucsc.hg19.chr16"</span><span class='op'>)</span>,
    threads <span class='op'>=</span> <span class='fl'>1</span>,
    numb_reads <span class='op'>=</span> <span class='fl'>10e10</span>,
    ref_gen <span class='op'>=</span> <span class='fu'>BSgenome.Hsapiens.UCSC.hg19</span><span class='fu'>::</span><span class='va'><a href='https://rdrr.io/pkg/BSgenome.Hsapiens.UCSC.hg19/man/package.html'>BSgenome.Hsapiens.UCSC.hg19</a></span>,
    sel_seqname <span class='op'>=</span> <span class='st'>"chr16"</span>
<span class='op'>)</span>

<span class='fu'><a href='https://rdrr.io/r/base/unlink.html'>unlink</a></span><span class='op'>(</span><span class='va'>path</span>, recursive<span class='op'>=</span><span class='cn'>TRUE</span><span class='op'>)</span>
<span class='op'>}</span>
</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Mireia Ramos-Rodriguez, Marc Subirana-Granes.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.9000.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


