<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Analyzing UMI-4C data with UMI4Cats • UMI4Cats</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Analyzing UMI-4C data with UMI4Cats">
<meta property="og:description" content="UMI4Cats">
<meta property="og:image" content="/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">UMI4Cats</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.99.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/umi4cats-usage.html">Analyzing UMI-4C data with UMI4Cats</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://gitlab.com/pasquali-lab/umi4cats/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Analyzing UMI-4C data with UMI4Cats</h1>
                        <h4 class="author">Mireia Ramos-Rodríguez &amp; Marc Subirana-Granés</h4>
            
            <h4 class="date">2020-04-06</h4>
      
      <small class="dont-index">Source: <a href="https://gitlab.com/pasquali-lab/umi4cats/blob/master/vignettes/umi4cats-usage.Rmd"><code>vignettes/umi4cats-usage.Rmd</code></a></small>
      <div class="hidden name"><code>umi4cats-usage.Rmd</code></div>

    </div>

    
    
<div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h1>
<p><img src="../man/figures/logo.png"></p>
<p>Hello stranger! If you are here, that means you’ve successfully completed the UMI-4C protocol and got some sequencing results.</p>
<div class="sourceCode" id="cb1"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">UMI4Cats</span>)</pre></body></html></div>
</div>
<div id="quick-start" class="section level1">
<h1 class="hasAnchor">
<a href="#quick-start" class="anchor"></a>Quick start</h1>
<div id="processing-fastq-files" class="section level2">
<h2 class="hasAnchor">
<a href="#processing-fastq-files" class="anchor"></a>Processing FASTQ files</h2>
<div class="sourceCode" id="cb2"><html><body><pre class="r"><span class="co">## 0) Download example data</span>
<span class="no">path</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/downloadUMI4CexampleData.html">downloadUMI4CexampleData</a></span>()

<span class="co">## 1) Generate Digested genome ----------------------------</span>
<span class="co"># The selected RE in this case is DpnII (|GATC), so the cs5p is "" and cs3p is GATC</span>
<span class="no">hg19_dpnii</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/digestGenome.html">digestGenome</a></span>(<span class="kw">cut_pos</span> <span class="kw">=</span> <span class="fl">0</span>,
                           <span class="kw">res_enz</span> <span class="kw">=</span> <span class="st">"GATC"</span>,
                           <span class="kw">name_RE</span> <span class="kw">=</span> <span class="st">"DpnII"</span>,
                           <span class="kw">ref_gen</span> <span class="kw">=</span> <span class="kw pkg">BSgenome.Hsapiens.UCSC.hg19</span><span class="kw ns">::</span><span class="no"><a href="https://rdrr.io/pkg/BSgenome.Hsapiens.UCSC.hg19/man/package.html">BSgenome.Hsapiens.UCSC.hg19</a></span>,
                           <span class="kw">out_path</span> <span class="kw">=</span> <span class="st">"digested_genome/"</span>)

<span class="co">## 2) Process UMI-4C fastq files --------------------------</span>
<span class="no">raw_dir</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span>(<span class="no">path</span>, <span class="st">"SOCS1"</span>, <span class="st">"fastq"</span>)

<span class="fu"><a href="../reference/contactsUMI4C.html">contactsUMI4C</a></span>(<span class="kw">fastq_dir</span> <span class="kw">=</span> <span class="no">raw_dir</span>,
              <span class="kw">wk_dir</span> <span class="kw">=</span> <span class="st">"SOCS1"</span>,
              <span class="kw">bait_seq</span> <span class="kw">=</span> <span class="st">"CCCAAATCGCCCAGACCAG"</span>,
              <span class="kw">bait_pad</span> <span class="kw">=</span> <span class="st">"GCGCG"</span>,
              <span class="kw">res_enz</span> <span class="kw">=</span> <span class="st">"GATC"</span>,
              <span class="kw">cut_pos</span> <span class="kw">=</span> <span class="fl">0</span>,
              <span class="kw">digested_genome</span> <span class="kw">=</span> <span class="no">hg19_dpnii</span>,
              <span class="kw">bowtie_index</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span>(<span class="no">path</span>, <span class="st">"ref_genome"</span>, <span class="st">"ucsc.hg19.chr16"</span>),
              <span class="kw">ref_gen</span> <span class="kw">=</span> <span class="kw pkg">BSgenome.Hsapiens.UCSC.hg19</span><span class="kw ns">::</span><span class="no"><a href="https://rdrr.io/pkg/BSgenome.Hsapiens.UCSC.hg19/man/package.html">BSgenome.Hsapiens.UCSC.hg19</a></span>,
             <span class="kw">threads</span> <span class="kw">=</span> <span class="fl">5</span>)</pre></body></html></div>
</div>
<div id="analyzing-umi-4c-results" class="section level2">
<h2 class="hasAnchor">
<a href="#analyzing-umi-4c-results" class="anchor"></a>Analyzing UMI-4C results</h2>
<div class="sourceCode" id="cb3"><html><body><pre class="r"><span class="co">## Get filtering and alignment stats ---------</span>
<span class="fu"><a href="../reference/statsUMI4C.html">statsUMI4C</a></span>(<span class="kw">wk_dir</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">"extdata"</span>, <span class="st">"SOCS1"</span>,
                               <span class="kw">package</span><span class="kw">=</span><span class="st">"UMI4Cats"</span>))</pre></body></html></div>
<p><img src="umi4cats-usage_files/figure-html/analysis-quick-start-1.png" width="672"></p>
<div class="sourceCode" id="cb4"><html><body><pre class="r">
<span class="co"># Load sample processed file paths</span>
<span class="no">files</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html">list.files</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">"extdata"</span>, <span class="st">"SOCS1"</span>, <span class="st">"count"</span>,
                                <span class="kw">package</span><span class="kw">=</span><span class="st">"UMI4Cats"</span>),
                    <span class="kw">pattern</span><span class="kw">=</span><span class="st">"*_counts.tsv.gz"</span>,
                    <span class="kw">full.names</span><span class="kw">=</span><span class="fl">TRUE</span>)

<span class="co"># Create colData including all relevant information</span>
<span class="no">colData</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span>(<span class="kw">sampleID</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html">gsub</a></span>(<span class="st">"_counts.tsv.gz"</span>, <span class="st">""</span>, <span class="fu"><a href="https://rdrr.io/r/base/basename.html">basename</a></span>(<span class="no">files</span>)),
                      <span class="kw">file</span> <span class="kw">=</span> <span class="no">files</span>,
                      <span class="kw">stringsAsFactors</span><span class="kw">=</span><span class="fl">FALSE</span>)

<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">tidyr</span>)
<span class="no">colData</span> <span class="kw">&lt;-</span> <span class="no">colData</span> <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://tidyr.tidyverse.org/reference/separate.html">separate</a></span>(<span class="no">sampleID</span>,
           <span class="kw">into</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"condition"</span>, <span class="st">"replicate"</span>, <span class="st">"viewpoint"</span>),
           <span class="kw">remove</span><span class="kw">=</span><span class="fl">FALSE</span>)

<span class="co"># Load UMI-4C data and generate UMI4C object</span>
<span class="no">umi</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/UMI4C.html">makeUMI4C</a></span>(<span class="kw">colData</span><span class="kw">=</span><span class="no">colData</span>,
                 <span class="kw">viewpoint_name</span><span class="kw">=</span><span class="st">"SOCS1"</span>)

<span class="co"># Perform differential test</span>
<span class="no">umi</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/fisherUMI4C.html">fisherUMI4C</a></span>(<span class="no">umi</span>,
                   <span class="kw">filter_low</span> <span class="kw">=</span> <span class="fl">20</span>)

<span class="co"># Plot results</span>
<span class="fu"><a href="../reference/plotUMI4C.html">plotUMI4C</a></span>(<span class="no">umi</span>,
          <span class="kw">ylim</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">0</span>,<span class="fl">10</span>),
          <span class="kw">xlim</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">11e6</span>, <span class="fl">11.5e6</span>)
          )</pre></body></html></div>
<p><img src="umi4cats-usage_files/figure-html/analysis-quick-start-2.png" width="672"></p>
</div>
</div>
<div id="experiment-design" class="section level1">
<h1 class="hasAnchor">
<a href="#experiment-design" class="anchor"></a>Experiment design</h1>
</div>
<div id="processing-umi-4c-fastq-files" class="section level1">
<h1 class="hasAnchor">
<a href="#processing-umi-4c-fastq-files" class="anchor"></a>Processing UMI-4C FASTQ files</h1>
<div id="reference-genome-digestion" class="section level2">
<h2 class="hasAnchor">
<a href="#reference-genome-digestion" class="anchor"></a>Reference genome digestion</h2>
<p>For the processing of the UMI-4C FastQ files it is necessary to construct a digested genome using the appropriate restriction enzyme. The <code>UMI4Cats</code> package includes the <code><a href="../reference/digestGenome.html">digestGenome()</a></code> function to make this process as easy as possible. The function uses a <code>BSgenome</code> object<a href="#fn1" class="footnote-ref" id="fnref1"><sup>1</sup></a> as reference genome and given a cutting sequence for the description enzyme, digests the reference genome.</p>
<p>The cutting position is defined by <code>cut_pos</code> a zero-based numeric integrer indicating the nucleotide position where restriction enzyme cuts.</p>
<p>In the following table you can see two examples of the different cutting sequences for <em>DpnII</em>, <em>Csp6I</em> and <em>HindIII</em>.</p>
<table class="table">
<thead><tr class="header">
<th>Restriction enzyme</th>
<th>Restriction seq</th>
<th>cut_pos</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>DpnII</td>
<td><code>|GATC</code></td>
<td>0</td>
</tr>
<tr class="even">
<td>Csp6I</td>
<td><code>G|TAC</code></td>
<td>1</td>
</tr>
<tr class="odd">
<td>HindIII</td>
<td><code>A|AGCTT</code></td>
<td>1</td>
</tr>
</tbody>
</table>
<p>For this example, we are using the hg19 <code>BSGenome</code> object and we are going to digest it using <em>DpnII</em>.</p>
<div class="sourceCode" id="cb5"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">BSgenome.Hsapiens.UCSC.hg19</span>)
<span class="no">refgen</span> <span class="kw">&lt;-</span> <span class="no">BSgenome.Hsapiens.UCSC.hg19</span>

<span class="no">hg19_dpnii</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/digestGenome.html">digestGenome</a></span>(<span class="kw">res_enz</span> <span class="kw">=</span> <span class="st">"GATC"</span>,
                           <span class="kw">cut_pos</span> <span class="kw">=</span> <span class="fl">0</span>,
                           <span class="kw">name_RE</span> <span class="kw">=</span> <span class="st">"dpnII"</span>,
                           <span class="kw">ref_gen</span> <span class="kw">=</span> <span class="no">refgen</span>,
                           <span class="kw">sel_chr</span> <span class="kw">=</span> <span class="st">"chr16"</span>,
                           <span class="kw">out_path</span> <span class="kw">=</span> <span class="st">"digested_genome/"</span>)

<span class="no">hg19_dpnii</span>
<span class="co">#&gt; [1] "digested_genome//BSgenome.Hsapiens.UCSC.hg19_dpnII"</span></pre></body></html></div>
<p>Additionally, the <code><a href="../reference/digestGenome.html">digestGenome()</a></code> function returns invisibly the path of the generated file, so you can redirect it to a variable (in this case <code>hg19_dpnii</code>).</p>
</div>
<div id="demultiplex-fastq-files-for-different-baits" class="section level2">
<h2 class="hasAnchor">
<a href="#demultiplex-fastq-files-for-different-baits" class="anchor"></a>Demultiplex FastQ files for different baits</h2>
<p>UMI4Cats package provides a function for demultiplexing parser for demultiplexing paired-end FastQ files. Barcodes and bait names need to be provided as a data.frame as follows in the next example:</p>
<div class="sourceCode" id="cb6"><html><body><pre class="r"><span class="no">barcodes</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span>(<span class="kw">sample</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"KLK3"</span>, <span class="st">"KLK6"</span>, <span class="st">"KLK7"</span>, <span class="st">"KLK10"</span>),
                      <span class="kw">barcode</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"ATGGTCTGGGCGCTGTCTTG"</span>,
                                <span class="st">"TATTCTTCCTCAGCCCACATCTT"</span>,
                                <span class="st">"GGATGAAGATTTTGGAGCCCAGC"</span>,
                                <span class="st">"GGGCGGGGATTGAACGC"</span>))

<span class="fu"><a href="../reference/demultiplexFastq.html">demultiplexFastq</a></span>(<span class="kw">barcodes</span><span class="kw">=</span><span class="no">barcodes</span>,
                <span class="kw">fastq</span><span class="kw">=</span><span class="st">"~/samples/sample_1_R1.fastq.gz"</span>,
                <span class="kw">out_path</span><span class="kw">=</span><span class="st">"raw_fastq"</span>)</pre></body></html></div>
</div>
<div id="umi-4c-fastq-files-processing" class="section level2">
<h2 class="hasAnchor">
<a href="#umi-4c-fastq-files-processing" class="anchor"></a>UMI-4C FastQ files processing</h2>
<p>This step converts paired-end reads in the FastQ files to UMI counts in the genome fragments generated using the restriction enzymes. The basic function for doing this analysis is <code><a href="../reference/contactsUMI4C.html">contactsUMI4C()</a></code> and should be ran in samples generated with the same viewpoint and restriction enzyme.</p>
<p>The function will consider all FastQ files in the same folder <code>fastq_dir</code> to be part of the same experiment (viewpoint + restriction enzyme). The paired-end FASTQ files in this folder have to have one of the follow extensions: <code>_RX.fastq</code>, <code>_RX.fq</code>, <code>_RX.fq.gz</code> or <code>_RX.fastq.gz</code>.</p>
<p>For every experiment analysis, the user needs to define 3 different sequences: - <strong>Bait/viewpoint sequence</strong> (<code>bait_seq</code>). This is the downstream primer sequence (DS primer) that matches the sequence of the queried bait. - <strong>Padding sequence</strong> (<code>bait_pad</code>). The padding sequence corresponds to the nucleotides betweeen the DS primer end and the restriction enzyme site. - <strong>Restriction enzyme sequence</strong> (<code>res_enz</code>). This sequence is the restriction sequence recognized by the selected restriction enzyme.</p>
<p><img src="https://media.springernature.com/full/springer-static/image/art%3A10.1038%2Fnmeth.3922/MediaObjects/41592_2016_Article_BFnmeth3922_Fig1_HTML.jpg?as=webp"></p>
<p>Additionally, it is necessary to define the restiction enzyme cutting postion (<code>cut_pos</code>) in the restriction sequence which corresponds to the site where the restriction enzyme cuts the sequence. For example, for DpnII (<code>|GATC</code>) <code>cut_pos = 0</code> and for Csp6I (<code>G|TAC</code>) <code>cut_pos = 1</code>. The function will also need the path for the digested genome (<code>digested_genome</code>) generated with the function <code><a href="../reference/digestGenome.html">digestGenome()</a></code>.</p>
<p>As <code><a href="../reference/contactsUMI4C.html">contactsUMI4C()</a></code> performs the alignment using <a href="http://bowtie-bio.sourceforge.net/bowtie2/index.shtml">Bowtie2</a>, you will also need to provide a reference genome in fasta format and the corresponding indexes generated with bowtie2<a href="#fn2" class="footnote-ref" id="fnref2"><sup>2</sup></a>, which need to be located in the same folder as the fasta file.</p>
<div class="sourceCode" id="cb7"><html><body><pre class="r"><span class="co">## Download sample dataset</span>
<span class="no">path</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/downloadUMI4CexampleData.html">downloadUMI4CexampleData</a></span>()
<span class="co">#&gt; Will begin downloading datasets to /var/folders/6s/7ydjk_816j12lnyvsx292_qr0000gn/T//Rtmp8IGrTX/file62401e3c2641</span>
<span class="co">#&gt; untar: using cmd = '/usr/bin/tar -xf '/var/folders/6s/7ydjk_816j12lnyvsx292_qr0000gn/T//Rtmp8IGrTX/file62401e3c2641' -C './''</span>
<span class="co">#&gt; Done writing UMI4C example files to ./</span>

<span class="co"># hg19_dpnii &lt;- "digested_genome/BSgenome.Hsapiens.UCSC.hg19_dpnII/"</span>
<span class="co"># path &lt;- "UMI4Cats_data/"</span>

<span class="co">## Processing of fastq files</span>
<span class="no">raw_dir</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span>(<span class="no">path</span>, <span class="st">"SOCS1"</span>, <span class="st">"fastq"</span>)

<span class="fu"><a href="../reference/contactsUMI4C.html">contactsUMI4C</a></span>(<span class="kw">fastq_dir</span> <span class="kw">=</span> <span class="no">raw_dir</span>,
              <span class="kw">wk_dir</span> <span class="kw">=</span> <span class="st">"SOCS1"</span>,
              <span class="kw">file_pattern</span><span class="kw">=</span><span class="st">"ctrl_hi19_SOCS1"</span>, <span class="co"># Select only one sample to reduce time</span>
              <span class="kw">bait_seq</span> <span class="kw">=</span> <span class="st">"CCCAAATCGCCCAGACCAG"</span>,
              <span class="kw">bait_pad</span> <span class="kw">=</span> <span class="st">"GCGCG"</span>,
              <span class="kw">res_enz</span> <span class="kw">=</span> <span class="st">"GATC"</span>,
              <span class="kw">cut_pos</span> <span class="kw">=</span> <span class="fl">0</span>,
              <span class="kw">digested_genome</span> <span class="kw">=</span> <span class="no">hg19_dpnii</span>,
              <span class="kw">bowtie_index</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span>(<span class="no">path</span>, <span class="st">"ref_genome"</span>, <span class="st">"ucsc.hg19.chr16"</span>),
              <span class="kw">ref_gen</span> <span class="kw">=</span> <span class="kw pkg">BSgenome.Hsapiens.UCSC.hg19</span><span class="kw ns">::</span><span class="no"><a href="https://rdrr.io/pkg/BSgenome.Hsapiens.UCSC.hg19/man/package.html">BSgenome.Hsapiens.UCSC.hg19</a></span>,
              <span class="kw">threads</span> <span class="kw">=</span> <span class="fl">10</span>)
<span class="co">#&gt; </span>
<span class="co">#&gt; [2020-04-06 17:42:44] Starting prepUMI4C using:</span>
<span class="co">#&gt;  &gt; Fastq directory:</span>
<span class="co">#&gt;  ./UMI4Cats_data/SOCS1/fastq </span>
<span class="co">#&gt;  &gt; Work directory: SOCS1 </span>
<span class="co">#&gt;  &gt; Bait sequence: CCCAAATCGCCCAGACCAG </span>
<span class="co">#&gt;  &gt; Bait pad: GCGCG </span>
<span class="co">#&gt;  &gt; Restriction enzyme: GATC </span>
<span class="co">#&gt;  &gt; Number of reads loaded into memory: 1e+11</span>
<span class="co">#&gt; [2020-04-06 17:42:49] Finished sample ctrl_hi19_SOCS1</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; [2020-04-06 17:42:49] Starting splitUMI4C using:</span>
<span class="co">#&gt;  &gt; Work directory: SOCS1 </span>
<span class="co">#&gt;  &gt; Cut position: 0 </span>
<span class="co">#&gt;  &gt; Restriction enzyme: GATC </span>
<span class="co">#&gt;  &gt; Number of reads loaded into memory: 1e+11</span>
<span class="co">#&gt; [2020-04-06 17:42:50] Finished sample ctrl_hi19_SOCS1_R1.fq.gz</span>
<span class="co">#&gt; [2020-04-06 17:42:51] Finished sample ctrl_hi19_SOCS1_R2.fq.gz</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; [2020-04-06 17:42:51] Starting alignmentUMI4C using:</span>
<span class="co">#&gt;  &gt; Work directory: SOCS1 </span>
<span class="co">#&gt;  &gt; Viewpoint position: chr16:11349721-11349748 </span>
<span class="co">#&gt;  &gt; Reference genome: ./UMI4Cats_data/ref_genome/ucsc.hg19.chr16 </span>
<span class="co">#&gt;  &gt; Number of threads: 10</span>
<span class="co">#&gt; [2020-04-06 17:42:55] Finished sample ctrl_hi19_SOCS1_R1.fq</span>
<span class="co">#&gt; [2020-04-06 17:42:59] Finished sample ctrl_hi19_SOCS1_R2.fq</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; [2020-04-06 17:42:59] Starting counterUMI4C using:</span>
<span class="co">#&gt;  &gt; Work directory: SOCS1 </span>
<span class="co">#&gt;  &gt; Viewpoint position: chr16:11349721-11349748 </span>
<span class="co">#&gt;  &gt; Restriction enzyme: GATC </span>
<span class="co">#&gt;  &gt; Digested genome: digested_genome//BSgenome.Hsapiens.UCSC.hg19_dpnII</span>
<span class="co">#&gt; [2020-04-06 17:43:07] Finished sample ctrl_hi19_SOCS1</span></pre></body></html></div>
<p><code><a href="../reference/contactsUMI4C.html">contactsUMI4C()</a></code> will run the following processes sequentially:</p>
<ol style="list-style-type: decimal">
<li>FastQ files preparation (<code>prepUMI4C</code>).</li>
<li>Split reads at restriction sequences (<code>splitUMI4C</code>).</li>
<li>Align split reads to the reference genome (<code>alignmentUMI4C</code>).</li>
<li>Algorithm for collapsing reads coming from same molecule, returning all fragment ends 10Mb around the viewpoint(<code>counterUMI4C</code>).</li>
</ol>
<p>For the preparation and splitting, the FastQ file is loaded into memory. If you are having problems with the memory size, you can change the number of lines to be loaded using the <code>numb_reads</code> parameter. See <code><a href="../reference/contactsUMI4C.html">?contactsUMI4C</a></code> for more information.</p>
<p>Finally, the output (UMI-4C counts) is a tsv file saved in <code>wk_dir/count</code>. This file will be the input for the data analysis in the following section.</p>
</div>
<div id="quality-control-measures" class="section level2">
<h2 class="hasAnchor">
<a href="#quality-control-measures" class="anchor"></a>Quality control measures</h2>
<p>Once the contact analysis has been ran, the statistics of the UMI-4C filtering, aligment and final number of UMIs can be generated from the logs generated by the <code><a href="../reference/contactsUMI4C.html">contactsUMI4C()</a></code> function.</p>
<p>By using these logs, the function <code><a href="../reference/statsUMI4C.html">statsUMI4C()</a></code> will produce a summary plot .</p>
<div class="sourceCode" id="cb8"><html><body><pre class="r"><span class="fu"><a href="../reference/statsUMI4C.html">statsUMI4C</a></span>(<span class="kw">wk_dir</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">"extdata"</span>, <span class="st">"SOCS1"</span>,
                                <span class="kw">package</span><span class="kw">=</span><span class="st">"UMI4Cats"</span>))</pre></body></html></div>
<p><img src="umi4cats-usage_files/figure-html/unnamed-chunk-5-1.png" width="672"></p>
</div>
</div>
<div id="analyzing-umi-4c-data" class="section level1">
<h1 class="hasAnchor">
<a href="#analyzing-umi-4c-data" class="anchor"></a>Analyzing UMI-4C data</h1>
<p>For this part of the analysis each bait should be analyzed separately. A different <code>UMI4C</code> object should be created for each analyzed bait.</p>
<div id="construct-the-umi4c-object" class="section level2">
<h2 class="hasAnchor">
<a href="#construct-the-umi4c-object" class="anchor"></a>Construct the <code>UMI4C</code> object</h2>
<p>The first step of the UMI-4C data analysis consists on loading the tables generated by the function <code>umi4CatsContacts</code> and use them to construct a <code>UMI4C</code>object. All these steps are perform automatically by the <code>makeUMI4C</code> functions.</p>
<p>The <code>makeUMI4C</code> will need as imput a data frame (<code>colData</code>) containing all relevant experiment information that will be needed for analyzing the data later on. The mandatory columns that this data frame should have are:</p>
<ol style="list-style-type: decimal">
<li>
<code>sampleID</code>: Unique identifier for the sample.</li>
<li>
<code>replicate</code>: Replicate identifier or number.</li>
<li>
<code>condition</code>: Condition for performing the diferential analysis. For example: “control” and “treatment”, two different cell types, etc. The condition column should only have <strong>two</strong> different values. If more condition variables are provided, the differential analysis will fail.</li>
<li>
<code>file</code>: Complete path and filename where the tsv files generated by <code>umi4CatsContacts</code> are saved.</li>
</ol>
<p>You can also include additional columns to <code>colData</code>, but the ones mentioned above are mandatory.</p>
<p>By default, the function will combine the samples using the <strong>condition</strong> column from <code>colData(umi)</code> (the colData data frame you provided for generating the <code>UMI4C</code> object), but you can change this behaviour using the <code>grouping</code> argument. You can give it one or several column names and will to the grouping based on those variables. For example, if you set <code>grouping="sampleID"</code> it will plot the individual profiles.</p>
<p>Additionally, the <code>makeUMI4C</code> function also contains other arguments that can be used if you want to tweak the default parameters of the analysis. See <code><a href="../reference/UMI4C.html">?makeUMI4C</a></code> to have a complete list and description of all the arguments.</p>
<div class="sourceCode" id="cb9"><html><body><pre class="r"><span class="co"># Load sample processed file paths</span>
<span class="no">files</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html">list.files</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">"extdata"</span>, <span class="st">"SOCS1"</span>, <span class="st">"count"</span>,
                                <span class="kw">package</span> <span class="kw">=</span> <span class="st">"UMI4Cats"</span>),
                    <span class="kw">pattern</span> <span class="kw">=</span> <span class="st">"*_counts.tsv"</span>,
                    <span class="kw">full.names</span> <span class="kw">=</span> <span class="fl">TRUE</span>)

<span class="co"># Create colData including all relevant information</span>
<span class="no">colData</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span>(<span class="kw">sampleID</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html">gsub</a></span>(<span class="st">"_counts.tsv.gz"</span>, <span class="st">""</span>, <span class="fu"><a href="https://rdrr.io/r/base/basename.html">basename</a></span>(<span class="no">files</span>)),
                      <span class="kw">file</span> <span class="kw">=</span> <span class="no">files</span>,
                      <span class="kw">stringsAsFactors</span> <span class="kw">=</span> <span class="fl">FALSE</span>)

<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">tidyr</span>)
<span class="no">colData</span> <span class="kw">&lt;-</span> <span class="no">colData</span> <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://tidyr.tidyverse.org/reference/separate.html">separate</a></span>(<span class="no">sampleID</span>,
           <span class="kw">into</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"condition"</span>, <span class="st">"replicate"</span>, <span class="st">"viewpoint"</span>),
           <span class="kw">remove</span> <span class="kw">=</span> <span class="fl">FALSE</span>)

<span class="co"># Load UMI-4C data and generate UMI4C object</span>
<span class="no">umi</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/UMI4C.html">makeUMI4C</a></span>(<span class="kw">colData</span> <span class="kw">=</span> <span class="no">colData</span>,
                 <span class="kw">viewpoint_name</span> <span class="kw">=</span> <span class="st">"SOCS1"</span>)

<span class="no">umi</span>
<span class="co">#&gt; class: UMI4C </span>
<span class="co">#&gt; dim: 2895 2 </span>
<span class="co">#&gt; metadata(7): bait scales ... region ref_umi4c</span>
<span class="co">#&gt; assays(6): umis norm_mat ... scale sd</span>
<span class="co">#&gt; rownames(2895): frag_26371 frag_26372 ... frag_29268 frag_29269</span>
<span class="co">#&gt; rowData names(2): id_contact position</span>
<span class="co">#&gt; colnames(2): ctrl cyt</span>
<span class="co">#&gt; colData names(5): condition sampleID replicate viewpoint file</span></pre></body></html></div>
<p>This function contains several basic steps for preparing and analzying the UMI-4C data:</p>
<ol style="list-style-type: decimal">
<li>Removing fragment ends around the bait. The default value can be changed using the <code>bait_exclusion</code> argument.</li>
<li>Focusing the scope of the analysis in a specific region around the bait. The default value can be changed using the <code>bait_expansion</code> argument.</li>
<li>Obtaining the normalization matrices that will be used to scale the samples to the reference sample. If you want to avoid this normalization step, you can set <code>normalized</code> to <code>FALSE</code>.</li>
<li>Calculating the domainograms for each sample.</li>
<li>Calculating the adaptative trend for each sample.</li>
</ol>
</div>
<div id="performing-differential-analysis" class="section level2">
<h2 class="hasAnchor">
<a href="#performing-differential-analysis" class="anchor"></a>Performing differential analysis</h2>
<p>Once the <code>UMI4C</code> object is generated, one can perform a differential analysis using the <code><a href="../reference/fisherUMI4C.html">fisherUMI4C()</a></code> function. This will take as input the <code>UMI4C</code> object. If no other argument is provided, it will bin the region around the viewpoint and perform Fisher’s Exact tests in those windows, comparing the number of reads in each condition at that specific window with the number of reads in the studied region. Finally, p-values will be corrected using the FDR method and regions with adjusted p-values &gt; 0.05 will be considered signficantly different. The <code>filter_low</code> arguments determines which windows are tested: if they do not contain at least the number of UMIs in <code>filter_low</code>, they will not be tested. If you want to disable the filtering you can set <code>filter_low=FALSE</code>. You can also change the size of the windows used for this analysis using the <code>window_size</code> argument. Check <code>?fisherUMI4()</code> for more information.</p>
<p>Additionally, one can select the regions of interest to query for differential contacts using the <code>query_regions</code> argument.</p>
<p>Results from this test can be retrieved using the <code><a href="../reference/UMI4C-methods.html">results()</a></code> method defined for <code>UMI4C</code> objects.</p>
<div class="sourceCode" id="cb10"><html><body><pre class="r"><span class="no">umi_dif</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/fisherUMI4C.html">fisherUMI4C</a></span>(<span class="no">umi</span>,
                   <span class="kw">filter_low</span><span class="kw">=</span><span class="fl">30</span>)

<span class="fu"><a href="../reference/UMI4C-methods.html">results</a></span>(<span class="no">umi_dif</span>)
<span class="co">#&gt; GRanges object with 198 ranges and 6 metadata columns:</span>
<span class="co">#&gt;         seqnames            ranges strand |          id    pvalue odds_ratio</span>
<span class="co">#&gt;            &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt; &lt;numeric&gt;  &lt;numeric&gt;</span>
<span class="co">#&gt;     [1]    chr16 10849721-10854720      * |    region_1      &lt;NA&gt;       &lt;NA&gt;</span>
<span class="co">#&gt;     [2]    chr16 10854721-10859720      * |    region_2      &lt;NA&gt;       &lt;NA&gt;</span>
<span class="co">#&gt;     [3]    chr16 10859721-10864720      * |    region_3      &lt;NA&gt;       &lt;NA&gt;</span>
<span class="co">#&gt;     [4]    chr16 10864721-10869720      * |    region_4      &lt;NA&gt;       &lt;NA&gt;</span>
<span class="co">#&gt;     [5]    chr16 10869721-10874720      * |    region_5      &lt;NA&gt;       &lt;NA&gt;</span>
<span class="co">#&gt;     ...      ...               ...    ... .         ...       ...        ...</span>
<span class="co">#&gt;   [194]    chr16 11824721-11829720      * |  region_194      &lt;NA&gt;       &lt;NA&gt;</span>
<span class="co">#&gt;   [195]    chr16 11829721-11834720      * |  region_195      &lt;NA&gt;       &lt;NA&gt;</span>
<span class="co">#&gt;   [196]    chr16 11834721-11839720      * |  region_196      &lt;NA&gt;       &lt;NA&gt;</span>
<span class="co">#&gt;   [197]    chr16 11839721-11844720      * |  region_197      &lt;NA&gt;       &lt;NA&gt;</span>
<span class="co">#&gt;   [198]    chr16 11844721-11849720      * |  region_198      &lt;NA&gt;       &lt;NA&gt;</span>
<span class="co">#&gt;         log2_odds_ratio      padj      sign</span>
<span class="co">#&gt;               &lt;numeric&gt; &lt;numeric&gt; &lt;logical&gt;</span>
<span class="co">#&gt;     [1]            &lt;NA&gt;      &lt;NA&gt;      &lt;NA&gt;</span>
<span class="co">#&gt;     [2]            &lt;NA&gt;      &lt;NA&gt;      &lt;NA&gt;</span>
<span class="co">#&gt;     [3]            &lt;NA&gt;      &lt;NA&gt;      &lt;NA&gt;</span>
<span class="co">#&gt;     [4]            &lt;NA&gt;      &lt;NA&gt;      &lt;NA&gt;</span>
<span class="co">#&gt;     [5]            &lt;NA&gt;      &lt;NA&gt;      &lt;NA&gt;</span>
<span class="co">#&gt;     ...             ...       ...       ...</span>
<span class="co">#&gt;   [194]            &lt;NA&gt;      &lt;NA&gt;      &lt;NA&gt;</span>
<span class="co">#&gt;   [195]            &lt;NA&gt;      &lt;NA&gt;      &lt;NA&gt;</span>
<span class="co">#&gt;   [196]            &lt;NA&gt;      &lt;NA&gt;      &lt;NA&gt;</span>
<span class="co">#&gt;   [197]            &lt;NA&gt;      &lt;NA&gt;      &lt;NA&gt;</span>
<span class="co">#&gt;   [198]            &lt;NA&gt;      &lt;NA&gt;      &lt;NA&gt;</span>
<span class="co">#&gt;   -------</span>
<span class="co">#&gt;   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span></pre></body></html></div>
</div>
<div id="plotting-the-umi4c-object" class="section level2">
<h2 class="hasAnchor">
<a href="#plotting-the-umi4c-object" class="anchor"></a>Plotting the <code>UMI4C</code> object</h2>
<p>Once the <code>UMI4C</code> object is created, you can plot it using the <code>plotUMI4C</code> function.</p>
<div class="sourceCode" id="cb11"><html><body><pre class="r"><span class="fu"><a href="../reference/plotUMI4C.html">plotUMI4C</a></span>(<span class="no">umi</span>)</pre></body></html></div>
<p><img src="umi4cats-usage_files/figure-html/unnamed-chunk-8-1.png" width="700"></p>
<p>If the <code>UMI4C</code> object contains information on the differential contacts, this data will be shown in the plot as well.</p>
<div class="sourceCode" id="cb12"><html><body><pre class="r"><span class="fu"><a href="../reference/plotUMI4C.html">plotUMI4C</a></span>(<span class="no">umi_dif</span>)</pre></body></html></div>
<p><img src="umi4cats-usage_files/figure-html/unnamed-chunk-9-1.png" width="700"></p>
</div>
<div id="accessing-information-from-the-umi4c-object" class="section level2">
<h2 class="hasAnchor">
<a href="#accessing-information-from-the-umi4c-object" class="anchor"></a>Accessing information from the <code>UMI4C</code> object</h2>
<p>The usual accessor functions from the SummarizedExperiment-class also work with the UMI-4C class (for example: <code>assay</code>, <code>colData</code>, etc.). Other additional accessors have been created to retrieve different information:</p>
<ul>
<li>
<code><a href="../reference/UMI4C-methods.html">dgram()</a></code>. Get a list of the domaingorams.</li>
<li>
<code><a href="../reference/UMI4C-methods.html">bait()</a></code>. Retrieve a GRanges object with the bait position.</li>
<li>
<code><a href="../reference/UMI4C-methods.html">trend()</a></code>. Obtain a data.frame in long format with the adaptative smoothen trend.</li>
<li>
<code><a href="../reference/UMI4C-methods.html">results()</a></code>. Retrieve results from the differential analysis. This only works if a differential analysis has been performed on the UMI4C object.</li>
</ul>
<div class="sourceCode" id="cb13"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="fu">assay</span>(<span class="no">umi</span>))
<span class="co">#&gt;            ctrl cyt</span>
<span class="co">#&gt; frag_26371    0   0</span>
<span class="co">#&gt; frag_26372    0   0</span>
<span class="co">#&gt; frag_26373    0   0</span>
<span class="co">#&gt; frag_26374    0   0</span>
<span class="co">#&gt; frag_26375    0   0</span>
<span class="co">#&gt; frag_26376    0   0</span>

<span class="fu">colData</span>(<span class="no">umi</span>)
<span class="co">#&gt; DataFrame with 2 rows and 5 columns</span>
<span class="co">#&gt;        condition                         sampleID   replicate    viewpoint</span>
<span class="co">#&gt;      &lt;character&gt;                      &lt;character&gt; &lt;character&gt;  &lt;character&gt;</span>
<span class="co">#&gt; ctrl        ctrl ctrl_hi19_SOCS1, ctrl_hi22_SOCS1  hi19, hi22 SOCS1, SOCS1</span>
<span class="co">#&gt; cyt          cyt   cyt_hi19_SOCS1, cyt_hi22_SOCS1  hi19, hi22 SOCS1, SOCS1</span>
<span class="co">#&gt;                                                                                                                                                                                                                                                      file</span>
<span class="co">#&gt;                                                                                                                                                                                                                                               &lt;character&gt;</span>
<span class="co">#&gt; ctrl /Library/Frameworks/R.framework/Versions/3.6/Resources/library/UMI4Cats/extdata/SOCS1/count/ctrl_hi19_SOCS1_counts.tsv.gz, /Library/Frameworks/R.framework/Versions/3.6/Resources/library/UMI4Cats/extdata/SOCS1/count/ctrl_hi22_SOCS1_counts.tsv.gz</span>
<span class="co">#&gt; cyt    /Library/Frameworks/R.framework/Versions/3.6/Resources/library/UMI4Cats/extdata/SOCS1/count/cyt_hi19_SOCS1_counts.tsv.gz, /Library/Frameworks/R.framework/Versions/3.6/Resources/library/UMI4Cats/extdata/SOCS1/count/cyt_hi22_SOCS1_counts.tsv.gz</span>

<span class="fu"><a href="../reference/UMI4C-methods.html">dgram</a></span>(<span class="no">umi</span>)
<span class="co">#&gt; List of length 2</span>
<span class="co">#&gt; names(2): ctrl cyt</span>

<span class="fu"><a href="../reference/UMI4C-methods.html">bait</a></span>(<span class="no">umi</span>)
<span class="co">#&gt; GRanges object with 1 range and 1 metadata column:</span>
<span class="co">#&gt;                   seqnames    ranges strand |        name</span>
<span class="co">#&gt;                      &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;</span>
<span class="co">#&gt;   ctrl_hi19_SOCS1    chr16  11349721      * |       SOCS1</span>
<span class="co">#&gt;   -------</span>
<span class="co">#&gt;   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span>

<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="fu"><a href="../reference/UMI4C-methods.html">trend</a></span>(<span class="no">umi</span>))
<span class="co">#&gt;   geo_coord     trend         sd scale id_contact sample</span>
<span class="co">#&gt; 1  10973745 0.3900000 0.07211103   150 frag_26766   ctrl</span>
<span class="co">#&gt; 2  10974076 0.3859060 0.07197185   149 frag_26767   ctrl</span>
<span class="co">#&gt; 3  10974410 0.3885135 0.07245814   148 frag_26768   ctrl</span>
<span class="co">#&gt; 4  10974743 0.3911565 0.07295106   147 frag_26769   ctrl</span>
<span class="co">#&gt; 5  10975075 0.3938356 0.07345072   146 frag_26770   ctrl</span>
<span class="co">#&gt; 6  10975404 0.3965517 0.07395728   145 frag_26771   ctrl</span>
<span class="co">#&gt;                           sampleID  replicate    viewpoint</span>
<span class="co">#&gt; 1 ctrl_hi19_SOCS1, ctrl_hi22_SOCS1 hi19, hi22 SOCS1, SOCS1</span>
<span class="co">#&gt; 2 ctrl_hi19_SOCS1, ctrl_hi22_SOCS1 hi19, hi22 SOCS1, SOCS1</span>
<span class="co">#&gt; 3 ctrl_hi19_SOCS1, ctrl_hi22_SOCS1 hi19, hi22 SOCS1, SOCS1</span>
<span class="co">#&gt; 4 ctrl_hi19_SOCS1, ctrl_hi22_SOCS1 hi19, hi22 SOCS1, SOCS1</span>
<span class="co">#&gt; 5 ctrl_hi19_SOCS1, ctrl_hi22_SOCS1 hi19, hi22 SOCS1, SOCS1</span>
<span class="co">#&gt; 6 ctrl_hi19_SOCS1, ctrl_hi22_SOCS1 hi19, hi22 SOCS1, SOCS1</span>
<span class="co">#&gt;                                                                                                                                                                                                                                                   file</span>
<span class="co">#&gt; 1 /Library/Frameworks/R.framework/Versions/3.6/Resources/library/UMI4Cats/extdata/SOCS1/count/ctrl_hi19_SOCS1_counts.tsv.gz, /Library/Frameworks/R.framework/Versions/3.6/Resources/library/UMI4Cats/extdata/SOCS1/count/ctrl_hi22_SOCS1_counts.tsv.gz</span>
<span class="co">#&gt; 2 /Library/Frameworks/R.framework/Versions/3.6/Resources/library/UMI4Cats/extdata/SOCS1/count/ctrl_hi19_SOCS1_counts.tsv.gz, /Library/Frameworks/R.framework/Versions/3.6/Resources/library/UMI4Cats/extdata/SOCS1/count/ctrl_hi22_SOCS1_counts.tsv.gz</span>
<span class="co">#&gt; 3 /Library/Frameworks/R.framework/Versions/3.6/Resources/library/UMI4Cats/extdata/SOCS1/count/ctrl_hi19_SOCS1_counts.tsv.gz, /Library/Frameworks/R.framework/Versions/3.6/Resources/library/UMI4Cats/extdata/SOCS1/count/ctrl_hi22_SOCS1_counts.tsv.gz</span>
<span class="co">#&gt; 4 /Library/Frameworks/R.framework/Versions/3.6/Resources/library/UMI4Cats/extdata/SOCS1/count/ctrl_hi19_SOCS1_counts.tsv.gz, /Library/Frameworks/R.framework/Versions/3.6/Resources/library/UMI4Cats/extdata/SOCS1/count/ctrl_hi22_SOCS1_counts.tsv.gz</span>
<span class="co">#&gt; 5 /Library/Frameworks/R.framework/Versions/3.6/Resources/library/UMI4Cats/extdata/SOCS1/count/ctrl_hi19_SOCS1_counts.tsv.gz, /Library/Frameworks/R.framework/Versions/3.6/Resources/library/UMI4Cats/extdata/SOCS1/count/ctrl_hi22_SOCS1_counts.tsv.gz</span>
<span class="co">#&gt; 6 /Library/Frameworks/R.framework/Versions/3.6/Resources/library/UMI4Cats/extdata/SOCS1/count/ctrl_hi19_SOCS1_counts.tsv.gz, /Library/Frameworks/R.framework/Versions/3.6/Resources/library/UMI4Cats/extdata/SOCS1/count/ctrl_hi22_SOCS1_counts.tsv.gz</span>

<span class="fu"><a href="../reference/UMI4C-methods.html">results</a></span>(<span class="no">umi_dif</span>)
<span class="co">#&gt; GRanges object with 198 ranges and 6 metadata columns:</span>
<span class="co">#&gt;         seqnames            ranges strand |          id    pvalue odds_ratio</span>
<span class="co">#&gt;            &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt; &lt;numeric&gt;  &lt;numeric&gt;</span>
<span class="co">#&gt;     [1]    chr16 10849721-10854720      * |    region_1      &lt;NA&gt;       &lt;NA&gt;</span>
<span class="co">#&gt;     [2]    chr16 10854721-10859720      * |    region_2      &lt;NA&gt;       &lt;NA&gt;</span>
<span class="co">#&gt;     [3]    chr16 10859721-10864720      * |    region_3      &lt;NA&gt;       &lt;NA&gt;</span>
<span class="co">#&gt;     [4]    chr16 10864721-10869720      * |    region_4      &lt;NA&gt;       &lt;NA&gt;</span>
<span class="co">#&gt;     [5]    chr16 10869721-10874720      * |    region_5      &lt;NA&gt;       &lt;NA&gt;</span>
<span class="co">#&gt;     ...      ...               ...    ... .         ...       ...        ...</span>
<span class="co">#&gt;   [194]    chr16 11824721-11829720      * |  region_194      &lt;NA&gt;       &lt;NA&gt;</span>
<span class="co">#&gt;   [195]    chr16 11829721-11834720      * |  region_195      &lt;NA&gt;       &lt;NA&gt;</span>
<span class="co">#&gt;   [196]    chr16 11834721-11839720      * |  region_196      &lt;NA&gt;       &lt;NA&gt;</span>
<span class="co">#&gt;   [197]    chr16 11839721-11844720      * |  region_197      &lt;NA&gt;       &lt;NA&gt;</span>
<span class="co">#&gt;   [198]    chr16 11844721-11849720      * |  region_198      &lt;NA&gt;       &lt;NA&gt;</span>
<span class="co">#&gt;         log2_odds_ratio      padj      sign</span>
<span class="co">#&gt;               &lt;numeric&gt; &lt;numeric&gt; &lt;logical&gt;</span>
<span class="co">#&gt;     [1]            &lt;NA&gt;      &lt;NA&gt;      &lt;NA&gt;</span>
<span class="co">#&gt;     [2]            &lt;NA&gt;      &lt;NA&gt;      &lt;NA&gt;</span>
<span class="co">#&gt;     [3]            &lt;NA&gt;      &lt;NA&gt;      &lt;NA&gt;</span>
<span class="co">#&gt;     [4]            &lt;NA&gt;      &lt;NA&gt;      &lt;NA&gt;</span>
<span class="co">#&gt;     [5]            &lt;NA&gt;      &lt;NA&gt;      &lt;NA&gt;</span>
<span class="co">#&gt;     ...             ...       ...       ...</span>
<span class="co">#&gt;   [194]            &lt;NA&gt;      &lt;NA&gt;      &lt;NA&gt;</span>
<span class="co">#&gt;   [195]            &lt;NA&gt;      &lt;NA&gt;      &lt;NA&gt;</span>
<span class="co">#&gt;   [196]            &lt;NA&gt;      &lt;NA&gt;      &lt;NA&gt;</span>
<span class="co">#&gt;   [197]            &lt;NA&gt;      &lt;NA&gt;      &lt;NA&gt;</span>
<span class="co">#&gt;   [198]            &lt;NA&gt;      &lt;NA&gt;      &lt;NA&gt;</span>
<span class="co">#&gt;   -------</span>
<span class="co">#&gt;   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span></pre></body></html></div>
</div>
</div>
<div class="footnotes">
<hr>
<ol>
<li id="fn1"><p>More information on <code>BSgenome</code> package and objects can be found <a href="https://bioconductor.org/packages/release/bioc/html/BSgenome.html">here</a>)<a href="#fnref1" class="footnote-back">↩</a></p></li>
<li id="fn2"><p>See the <a href="http://bowtie-bio.sourceforge.net/bowtie2/manual.shtml#getting-started-with-bowtie-2-lambda-phage-example">getting started section</a> on the Bowtie2 page for more information on how to generate the index for the reference genome.<a href="#fnref2" class="footnote-back">↩</a></p></li>
</ol>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Mireia Ramos-Rodriguez, Marc Subirana-Granes.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.0.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
