<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Analyzing UMI-4C data with UMI4Cats • UMI4Cats</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Analyzing UMI-4C data with UMI4Cats">
<meta property="og:description" content="UMI4Cats">
<meta property="og:image" content="/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">UMI4Cats</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.99.14</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/UMI4Cats.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/Pasquali-lab/UMI4Cats/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="UMI4Cats_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Analyzing UMI-4C data with UMI4Cats</h1>
                        <h4 class="author">Mireia Ramos-Rodríguez and Marc Subirana-Granés</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/Pasquali-lab/UMI4Cats/blob/master/vignettes/UMI4Cats.Rmd"><code>vignettes/UMI4Cats.Rmd</code></a></small>
      <div class="hidden name"><code>UMI4Cats.Rmd</code></div>

    </div>

    
    
<p><img src="../man/figures/UMI4Cats.png" width="10%" style="display: block; margin: auto;"></p>
<div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h1>
<p>Hello stranger! If you are here, that means you’ve successfully completed the UMI-4C protocol and got some sequencing results! This vignette has the objective to guide you through performing a simple analysis of your brand-new UMI-4C contact data. Let’s dive in!</p>
<div class="sourceCode" id="cb1"><pre class="downlit">
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/Pasquali-lab/UMI4Cats">UMI4Cats</a></span><span class="op">)</span></pre></div>
<div id="overview-of-the-package" class="section level2">
<h2 class="hasAnchor">
<a href="#overview-of-the-package" class="anchor"></a>Overview of the package</h2>
<div class="figure" style="text-align: center">
<img src="figures/scheme.png" alt="Overview of the different functions included in the UMI4Cats package to analyze UMI-4C data." width="60%"><p class="caption">
Overview of the different functions included in the UMI4Cats package to analyze UMI-4C data.
</p>
</div>
</div>
<div id="about-the-experimental-design" class="section level2">
<h2 class="hasAnchor">
<a href="#about-the-experimental-design" class="anchor"></a>About the experimental design</h2>
<p>One of the strengths of the UMI-4C assay <span class="citation">(Schwartzman et al. 2016)</span> is that of reducing the PCR duplication bias allowing a more accurate quantification of chromatin interactions. For this reason, UMI-4C is mostly used when trying to compare changes in chromatin interactions between two conditions, cell types or developmental stages.</p>
<p>Taking into account this main application, UMI4Cats has been developed to facilitate the differential analysis between conditions at a given viewpoint of interest. When analyzing your data with this package, you should take into account the following points:</p>
<ul>
<li><p>Each analysis (and <code>UMI4C</code> object) should correspond to the <strong>same viewpoint</strong>. If you are analyzing different viewpoints in the same or different loci, you need to analyze them separately.</p></li>
<li><p>The UMI4Cats package is mostly oriented to the performance of differential analysis. For this reason, in the analysis step, it requires to <strong>annotate each sample “replicate” and “condition”</strong>. However, if you only have one replicate or condition you will be able to perform the analysis and visualization of the data but not the differential analysis.</p></li>
<li><p>When performing the differential analysis, UMI4Cats is only able to deal with a <strong>condition with 2 different levels</strong>. If you have more than two conditions, you should produce different UMI4C objects and perform pairwise comparisons.</p></li>
</ul>
</div>
<div id="about-the-example-datasets" class="section level2">
<h2 class="hasAnchor">
<a href="#about-the-example-datasets" class="anchor"></a>About the example datasets</h2>
<p>The datasets used in this vignette (from <span class="citation">Ramos-Rodríguez et al. (2019)</span>) are available for download if you want to reproduce the contents of this vignette.</p>
<p>Briefly, the datasets correspond to human pancreatic islets exposed (<code>cyt</code>) or not (<code>ctrl</code>) to pro-inflammatory cytokines for 48 hours. In this example we are using the UMI-4C data generated from two different biological replicates (HI24 and HI32) using the promoter of the <em>CIITA</em> gene as viewpoint.</p>
<p>Sample datasets can be downloaded using the <code><a href="../reference/downloadUMI4CexampleData.html">downloadUMI4CexampleData()</a></code> function. When used without arguments, will download the full sample fastq files containing 200K reads. However, in order to reduce computing time, the <em>Processing UMI-4C FASTQ files</em> section in this vignette use a reduced sample file containing 100 reads, which can be downloaded using <code><a href="../reference/downloadUMI4CexampleData.html">downloadUMI4CexampleData(reduced = TRUE)</a></code>. The following sections addressing analysis and visualization of such data use the processed files from the 200K fastq files, which are also included inside the package and can be accessed using the <code>system.file</code> functions.</p>
</div>
</div>
<div id="quick-start" class="section level1">
<h1 class="hasAnchor">
<a href="#quick-start" class="anchor"></a>Quick start</h1>
<p>In this section we summarize a complete analysis using the examples provided in this package.</p>
<div class="sourceCode" id="cb2"><pre class="downlit">
<span class="co">## 0) Download example data -------------------------------</span>
<span class="va">path</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/downloadUMI4CexampleData.html">downloadUMI4CexampleData</a></span><span class="op">(</span><span class="op">)</span>

<span class="co">## 1) Generate Digested genome ----------------------------</span>
<span class="co"># The selected RE in this case is DpnII (|GATC), so the cut_pos is 0, and the res_enz "GATC".</span>
<span class="va">hg19_dpnii</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/digestGenome.html">digestGenome</a></span><span class="op">(</span>
    cut_pos <span class="op">=</span> <span class="fl">0</span>,
    res_enz <span class="op">=</span> <span class="st">"GATC"</span>,
    name_RE <span class="op">=</span> <span class="st">"DpnII"</span>,
    ref_gen <span class="op">=</span> <span class="fu">BSgenome.Hsapiens.UCSC.hg19</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/BSgenome.Hsapiens.UCSC.hg19/man/package.html">BSgenome.Hsapiens.UCSC.hg19</a></span>,
    out_path <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/tempfile.html">tempdir</a></span><span class="op">(</span><span class="op">)</span>, <span class="st">"digested_genome/"</span><span class="op">)</span>
<span class="op">)</span>

<span class="co">## 2) Process UMI-4C fastq files --------------------------</span>
<span class="va">raw_dir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span><span class="op">(</span><span class="va">path</span>, <span class="st">"CIITA"</span>, <span class="st">"fastq"</span><span class="op">)</span>

<span class="fu"><a href="../reference/contactsUMI4C.html">contactsUMI4C</a></span><span class="op">(</span>
    fastq_dir <span class="op">=</span> <span class="va">raw_dir</span>,
    wk_dir <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span><span class="op">(</span><span class="va">path</span>, <span class="st">"CIITA"</span><span class="op">)</span>,
    bait_seq <span class="op">=</span> <span class="st">"GGACAAGCTCCCTGCAACTCA"</span>,
    bait_pad <span class="op">=</span> <span class="st">"GGACTTGCA"</span>,
    res_enz <span class="op">=</span> <span class="st">"GATC"</span>,
    cut_pos <span class="op">=</span> <span class="fl">0</span>,
    digested_genome <span class="op">=</span> <span class="va">hg19_dpnii</span>,
    bowtie_index <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span><span class="op">(</span><span class="va">path</span>, <span class="st">"ref_genome"</span>, <span class="st">"ucsc.hg19.chr16"</span><span class="op">)</span>,
    ref_gen <span class="op">=</span> <span class="fu">BSgenome.Hsapiens.UCSC.hg19</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/BSgenome.Hsapiens.UCSC.hg19/man/package.html">BSgenome.Hsapiens.UCSC.hg19</a></span>,
    threads <span class="op">=</span> <span class="fl">5</span>
<span class="op">)</span>

<span class="co">## 3) Get filtering and alignment stats -------------------</span>
<span class="fu"><a href="../reference/statsUMI4C.html">statsUMI4C</a></span><span class="op">(</span>wk_dir <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span><span class="op">(</span><span class="va">path</span>, <span class="st">"CIITA"</span><span class="op">)</span><span class="op">)</span>

<span class="co">## 4) Analyze the results ---------------------------------</span>
<span class="co"># Load sample processed file paths</span>
<span class="va">files</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html">list.files</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span><span class="op">(</span><span class="va">path</span>, <span class="st">"CIITA"</span>, <span class="st">"count"</span><span class="op">)</span>,
                    pattern <span class="op">=</span> <span class="st">"*_counts.tsv.gz"</span>,
                    full.names <span class="op">=</span> <span class="cn">TRUE</span>
<span class="op">)</span>

<span class="co"># Create colData including all relevant information</span>
<span class="va">colData</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span>
    sampleID <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html">gsub</a></span><span class="op">(</span><span class="st">"_counts.tsv.gz"</span>, <span class="st">""</span>, <span class="fu"><a href="https://rdrr.io/r/base/basename.html">basename</a></span><span class="op">(</span><span class="va">files</span><span class="op">)</span><span class="op">)</span>,
    file <span class="op">=</span> <span class="va">files</span>,
    stringsAsFactors <span class="op">=</span> <span class="cn">FALSE</span>
<span class="op">)</span>

<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyr.tidyverse.org">tidyr</a></span><span class="op">)</span>
<span class="va">colData</span> <span class="op">&lt;-</span> <span class="va">colData</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://tidyr.tidyverse.org/reference/separate.html">separate</a></span><span class="op">(</span><span class="va">sampleID</span>,
        into <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"condition"</span>, <span class="st">"replicate"</span>, <span class="st">"viewpoint"</span><span class="op">)</span>,
        remove <span class="op">=</span> <span class="cn">FALSE</span>
    <span class="op">)</span>

<span class="co"># Load UMI-4C data and generate UMI4C object</span>
<span class="va">umi</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/UMI4C.html">makeUMI4C</a></span><span class="op">(</span>
    colData <span class="op">=</span> <span class="va">colData</span>,
    viewpoint_name <span class="op">=</span> <span class="st">"CIITA"</span>
<span class="op">)</span>

<span class="co">## 5) Perform differential test ---------------------------</span>
<span class="va">umi</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/fisherUMI4C.html">fisherUMI4C</a></span><span class="op">(</span><span class="va">umi</span>,
    filter_low <span class="op">=</span> <span class="fl">30</span>
<span class="op">)</span>

<span class="co">## 6)  Plot results ---------------------------------------</span>
<span class="fu"><a href="../reference/plotUMI4C.html">plotUMI4C</a></span><span class="op">(</span><span class="va">umi</span>,
    ylim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">10</span><span class="op">)</span> <span class="co"># ,</span>
    <span class="co"># xlim=c(11e6, 11.5e6)</span>
<span class="op">)</span></pre></div>
</div>
<div id="preparing-necessary-files" class="section level1">
<h1 class="hasAnchor">
<a href="#preparing-necessary-files" class="anchor"></a>Preparing necessary files</h1>
<div id="demultiplexing-fastq-files-containing-multiple-baits" class="section level2">
<h2 class="hasAnchor">
<a href="#demultiplexing-fastq-files-containing-multiple-baits" class="anchor"></a>Demultiplexing fastQ files containing multiple baits</h2>
<p>One of the many advantages of using the UMI-4C protocol is that it allows multiplexing of different baits using the same sample.</p>
<p>To facilitate the analysis, UMI4Cats provides a function for demultiplexing the paired-end FastQ files returned by the sequencing facility: <code>demultiplexFastq</code>.</p>
<p>This function requieres as input the name of the R1 file – it will automatically detect the R2 –, the barcode sequences and a path for the outputs. The barcode sequences and names to be used for each output sequences need to be provided as a <code>data.frame</code> with column names <code>sample</code> and <code>barcode</code>.</p>
<div class="sourceCode" id="cb3"><pre class="downlit">
<span class="va">path</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/downloadUMI4CexampleData.html">downloadUMI4CexampleData</a></span><span class="op">(</span>reduced<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span>

<span class="va">fastq</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span><span class="op">(</span><span class="va">path</span>, <span class="st">"CIITA"</span>, <span class="st">"fastq"</span>, <span class="st">"ctrl_hi24_CIITA_R1.fastq.gz"</span><span class="op">)</span>

<span class="va">barcodes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span>
    sample <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"CIITA"</span><span class="op">)</span>,
    barcode <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"GGACAAGCTCCCTGCAACTCA"</span><span class="op">)</span>
<span class="op">)</span>

<span class="fu"><a href="../reference/demultiplexFastq.html">demultiplexFastq</a></span><span class="op">(</span>
    barcodes <span class="op">=</span> <span class="va">barcodes</span>,
    fastq <span class="op">=</span> <span class="va">fastq</span>,
    out_path <span class="op">=</span> <span class="va">path</span>
<span class="op">)</span></pre></div>
</div>
<div id="reference-genome-digestion" class="section level2">
<h2 class="hasAnchor">
<a href="#reference-genome-digestion" class="anchor"></a>Reference genome digestion</h2>
<p>For the processing of the UMI-4C fastQ files it is necessary to construct a digested genome using the same restriction enzyme used in the UMI-4C experiments.</p>
<p>The <code>UMI4Cats</code> package includes the <code><a href="../reference/digestGenome.html">digestGenome()</a></code> function to make this process as easy as possible. The function uses a <code>BSgenome</code> object<a href="#fn1" class="footnote-ref" id="fnref1"><sup>1</sup></a> as reference genome which is digested by a given restriction enzyme cutting sequence (<code>res_enz</code>).</p>
<p>Besides the restriction sequence, it is also necessary to provide, as a zero-based numeric integer, the position at which the restriction enzyme cuts (<code>cut_pos</code>).</p>
<p>In the following table you can see three examples of the different cutting sequences for <em>DpnII</em>, <em>Csp6I</em> and <em>HindIII</em>.</p>
<table class="table">
<thead><tr class="header">
<th>Restriction enzyme</th>
<th>Restriction seq</th>
<th><code>res_enz</code></th>
<th><code>cut_pos</code></th>
</tr></thead>
<tbody>
<tr class="odd">
<td>DpnII</td>
<td>:<code>GATC</code>
</td>
<td>GATC</td>
<td>0</td>
</tr>
<tr class="even">
<td>Csp6I</td>
<td>
<code>G</code>:<code>TAC</code>
</td>
<td>GTAC</td>
<td>1</td>
</tr>
<tr class="odd">
<td>HindIII</td>
<td>
<code>A</code>:<code>AGCTT</code>
</td>
<td>AAGCTT</td>
<td>1</td>
</tr>
</tbody>
</table>
<p>For this example, we are using the hg19 <code>BSGenome</code> object and we are going to digest it using the <em>DpnII</em> enzyme.</p>
<div class="sourceCode" id="cb4"><pre class="downlit">
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">BSgenome.Hsapiens.UCSC.hg19</span><span class="op">)</span>
<span class="va">refgen</span> <span class="op">&lt;-</span> <span class="va">BSgenome.Hsapiens.UCSC.hg19</span>

<span class="va">hg19_dpnii</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/digestGenome.html">digestGenome</a></span><span class="op">(</span>
    res_enz <span class="op">=</span> <span class="st">"GATC"</span>,
    cut_pos <span class="op">=</span> <span class="fl">0</span>,
    name_RE <span class="op">=</span> <span class="st">"dpnII"</span>,
    ref_gen <span class="op">=</span> <span class="va">refgen</span>,
    sel_chr <span class="op">=</span> <span class="st">"chr16"</span>,
    out_path <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/tempfile.html">tempdir</a></span><span class="op">(</span><span class="op">)</span>, <span class="st">"digested_genome/"</span><span class="op">)</span>
<span class="op">)</span>

<span class="va">hg19_dpnii</span>
<span class="co">#&gt; [1] "/tmp/Rtmpuy4ZfT/digested_genome//BSgenome.Hsapiens.UCSC.hg19_dpnII"</span></pre></div>
<p>The digested genome will be saved in the folder provided in <code>out_path</code> as RData objects divided by chromosome. The path of the digested genome files is outputed by the function, so you can save it as a variable (in this case <code>hg19_dpnii</code>) and use it for downstream analyses.</p>
</div>
</div>
<div id="processing-umi-4c-fastq-files" class="section level1">
<h1 class="hasAnchor">
<a href="#processing-umi-4c-fastq-files" class="anchor"></a>Processing UMI-4C FASTQ files</h1>
<p>This step converts FastQ files of paired-end reads to UMI counts in the fragments resulted from the <em>in silico</em> digestion of the genome. The function for doing this analysis is <code><a href="../reference/contactsUMI4C.html">contactsUMI4C()</a></code> and should be ran in samples generated with the same viewpoint and restriction enzyme.</p>
<p>The function will consider all FastQ files in the same folder <code>fastq_dir</code> to be part of the same experiment (viewpoint + restriction enzyme). However, if you want to specify a subset of samples to perform the analysis you can do so by using the <code>file_pattern</code> argument. This way, only the files matching the specified pattern will be use as input for the analyses.</p>
<p>The different pairs for each sample should be named <code>_R1</code> or <code>_R2</code> and the function only accepts the following fastq suffixes: <code>.fastq</code>, <code>.fq</code>, <code>.fq.gz</code> or <code>.fastq.gz</code>.</p>
<p>For each analysis, the user needs to define 3 different sequences:</p>
<ul>
<li>
<strong>Bait/viewpoint sequence</strong> (<code>bait_seq</code>). This is the downstream primer sequence (DS primer) that matches the sequence of the queried bait.</li>
<li>
<strong>Padding sequence</strong> (<code>bait_pad</code>). The padding sequence corresponds to the nucleotides between the DS primer end and the restriction enzyme site.</li>
<li>
<strong>Restriction enzyme sequence</strong> (<code>res_enz</code>). This sequence is the restriction sequence recognized by the selected restriction enzyme.</li>
</ul>
<div class="figure" style="text-align: center">
<img src="figures/read_scheme.png" alt="Schematic of a UMI-4C read detailing the different elements that need to be used as input for processing the data." width="60%"><p class="caption">
Schematic of a UMI-4C read detailing the different elements that need to be used as input for processing the data.
</p>
</div>
<p>Additionally, it is necessary to define the restiction enzyme cutting postion (<code>cut_pos</code>) as previously did for generating the digested genome. The function will also need the path for the digested genome (<code>digested_genome</code>) generated with the function <code><a href="../reference/digestGenome.html">digestGenome()</a></code>.</p>
<p>As <code><a href="../reference/contactsUMI4C.html">contactsUMI4C()</a></code> performs the alignment using <a href="http://bowtie-bio.sourceforge.net/bowtie2/index.shtml">Bowtie2</a>, you will need to provide the reference genome indexes generated with bowtie2<a href="#fn2" class="footnote-ref" id="fnref2"><sup>2</sup></a>. Remember to make sure that your reference genome and digested genome correspond to the same build (in this example, hg19).</p>
<div class="sourceCode" id="cb5"><pre class="downlit">
<span class="co">## Use reduced example to make vignette faster</span>
<span class="co">## If you want to download the full dataset, set reduced = FALSE or remove</span>
<span class="co">## the reduce argument.</span>
<span class="va">path</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/downloadUMI4CexampleData.html">downloadUMI4CexampleData</a></span><span class="op">(</span>reduced<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span>
<span class="co">#&gt; untar: using cmd = '/usr/bin/tar -xf '/tmp/Rtmpuy4ZfT/BiocFileCache/273f5e3ab051_24309485' -C '/tmp/Rtmpuy4ZfT''</span>
<span class="va">raw_dir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span><span class="op">(</span><span class="va">path</span>, <span class="st">"CIITA"</span>, <span class="st">"fastq"</span><span class="op">)</span>
<span class="va">index_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span><span class="op">(</span><span class="va">path</span>, <span class="st">"ref_genome"</span>, <span class="st">"ucsc.hg19.chr16"</span><span class="op">)</span>

<span class="co">## Run main function to process UMI-4C contacts</span>
<span class="fu"><a href="../reference/contactsUMI4C.html">contactsUMI4C</a></span><span class="op">(</span>
    fastq_dir <span class="op">=</span> <span class="va">raw_dir</span>,
    wk_dir <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span><span class="op">(</span><span class="va">path</span>, <span class="st">"CIITA"</span><span class="op">)</span>,
    file_pattern <span class="op">=</span> <span class="st">"ctrl_hi24_CIITA"</span>, <span class="co"># Select only one sample to reduce running time</span>
    bait_seq <span class="op">=</span> <span class="st">"GGACAAGCTCCCTGCAACTCA"</span>,
    bait_pad <span class="op">=</span> <span class="st">"GGACTTGCA"</span>,
    res_enz <span class="op">=</span> <span class="st">"GATC"</span>,
    cut_pos <span class="op">=</span> <span class="fl">0</span>,
    digested_genome <span class="op">=</span> <span class="va">hg19_dpnii</span>,
    bowtie_index <span class="op">=</span> <span class="va">index_path</span>,
    ref_gen <span class="op">=</span> <span class="fu">BSgenome.Hsapiens.UCSC.hg19</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/BSgenome.Hsapiens.UCSC.hg19/man/package.html">BSgenome.Hsapiens.UCSC.hg19</a></span>,
    sel_seqname <span class="op">=</span> <span class="st">"chr16"</span>,
    threads <span class="op">=</span> <span class="fl">2</span>
<span class="op">)</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; [2020-10-09 19:26:49] Starting prepUMI4C using:</span>
<span class="co">#&gt;  &gt; Fastq directory:</span>
<span class="co">#&gt;  /tmp/Rtmpuy4ZfT/UMI4Cats_data_reduced/CIITA/fastq </span>
<span class="co">#&gt;  &gt; Work directory: /tmp/Rtmpuy4ZfT/UMI4Cats_data_reduced/CIITA </span>
<span class="co">#&gt;  &gt; Bait sequence: GGACAAGCTCCCTGCAACTCA </span>
<span class="co">#&gt;  &gt; Bait pad: GGACTTGCA </span>
<span class="co">#&gt;  &gt; Restriction enzyme: GATC </span>
<span class="co">#&gt;  &gt; Number of reads loaded into memory: 1e+11</span>
<span class="co">#&gt; [2020-10-09 19:26:50] Finished sample ctrl_hi24_CIITA</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; [2020-10-09 19:26:50] Starting splitUMI4C using:</span>
<span class="co">#&gt;  &gt; Work directory: /tmp/Rtmpuy4ZfT/UMI4Cats_data_reduced/CIITA </span>
<span class="co">#&gt;  &gt; Cut position: 0 </span>
<span class="co">#&gt;  &gt; Restriction enzyme: GATC </span>
<span class="co">#&gt;  &gt; Number of reads loaded into memory: 1e+11</span>
<span class="co">#&gt; [2020-10-09 19:26:50] Finished sample ctrl_hi24_CIITA_R1.fq.gz</span>
<span class="co">#&gt; [2020-10-09 19:26:51] Finished sample ctrl_hi32_CIITA_R1.fq.gz</span>
<span class="co">#&gt; [2020-10-09 19:26:51] Finished sample cyt_hi24_CIITA_R1.fq.gz</span>
<span class="co">#&gt; [2020-10-09 19:26:51] Finished sample cyt_hi32_CIITA_R1.fq.gz</span>
<span class="co">#&gt; [2020-10-09 19:26:51] Finished sample ctrl_hi24_CIITA_R2.fq.gz</span>
<span class="co">#&gt; [2020-10-09 19:26:51] Finished sample ctrl_hi32_CIITA_R2.fq.gz</span>
<span class="co">#&gt; [2020-10-09 19:26:52] Finished sample cyt_hi24_CIITA_R2.fq.gz</span>
<span class="co">#&gt; [2020-10-09 19:26:52] Finished sample cyt_hi32_CIITA_R2.fq.gz</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; [2020-10-09 19:26:52] Starting alignmentUMI4C using:</span>
<span class="co">#&gt;  &gt; Work directory: /tmp/Rtmpuy4ZfT/UMI4Cats_data_reduced/CIITA </span>
<span class="co">#&gt;  &gt; Viewpoint position: chr16:10972515-10972548 </span>
<span class="co">#&gt;  &gt; Reference genome: /tmp/Rtmpuy4ZfT/UMI4Cats_data_reduced/ref_genome/ucsc.hg19.chr16 </span>
<span class="co">#&gt;  &gt; Number of threads: 2</span>
<span class="co">#&gt; [2020-10-09 19:26:52] Finished sample ctrl_hi24_CIITA_R1.fq</span>
<span class="co">#&gt; [2020-10-09 19:26:53] Finished sample ctrl_hi24_CIITA_R2.fq</span>
<span class="co">#&gt; [2020-10-09 19:26:53] Finished sample ctrl_hi32_CIITA_R1.fq</span>
<span class="co">#&gt; [2020-10-09 19:26:53] Finished sample ctrl_hi32_CIITA_R2.fq</span>
<span class="co">#&gt; [2020-10-09 19:26:53] Finished sample cyt_hi24_CIITA_R1.fq</span>
<span class="co">#&gt; [2020-10-09 19:26:54] Finished sample cyt_hi24_CIITA_R2.fq</span>
<span class="co">#&gt; [2020-10-09 19:26:54] Finished sample cyt_hi32_CIITA_R1.fq</span>
<span class="co">#&gt; [2020-10-09 19:26:54] Finished sample cyt_hi32_CIITA_R2.fq</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; [2020-10-09 19:26:54] Starting counterUMI4C using:</span>
<span class="co">#&gt;  &gt; Work directory: /tmp/Rtmpuy4ZfT/UMI4Cats_data_reduced/CIITA </span>
<span class="co">#&gt;  &gt; Viewpoint position: chr16:10972515-10972548 </span>
<span class="co">#&gt;  &gt; Restriction enzyme: GATC </span>
<span class="co">#&gt;  &gt; Digested genome: /tmp/Rtmpuy4ZfT/digested_genome//BSgenome.Hsapiens.UCSC.hg19_dpnII</span>
<span class="co">#&gt; [2020-10-09 19:26:55] Finished sample ctrl_hi24_CIITA</span>
<span class="co">#&gt; [2020-10-09 19:26:56] Finished sample ctrl_hi32_CIITA</span>
<span class="co">#&gt; [2020-10-09 19:26:57] Finished sample cyt_hi24_CIITA</span>
<span class="co">#&gt; [2020-10-09 19:26:58] Finished sample cyt_hi32_CIITA</span></pre></div>
<p>Internally, <code><a href="../reference/contactsUMI4C.html">contactsUMI4C()</a></code> will then run sequentially the following processes:</p>
<ol style="list-style-type: decimal">
<li>FastQ files preparation (<code>prepUMI4C</code>). In this preprocessing step, only reads containing the <code>bait_seq</code> + <code>bait_pad</code> + <code>res_enz</code> will be selected. Also, reads with mean Phread quality scores &lt; 20 will be filtered out.</li>
<li>Split reads at restriction sites (<code>splitUMI4C</code>). Using the <code>res_enz</code> sequence, all reads from R1 and R2 will be split to mimic the fragments that were generated experimentally.</li>
<li>Align split reads to the reference genome (<code>alignmentUMI4C</code>).</li>
<li>Collapse reads using the UMI (<code>counterUMI4C</code>). This step is done to count molecular events thus reducing artifacts due to PCR duplicates. The function will return all fragment ends 10Mb around the viewpoint.</li>
</ol>
<p><strong>Note on memory usage</strong>: For the preparation and splitting, the FastQ file is loaded into memory. If you are having problems with the memory size available in your computer, you can change the number of lines to be loaded using the <code>numb_reads</code> parameter. See <code><a href="../reference/contactsUMI4C.html">?contactsUMI4C</a></code> for more information.</p>
<p>Finally, the output (UMI-4C counts) is a compressed tsv file per sample that will be saved in <code>wk_dir/count</code>. These files will be used as input for the analyses performed in the following section.</p>
<div id="quality-control-measures" class="section level2">
<h2 class="hasAnchor">
<a href="#quality-control-measures" class="anchor"></a>Quality control measures</h2>
<p>Once the processing step has been ran, the statistics of the UMI-4C filtering, alignment and final number of UMIs can be generated from the logs generated by the <code><a href="../reference/contactsUMI4C.html">contactsUMI4C()</a></code> function. By using these logs, the function <code><a href="../reference/statsUMI4C.html">statsUMI4C()</a></code> will produce a summary plot and a summary table with all statistics (in <code>wk_dir/logs/stats_summary.txt</code>).</p>
<p>For demonstration purposes, as we used a reduced version of the fastq files to reduce comptutation time, we are going to load the processed files included with the package.</p>
<div class="sourceCode" id="cb6"><pre class="downlit">
<span class="co"># Using the full dataset included in the package</span>
<span class="fu"><a href="../reference/statsUMI4C.html">statsUMI4C</a></span><span class="op">(</span>wk_dir <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"CIITA"</span>,
    package <span class="op">=</span> <span class="st">"UMI4Cats"</span>
<span class="op">)</span><span class="op">)</span></pre></div>
<p><img src="UMI4Cats_files/figure-html/stats-1.png" width="60%" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb7"><pre class="downlit">

<span class="va">stats</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.delim</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"CIITA"</span>, <span class="st">"logs"</span>, <span class="st">"stats_summary.txt"</span>,
    package <span class="op">=</span> <span class="st">"UMI4Cats"</span>
<span class="op">)</span><span class="op">)</span>

<span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html">kable</a></span><span class="op">(</span><span class="va">stats</span><span class="op">)</span></pre></div>
<table class="table">
<colgroup>
<col width="15%">
<col width="14%">
<col width="17%">
<col width="14%">
<col width="13%">
<col width="9%">
<col width="11%">
<col width="4%">
</colgroup>
<thead><tr class="header">
<th align="left">sample_id</th>
<th align="right">specific_reads</th>
<th align="right">nonspecific_reads</th>
<th align="right">filtered_reads</th>
<th align="right">filtout_reads</th>
<th align="right">al_mapped</th>
<th align="right">al_unmapped</th>
<th align="right">umi</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">ctrl_hi24_CIITA</td>
<td align="right">183831</td>
<td align="right">16169</td>
<td align="right">173342</td>
<td align="right">10489</td>
<td align="right">525291</td>
<td align="right">24248</td>
<td align="right">1429</td>
</tr>
<tr class="even">
<td align="left">ctrl_hi32_CIITA</td>
<td align="right">183759</td>
<td align="right">16241</td>
<td align="right">181789</td>
<td align="right">1970</td>
<td align="right">554708</td>
<td align="right">22643</td>
<td align="right">2881</td>
</tr>
<tr class="odd">
<td align="left">cyt_hi24_CIITA</td>
<td align="right">181278</td>
<td align="right">18722</td>
<td align="right">178473</td>
<td align="right">2805</td>
<td align="right">541994</td>
<td align="right">27625</td>
<td align="right">1743</td>
</tr>
<tr class="even">
<td align="left">cyt_hi32_CIITA</td>
<td align="right">183770</td>
<td align="right">16230</td>
<td align="right">182106</td>
<td align="right">1664</td>
<td align="right">555015</td>
<td align="right">23484</td>
<td align="right">2589</td>
</tr>
</tbody>
</table>
<p>The quality control measures summarized both in the plot and the table are:</p>
<ul>
<li>
<strong>Specific reads</strong>. Corresponds to the number of reads that contained the full viewpoint sequence (<code>bait_seq</code> + <code>bait_pad</code> + <code>res_enz</code>).</li>
<li>
<strong>Filtered reads</strong>. A quality filtering to ensure that filtered (or kept) reads have a mean Phred quality score <code>&gt;= 20</code>.</li>
<li>
<strong>Mapping stats</strong>. Indicates how many split reads are mapped or unmapped to the reference genome.</li>
<li>
<strong>UMIs</strong>. Shows the final number of molecular contacts detected.</li>
</ul>
</div>
</div>
<div id="analyzing-umi-4c-data" class="section level1">
<h1 class="hasAnchor">
<a href="#analyzing-umi-4c-data" class="anchor"></a>Analyzing UMI-4C data</h1>
<p>After processing the fastQ reads and obtaining tables summarizing the contacts and the number of UMIs supporting each interaction with the viewpoint, the next step is that of analyzing these data, detect differential contacts and visualize the genomic interactions.</p>
<div id="build-the-umi4c-object" class="section level2">
<h2 class="hasAnchor">
<a href="#build-the-umi4c-object" class="anchor"></a>Build the <code>UMI4C</code> object</h2>
<p>The first step of the UMI-4C data analysis consists in loading the tables generated by the function <code><a href="../reference/contactsUMI4C.html">contactsUMI4C()</a></code> and use them to construct a <code>UMI4C</code> object. All these steps are performed automatically by the <code><a href="../reference/UMI4C.html">makeUMI4C()</a></code> function.</p>
<p>The <code>makeUMI4C</code> will need as input a data frame (<code>colData</code>) containing all relevant experiment information that will be needed for analyzing the data later on. The mandatory columns that this data frame should have are:</p>
<ol style="list-style-type: decimal">
<li>
<code>sampleID</code>: Unique identifier for the sample.</li>
<li>
<code>replicate</code>: Replicate identifier or number.</li>
<li>
<code>condition</code>: Condition for performing the differential analysis. For example: “control” and “treatment”, two different cell types, etc. The condition column should only have <strong>two</strong> different values. If more condition variables are provided, the differential analysis will fail.</li>
<li>
<code>file</code>: Complete path and filename where the tsv files generated by <code>contactsUMI4C</code> are saved.</li>
</ol>
<p>You can also include other additional columns to <code>colData</code>.</p>
<p>By default, the function will combine the samples using the <strong>condition</strong> column from <code>colData</code>, but you can change this behavior using the <code>grouping</code> argument in <code><a href="../reference/UMI4C.html">makeUMI4C()</a></code>. For example, if you set <code>grouping="sampleID"</code> it will plot the individual profiles or if you set <code>grouping="replicate"</code> your samples will be summarized according to the replicates you provided in <code>colData</code>. By default, the reference group will be the one with less number of UMIs. You can override this behaviour by setting the <code>ref_umi4c</code> argument, which should be one of the groups present in the column used as <code>grouping</code>.</p>
<p>Additionally, the <code>makeUMI4C</code> function also contains other arguments that can be used if you want to tweak the default parameters of the analysis. See <code><a href="../reference/UMI4C.html">?makeUMI4C</a></code> to have a complete list and description of all the arguments.</p>
<div class="sourceCode" id="cb8"><pre class="downlit">
<span class="co"># Load sample processed file paths</span>
<span class="va">files</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html">list.files</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"CIITA"</span>, <span class="st">"count"</span>, package<span class="op">=</span><span class="st">"UMI4Cats"</span><span class="op">)</span>,
                    pattern <span class="op">=</span> <span class="st">"*_counts.tsv.gz"</span>,
                    full.names <span class="op">=</span> <span class="cn">TRUE</span>
<span class="op">)</span>

<span class="co"># Create colData including all relevant information</span>
<span class="va">colData</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span>
    sampleID <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html">gsub</a></span><span class="op">(</span><span class="st">"_counts.tsv.gz"</span>, <span class="st">""</span>, <span class="fu"><a href="https://rdrr.io/r/base/basename.html">basename</a></span><span class="op">(</span><span class="va">files</span><span class="op">)</span><span class="op">)</span>,
    file <span class="op">=</span> <span class="va">files</span>,
    stringsAsFactors <span class="op">=</span> <span class="cn">FALSE</span>
<span class="op">)</span>

<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyr.tidyverse.org">tidyr</a></span><span class="op">)</span>
<span class="va">colData</span> <span class="op">&lt;-</span> <span class="va">colData</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://tidyr.tidyverse.org/reference/separate.html">separate</a></span><span class="op">(</span><span class="va">sampleID</span>,
        into <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"condition"</span>, <span class="st">"replicate"</span>, <span class="st">"viewpoint"</span><span class="op">)</span>,
        remove <span class="op">=</span> <span class="cn">FALSE</span>
    <span class="op">)</span>

<span class="co"># Load UMI-4C data and generate UMI4C object</span>
<span class="va">umi</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/UMI4C.html">makeUMI4C</a></span><span class="op">(</span>
    colData <span class="op">=</span> <span class="va">colData</span>,
    viewpoint_name <span class="op">=</span> <span class="st">"CIITA"</span>,
    grouping <span class="op">=</span> <span class="st">"condition"</span>,
    ref_umi4c <span class="op">=</span> <span class="st">"ctrl"</span>
<span class="op">)</span>

<span class="va">umi</span>
<span class="co">#&gt; class: UMI4C </span>
<span class="co">#&gt; dim: 2838 2 </span>
<span class="co">#&gt; metadata(7): bait scales ... region ref_umi4c</span>
<span class="co">#&gt; assays(6): umi norm_mat ... scale sd</span>
<span class="co">#&gt; rownames(2838): frag_26081 frag_26082 ... frag_28925 frag_28926</span>
<span class="co">#&gt; rowData names(2): id_contact position</span>
<span class="co">#&gt; colnames(2): ctrl cyt</span>
<span class="co">#&gt; colData names(5): condition sampleID replicate viewpoint file</span></pre></div>
<p>The <code>makeUMI4C</code> function will perform the following steps to generate the <code>UMI4C</code> object:</p>
<ol style="list-style-type: decimal">
<li>
<strong>Remove fragment ends around the bait</strong>, as they are generally biased because of their proximity to the viewpoint. The value of the region that will be excluded from the analysis can be specified using the <code>bait_exclusion</code> argument. The default is 3kb.</li>
<li>
<strong>Focus the scope</strong> of the analysis in a specific genomic region around the bait, by default this is a 2Mb window centerd on the viewpoint. The default value can be changed using the <code>bait_expansion</code> argument.</li>
<li>Sum the UMIs of the different samples belonging to the same group (defined by the <code>grouping</code> variable).</li>
<li>
<strong>Obtain the normalization matrices</strong> that will be used to scale the groups to the reference, by default the group with less UMIs. If you want to avoid this normalization step, you can set <code>normalized</code> to <code>FALSE</code>.</li>
<li>Calculate the <strong>domainograms</strong> for each group.</li>
<li>Calculate the <strong>adaptive trend</strong> for each group.</li>
</ol>
<div id="accessing-information-from-the-umi4c-object" class="section level3">
<h3 class="hasAnchor">
<a href="#accessing-information-from-the-umi4c-object" class="anchor"></a>Accessing information from the <code>UMI4C</code> object</h3>
<p>The usual accessor functions from the <code>SummarizedExperiment-class</code><a href="#fn3" class="footnote-ref" id="fnref3"><sup>3</sup></a> also work with the UMI-4C class (for example: <code>assay</code>, <code>colData</code>, etc.). Other additional accessors have been created to retrieve different information:</p>
<ul>
<li>
<code><a href="../reference/dgram.html">dgram()</a></code>. Get a list of the domaingorams for each group.</li>
<li>
<code><a href="../reference/dgram.html">bait()</a></code>. Retrieve a GRanges object with the bait position.</li>
<li>
<code><a href="../reference/dgram.html">trend()</a></code>. Obtain a data.frame in long format with the adaptive smoothing trend.</li>
<li>
<code><a href="../reference/dgram.html">results()</a></code>. Retrieve results from the differential analysis. This only works if a differential analysis has been performed on the UMI4C object.</li>
</ul>
<div class="sourceCode" id="cb9"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="va">umi</span><span class="op">)</span><span class="op">)</span> <span class="co"># Retrieve raw UMIs</span>
<span class="co">#&gt;            ctrl cyt</span>
<span class="co">#&gt; frag_26081    0   0</span>
<span class="co">#&gt; frag_26082    0   0</span>
<span class="co">#&gt; frag_26083    0   0</span>
<span class="co">#&gt; frag_26084    0   3</span>
<span class="co">#&gt; frag_26085    0   0</span>
<span class="co">#&gt; frag_26086    0   0</span>

<span class="fu">colData</span><span class="op">(</span><span class="va">umi</span><span class="op">)</span> <span class="co"># Retrieve column information</span>
<span class="co">#&gt; DataFrame with 2 rows and 5 columns</span>
<span class="co">#&gt;        condition               sampleID   replicate    viewpoint</span>
<span class="co">#&gt;      &lt;character&gt;            &lt;character&gt; &lt;character&gt;  &lt;character&gt;</span>
<span class="co">#&gt; ctrl        ctrl ctrl_hi24_CIITA, ctr..  hi24, hi32 CIITA, CIITA</span>
<span class="co">#&gt; cyt          cyt cyt_hi24_CIITA, cyt_..  hi24, hi32 CIITA, CIITA</span>
<span class="co">#&gt;                        file</span>
<span class="co">#&gt;                 &lt;character&gt;</span>
<span class="co">#&gt; ctrl /usr/local/lib/R/hos..</span>
<span class="co">#&gt; cyt  /usr/local/lib/R/hos..</span>

<span class="fu">rowRanges</span><span class="op">(</span><span class="va">umi</span><span class="op">)</span> <span class="co"># Retrieve fragment coordinates</span>
<span class="co">#&gt; GRanges object with 2838 ranges and 2 metadata columns:</span>
<span class="co">#&gt;              seqnames    ranges strand |  id_contact    position</span>
<span class="co">#&gt;                 &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt; &lt;character&gt;</span>
<span class="co">#&gt;   frag_26081    chr16  10470999      * |  frag_26081    upstream</span>
<span class="co">#&gt;   frag_26082    chr16  10471013      * |  frag_26082    upstream</span>
<span class="co">#&gt;   frag_26083    chr16  10471023      * |  frag_26083    upstream</span>
<span class="co">#&gt;   frag_26084    chr16  10471233      * |  frag_26084    upstream</span>
<span class="co">#&gt;   frag_26085    chr16  10472222      * |  frag_26085    upstream</span>
<span class="co">#&gt;          ...      ...       ...    ... .         ...         ...</span>
<span class="co">#&gt;   frag_28922    chr16  11469408      * |  frag_28922  downstream</span>
<span class="co">#&gt;   frag_28923    chr16  11469834      * |  frag_28923  downstream</span>
<span class="co">#&gt;   frag_28924    chr16  11470052      * |  frag_28924  downstream</span>
<span class="co">#&gt;   frag_28925    chr16  11470627      * |  frag_28925  downstream</span>
<span class="co">#&gt;   frag_28926    chr16  11470914      * |  frag_28926  downstream</span>
<span class="co">#&gt;   -------</span>
<span class="co">#&gt;   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span>

<span class="fu"><a href="../reference/dgram.html">dgram</a></span><span class="op">(</span><span class="va">umi</span><span class="op">)</span> <span class="co"># Retrieve domainograms</span>
<span class="co">#&gt; List of length 2</span>
<span class="co">#&gt; names(2): ctrl cyt</span>

<span class="fu"><a href="../reference/dgram.html">bait</a></span><span class="op">(</span><span class="va">umi</span><span class="op">)</span> <span class="co"># Retrieve bait coordinates</span>
<span class="co">#&gt; GRanges object with 1 range and 1 metadata column:</span>
<span class="co">#&gt;                   seqnames    ranges strand |        name</span>
<span class="co">#&gt;                      &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;</span>
<span class="co">#&gt;   ctrl_hi24_CIITA    chr16  10970996      * |       CIITA</span>
<span class="co">#&gt;   -------</span>
<span class="co">#&gt;   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span>

<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="fu"><a href="../reference/dgram.html">trend</a></span><span class="op">(</span><span class="va">umi</span><span class="op">)</span><span class="op">)</span> <span class="co"># Retrieve adaptive smoothing trend</span>
<span class="co">#&gt;   geo_coord     trend         sd scale id_contact sample</span>
<span class="co">#&gt; 1  10514782 0.3897638 0.07834547   127 frag_26208   ctrl</span>
<span class="co">#&gt; 2  10515114 0.3928571 0.07896726   126 frag_26209   ctrl</span>
<span class="co">#&gt; 3  10515510 0.3897638 0.07834547   127 frag_26210   ctrl</span>
<span class="co">#&gt; 4  10515907 0.3867188 0.07773339   128 frag_26211   ctrl</span>
<span class="co">#&gt; 5  10516302 0.3837209 0.07713081   129 frag_26212   ctrl</span>
<span class="co">#&gt; 6  10516697 0.3884615 0.07730674   130 frag_26213   ctrl</span>
<span class="co">#&gt;                           sampleID  replicate    viewpoint</span>
<span class="co">#&gt; 1 ctrl_hi24_CIITA, ctrl_hi32_CIITA hi24, hi32 CIITA, CIITA</span>
<span class="co">#&gt; 2 ctrl_hi24_CIITA, ctrl_hi32_CIITA hi24, hi32 CIITA, CIITA</span>
<span class="co">#&gt; 3 ctrl_hi24_CIITA, ctrl_hi32_CIITA hi24, hi32 CIITA, CIITA</span>
<span class="co">#&gt; 4 ctrl_hi24_CIITA, ctrl_hi32_CIITA hi24, hi32 CIITA, CIITA</span>
<span class="co">#&gt; 5 ctrl_hi24_CIITA, ctrl_hi32_CIITA hi24, hi32 CIITA, CIITA</span>
<span class="co">#&gt; 6 ctrl_hi24_CIITA, ctrl_hi32_CIITA hi24, hi32 CIITA, CIITA</span>
<span class="co">#&gt;                                                                                                                                                                                           file</span>
<span class="co">#&gt; 1 /usr/local/lib/R/host-site-library/UMI4Cats/extdata/CIITA/count/ctrl_hi24_CIITA_counts.tsv.gz, /usr/local/lib/R/host-site-library/UMI4Cats/extdata/CIITA/count/ctrl_hi32_CIITA_counts.tsv.gz</span>
<span class="co">#&gt; 2 /usr/local/lib/R/host-site-library/UMI4Cats/extdata/CIITA/count/ctrl_hi24_CIITA_counts.tsv.gz, /usr/local/lib/R/host-site-library/UMI4Cats/extdata/CIITA/count/ctrl_hi32_CIITA_counts.tsv.gz</span>
<span class="co">#&gt; 3 /usr/local/lib/R/host-site-library/UMI4Cats/extdata/CIITA/count/ctrl_hi24_CIITA_counts.tsv.gz, /usr/local/lib/R/host-site-library/UMI4Cats/extdata/CIITA/count/ctrl_hi32_CIITA_counts.tsv.gz</span>
<span class="co">#&gt; 4 /usr/local/lib/R/host-site-library/UMI4Cats/extdata/CIITA/count/ctrl_hi24_CIITA_counts.tsv.gz, /usr/local/lib/R/host-site-library/UMI4Cats/extdata/CIITA/count/ctrl_hi32_CIITA_counts.tsv.gz</span>
<span class="co">#&gt; 5 /usr/local/lib/R/host-site-library/UMI4Cats/extdata/CIITA/count/ctrl_hi24_CIITA_counts.tsv.gz, /usr/local/lib/R/host-site-library/UMI4Cats/extdata/CIITA/count/ctrl_hi32_CIITA_counts.tsv.gz</span>
<span class="co">#&gt; 6 /usr/local/lib/R/host-site-library/UMI4Cats/extdata/CIITA/count/ctrl_hi24_CIITA_counts.tsv.gz, /usr/local/lib/R/host-site-library/UMI4Cats/extdata/CIITA/count/ctrl_hi32_CIITA_counts.tsv.gz</span></pre></div>
</div>
</div>
<div id="performing-differential-analysis" class="section level2">
<h2 class="hasAnchor">
<a href="#performing-differential-analysis" class="anchor"></a>Performing differential analysis</h2>
<p>Once the <code>UMI4C</code> object is generated, you can perform a differential analysis using the <code><a href="../reference/fisherUMI4C.html">fisherUMI4C()</a></code> function.</p>
<p>First of all, queried regions will be filtered according to the median number of UMIs present in the <code>filter_low</code> parameter. You can reduce this number or disable filtering using <code>filter_low = FALSE</code>.</p>
<p>Then, a contingency table for each region where the differential test should be performed will be created, where the group stored in <code>metadata(umi)$ref_umi4c</code> will be used as references. The values on the contingency table correspond to the following:</p>
<table class="table">
<thead><tr class="header">
<th>Group</th>
<th>Query region</th>
<th>Region</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>Reference</td>
<td><span class="math inline">\(n1\)</span></td>
<td><span class="math inline">\(N1 - n1\)</span></td>
</tr>
<tr class="even">
<td>Condition</td>
<td><span class="math inline">\(n2\)</span></td>
<td><span class="math inline">\(N2 - n2\)</span></td>
</tr>
</tbody>
</table>
<p>Where <span class="math inline">\(N1\)</span> and <span class="math inline">\(N2\)</span> correspond to the total number of UMIs in the whole analyzed region (<code>metadata(umi)$region</code>) and <span class="math inline">\(n1\)</span> and <span class="math inline">\(n2\)</span> correspond to the total number of UMIs in the query region that is to be tested.</p>
<p>After all the Fisher’s Exact Tests are performed, p-values are adjusted using the FDR method. Query regions with adjusted p-values &gt; 0.05 will be considered significantly different. Check <code>?fisherUMI4()</code> for more information and additional arguments you can provide to the function.</p>
<div id="differential-analysis-in-a-specific-set-of-regions" class="section level3">
<h3 class="hasAnchor">
<a href="#differential-analysis-in-a-specific-set-of-regions" class="anchor"></a>Differential analysis in a specific set of regions</h3>
<p>Many times, UMI-4C is used to test whether a certain set of promoter-enhancer contacts differ between conditions or cell types. For this cases, the <code>fisherUMI4c()</code> with the argument <code>query_regions</code> set to those regions of interest can be used. Your query regions should be provided as a <code>GRanges</code> object. If your set of query regions is narrow (~1kb), you can resize them to the desired width using the <code>resize</code> argument.</p>
<div class="sourceCode" id="cb10"><pre class="downlit">
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">GenomicRanges</span><span class="op">)</span>

<span class="co"># Provide your own query regions as GRanges objects</span>
<span class="va">enhancers</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html">GRanges</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span>
    <span class="st">"chr16:10925006-10928900"</span>,
    <span class="st">"chr16:11102721-11103700"</span>
<span class="op">)</span><span class="op">)</span>

<span class="co"># Perform differential test</span>
<span class="va">umi_dif1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/fisherUMI4C.html">fisherUMI4C</a></span><span class="op">(</span><span class="va">umi</span>,
    query_regions <span class="op">=</span> <span class="va">enhancers</span>,
    filter_low <span class="op">=</span> <span class="fl">20</span>,
    resize <span class="op">=</span> <span class="fl">5e3</span>
<span class="op">)</span></pre></div>
</div>
<div id="differential-analysis-in-the-binned-region" class="section level3">
<h3 class="hasAnchor">
<a href="#differential-analysis-in-the-binned-region" class="anchor"></a>Differential analysis in the binned region</h3>
<p>In case you do not have a candidate set of enhancer regions, the <code><a href="../reference/fisherUMI4C.html">fisherUMI4C()</a></code> function will bin the experiment region in bin of <code>window_size</code> size (default is 5kb). This will then be the query regions where the differential test will be performed, after filtering for number of UMIs with the <code>filter_low</code> argument.</p>
<div class="sourceCode" id="cb11"><pre class="downlit">
<span class="co"># Perform differential test</span>
<span class="va">umi_dif2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/fisherUMI4C.html">fisherUMI4C</a></span><span class="op">(</span><span class="va">umi</span>,
    filter_low <span class="op">=</span> <span class="fl">20</span>,
    window_size <span class="op">=</span> <span class="fl">5e3</span>
<span class="op">)</span></pre></div>
</div>
<div id="retrieve-differential-analysis-results" class="section level3">
<h3 class="hasAnchor">
<a href="#retrieve-differential-analysis-results" class="anchor"></a>Retrieve differential analysis results</h3>
<p>Results from this test can be retrieved using the <code><a href="../reference/dgram.html">results()</a></code> on the <code>UMI4C</code> object returned by the test function.</p>
<div class="sourceCode" id="cb12"><pre class="downlit">
<span class="fu"><a href="../reference/dgram.html">results</a></span><span class="op">(</span><span class="va">umi_dif1</span>, ordered <span class="op">=</span> <span class="cn">TRUE</span>, counts <span class="op">=</span> <span class="cn">TRUE</span>, format <span class="op">=</span> <span class="st">"data.frame"</span><span class="op">)</span>
<span class="co">#&gt;   seqnames    start      end       id umis_ref umis_cond       pvalue</span>
<span class="co">#&gt; 1    chr16 10924453 10929452 region_1       20        54 9.243518e-05</span>
<span class="co">#&gt; 2    chr16 11100711 11105710 region_2        6         9           NA</span>
<span class="co">#&gt;   odds_ratio log2_odds_ratio         padj sign</span>
<span class="co">#&gt; 1   2.673203        1.418569 9.243518e-05 TRUE</span>
<span class="co">#&gt; 2         NA              NA           NA   NA</span>

<span class="fu"><a href="../reference/dgram.html">results</a></span><span class="op">(</span><span class="va">umi_dif2</span>, ordered <span class="op">=</span> <span class="cn">TRUE</span>, counts <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="co">#&gt; GRanges object with 198 ranges and 8 metadata columns:</span>
<span class="co">#&gt;         seqnames            ranges strand |          id  umis_ref umis_cond</span>
<span class="co">#&gt;            &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt; &lt;integer&gt; &lt;integer&gt;</span>
<span class="co">#&gt;     [1]    chr16 10925996-10930995      * |   region_92        31        83</span>
<span class="co">#&gt;     [2]    chr16 10920996-10925995      * |   region_91        21        42</span>
<span class="co">#&gt;     [3]    chr16 10930996-10935995      * |   region_93        20        38</span>
<span class="co">#&gt;     [4]    chr16 10940996-10945995      * |   region_95        37        20</span>
<span class="co">#&gt;     [5]    chr16 10900996-10905995      * |   region_87        16        30</span>
<span class="co">#&gt;     ...      ...               ...    ... .         ...       ...       ...</span>
<span class="co">#&gt;   [194]    chr16 11445996-11450995      * |  region_194         0         2</span>
<span class="co">#&gt;   [195]    chr16 11450996-11455995      * |  region_195         7         6</span>
<span class="co">#&gt;   [196]    chr16 11455996-11460995      * |  region_196         3         5</span>
<span class="co">#&gt;   [197]    chr16 11460996-11465995      * |  region_197         7         0</span>
<span class="co">#&gt;   [198]    chr16 11465996-11470995      * |  region_198         4         3</span>
<span class="co">#&gt;              pvalue odds_ratio log2_odds_ratio        padj      sign</span>
<span class="co">#&gt;           &lt;numeric&gt;  &lt;numeric&gt;       &lt;numeric&gt;   &lt;numeric&gt; &lt;logical&gt;</span>
<span class="co">#&gt;     [1] 1.48733e-06   2.670392        1.417052 3.71833e-05      TRUE</span>
<span class="co">#&gt;     [2] 1.09773e-02   1.969685        0.977965 1.37217e-01     FALSE</span>
<span class="co">#&gt;     [3] 2.44768e-02   1.868954        0.902231 1.52980e-01     FALSE</span>
<span class="co">#&gt;     [4] 2.28298e-02   0.524297       -0.931545 1.52980e-01     FALSE</span>
<span class="co">#&gt;     [5] 5.37197e-02   1.841436        0.880831 2.68598e-01     FALSE</span>
<span class="co">#&gt;     ...         ...        ...             ...         ...       ...</span>
<span class="co">#&gt;   [194]          NA         NA              NA          NA      &lt;NA&gt;</span>
<span class="co">#&gt;   [195]          NA         NA              NA          NA      &lt;NA&gt;</span>
<span class="co">#&gt;   [196]          NA         NA              NA          NA      &lt;NA&gt;</span>
<span class="co">#&gt;   [197]          NA         NA              NA          NA      &lt;NA&gt;</span>
<span class="co">#&gt;   [198]          NA         NA              NA          NA      &lt;NA&gt;</span>
<span class="co">#&gt;   -------</span>
<span class="co">#&gt;   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span></pre></div>
<p>The parameter <code>counts</code> indicates wheter raw counts used for the test should be outputed. <code>umis_ref</code> corresponds to the number of raw UMIs from the sample/group used as reference (accessible through <code>metadata(umi_dif)$ref_umi4c</code>).</p>
</div>
</div>
</div>
<div id="visualizing-umi-4c-contact-data" class="section level1">
<h1 class="hasAnchor">
<a href="#visualizing-umi-4c-contact-data" class="anchor"></a>Visualizing UMI-4C contact data</h1>
<p>Once the <code>UMI4C</code> object is created, you can plot it using the <code>plotUMI4C</code> function.</p>
<p>The gene annotations will be extracted from the <code>TxDb.Hsapiens.UCSC.hg19.knownGene</code> package by default. Make sure that the annotations coincide with your reference genome. You can check the package <a href="https://bioconductor.org/packages/release/bioc/html/GenomicFeatures.html"><code>GenomicFeatures</code></a> for more information on available <code>TxDb</code> objects.</p>
<div class="sourceCode" id="cb13"><pre class="downlit">
<span class="fu"><a href="../reference/plotUMI4C.html">plotUMI4C</a></span><span class="op">(</span><span class="va">umi</span>,
    TxDb <span class="op">=</span> <span class="fu">TxDb.Hsapiens.UCSC.hg19.knownGene</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/TxDb.Hsapiens.UCSC.hg19.knownGene/man/package.html">TxDb.Hsapiens.UCSC.hg19.knownGene</a></span>
<span class="op">)</span></pre></div>
<p><img src="UMI4Cats_files/figure-html/plot-umi4c-1.png" width="60%" style="display: block; margin: auto;"></p>
<p>The domainogram plotting is controlled by the <code>dgram_plot</code> argument. If you set it to <code>FALSE</code>, the domainogram will not be plotted.</p>
<div class="sourceCode" id="cb14"><pre class="downlit">
<span class="fu"><a href="../reference/plotUMI4C.html">plotUMI4C</a></span><span class="op">(</span><span class="va">umi</span>, dgram_plot <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></pre></div>
<p><img src="UMI4Cats_files/figure-html/plot-nodgram-1.png" width="60%" style="display: block; margin: auto;"></p>
<p>In case you are interested in plotting the profiles of the different samples contained in your experiment, you will need to create another <code>UMI4C</code> object setting <code>grouping=sampleID</code> and then you will be able to plot this information:</p>
<div class="sourceCode" id="cb15"><pre class="downlit">
<span class="va">umi_reps</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/UMI4C.html">makeUMI4C</a></span><span class="op">(</span>
    colData <span class="op">=</span> <span class="va">colData</span>,
    viewpoint_name <span class="op">=</span> <span class="st">"CIITA"</span>,
    grouping <span class="op">=</span> <span class="st">"sampleID"</span>
<span class="op">)</span>

<span class="fu"><a href="../reference/plotUMI4C.html">plotUMI4C</a></span><span class="op">(</span><span class="va">umi_reps</span><span class="op">)</span></pre></div>
<p><img src="UMI4Cats_files/figure-html/plot-reps-1.png" width="60%" style="display: block; margin: auto;"></p>
<p>If the <code>UMI4C</code> object contains information on the differential contacts, this data will be shown in the plot as well.</p>
<div class="sourceCode" id="cb16"><pre class="downlit">
<span class="fu"><a href="../reference/plotUMI4C.html">plotUMI4C</a></span><span class="op">(</span><span class="va">umi_dif1</span><span class="op">)</span></pre></div>
<p><img src="UMI4Cats_files/figure-html/plot-dif-1.png" width="60%" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb17"><pre class="downlit">
<span class="fu"><a href="../reference/plotUMI4C.html">plotUMI4C</a></span><span class="op">(</span><span class="va">umi_dif2</span><span class="op">)</span></pre></div>
<p><img src="UMI4Cats_files/figure-html/plot-dif-2.png" width="60%" style="display: block; margin: auto;"></p>
<p>There are several different arguments you can provide to <code>plotUMI4C</code> to modify the output plot. You can check them at <code><a href="../reference/plotUMI4C.html">?plotUMI4C</a></code>.</p>
<p>The <code>plotUMI4C</code> function is a wrapper for separate functions that plot the different elements included in the figure. You can use each of the functions separately if you are interesting in combining them differently or with other ggplot2 objects. Check each function documentation at <code><a href="../reference/plotTrend.html">?plotTrend</a></code>, <code><a href="../reference/plotGenes.html">?plotGenes</a></code>, <code><a href="../reference/plotDomainogram.html">?plotDomainogram</a></code> and <code><a href="../reference/plotDifferential.html">?plotDifferential</a></code>.</p>
</div>
<div id="references" class="section level1 unnumbered">
<h1 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h1>
<div id="refs" class="references">
<div id="ref-Ramos-Rodriguez2019">
<p>Ramos-Rodríguez, Mireia, Helena Raurell-Vila, Maikel L. Colli, Maria Inês Alvelos, Marc Subirana-Granés, Jonàs Juan-Mateu, Richard Norris, et al. 2019. “The impact of proinflammatory cytokines on the <span class="math inline">\(\beta\)</span>-cell regulatory landscape provides insights into the genetics of type 1 diabetes.” <em>Nature Genetics</em> 51 (11): 1588–95. <a href="https://doi.org/10.1038/s41588-019-0524-6">https://doi.org/10.1038/s41588-019-0524-6</a>.</p>
</div>
<div id="ref-Schwartzman2016">
<p>Schwartzman, Omer, Zohar Mukamel, Noa Oded-Elkayam, Pedro Olivares-Chauvet, Yaniv Lubling, Gilad Landan, Shai Izraeli, and Amos Tanay. 2016. “UMI-4C for quantitative and targeted chromosomal contact profiling.” <em>Nature Methods</em> 13 (8): 685–91. <a href="https://doi.org/10.1038/nmeth.3922">https://doi.org/10.1038/nmeth.3922</a>.</p>
</div>
</div>
</div>
<div class="footnotes">
<hr>
<ol>
<li id="fn1"><p>More information on <code>BSgenome</code> package and objects can be found <a href="https://bioconductor.org/packages/release/bioc/html/BSgenome.html">here</a><a href="#fnref1" class="footnote-back">↩︎</a></p></li>
<li id="fn2"><p>See the <a href="http://bowtie-bio.sourceforge.net/bowtie2/manual.shtml#getting-started-with-bowtie-2-lambda-phage-example">getting started section</a> on the Bowtie2 page for more information on how to generate the index for the reference genome.<a href="#fnref2" class="footnote-back">↩︎</a></p></li>
<li id="fn3"><p>See more about the SummarizedExperiment class <a href="https://bioconductor.org/packages/release/bioc/html/SummarizedExperiment.html">here</a><a href="#fnref3" class="footnote-back">↩︎</a></p></li>
</ol>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Mireia Ramos-Rodriguez, Marc Subirana-Granes.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.9000.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
