<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Analyzing UMI-4C data with UMI4Cats • UMI4Cats</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Analyzing UMI-4C data with UMI4Cats">
<meta property="og:description" content="UMI4Cats">
<meta property="og:image" content="/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">UMI4Cats</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.1.4</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/UMI4Cats.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/Pasquali-lab/UMI4Cats/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="UMI4Cats_files/header-attrs-2.6/header-attrs.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Analyzing UMI-4C data with UMI4Cats</h1>
                        <h4 class="author">Mireia Ramos-Rodríguez and Marc Subirana-Granés</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/Pasquali-lab/UMI4Cats/blob/master/vignettes/UMI4Cats.Rmd"><code>vignettes/UMI4Cats.Rmd</code></a></small>
      <div class="hidden name"><code>UMI4Cats.Rmd</code></div>

    </div>

    
    
<p><img src="../man/figures/UMI4Cats.png" width="10%" style="display: block; margin: auto;"></p>
<div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h1>
<p>Hello stranger! If you are here, that means you’ve successfully completed the UMI-4C protocol and got some sequencing results! The objective of this vignette is to guide you through a simple analysis of your brand-new UMI-4C contact data. Let’s dive in!</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/Pasquali-lab/UMI4Cats">UMI4Cats</a></span><span class="op">)</span></code></pre></div>
<div id="overview-of-the-package" class="section level2">
<h2 class="hasAnchor">
<a href="#overview-of-the-package" class="anchor"></a>Overview of the package</h2>
<div class="figure" style="text-align: center">
<img src="figures/scheme.png" alt="Overview of the different functions included in the UMI4Cats package to analyze UMI-4C data." width="80%"><p class="caption">
Overview of the different functions included in the UMI4Cats package to analyze UMI-4C data.
</p>
</div>
</div>
<div id="about-the-experimental-design" class="section level2">
<h2 class="hasAnchor">
<a href="#about-the-experimental-design" class="anchor"></a>About the experimental design</h2>
<p>One of the strengths of the UMI-4C assay <span class="citation">(Schwartzman et al. 2016)</span> is that of reducing the PCR duplication bias, allowing a more accurate quantification of chromatin interactions. For this reason, UMI-4C is mostly used when trying to compare changes in chromatin interactions between two conditions, cell types or developmental stages.</p>
<p>Taking into account this main application, UMI4Cats has been developed to facilitate the differential analysis between conditions at a given viewpoint of interest. When analyzing your data with this package, you should take into account the following points:</p>
<ul>
<li><p>Each analysis (and <code>UMI4C</code> object) should correspond to the <strong>same viewpoint</strong>. If you are analyzing different viewpoints in the same or different loci, you need to analyze them separately.</p></li>
<li><p>The UMI4Cats package is mostly oriented to the performance of differential analysis. If you don’t have replicates yet or want to focus your analysis on a specific set of regions (like enhancers) we recommend you to use Fisher’s Exact Test (<code><a href="../reference/fisherUMI4C.html">fisherUMI4C()</a></code>). If, on the other hand, you have several replicates, you might benefit from using a DESeq2 differential test specific for UMI-4C data (<code><a href="../reference/differentialNbinomWaldTestUMI4C.html">differentialNbinomWaldTestUMI4C()</a></code>) or you can continue to use Fisher’s exact test.</p></li>
<li><p>When performing the differential analysis, UMI4Cats is only able to deal with a <strong>condition with 2 different levels</strong>. If you have more than two conditions, you should produce different UMI4C objects and perform pairwise comparisons.</p></li>
</ul>
</div>
<div id="about-the-example-datasets" class="section level2">
<h2 class="hasAnchor">
<a href="#about-the-example-datasets" class="anchor"></a>About the example datasets</h2>
<p>The datasets used in this vignette (obtained from <span class="citation">Ramos-Rodríguez et al. (2019)</span>) are available for download if you want to reproduce the contents of this vignette through the <code><a href="../reference/downloadUMI4CexampleData.html">downloadUMI4CexampleData()</a></code>.</p>
<p>Briefly, the datasets correspond to human pancreatic islets exposed (<code>cyt</code>) or not (<code>ctrl</code>) to pro-inflammatory cytokines for 48 hours. In this example we are using the UMI-4C data generated from two different biological replicates (HI24 and HI32) using the promoter of the <em>CIITA</em> gene as viewpoint.</p>
<p>Sample datasets can be downloaded using the <code><a href="../reference/downloadUMI4CexampleData.html">downloadUMI4CexampleData()</a></code> function. When used without arguments, will download the full sample fastq files containing 200K reads. However, in order to reduce computing time, the <em>Processing UMI-4C FASTQ files</em> section in this vignette uses a reduced sample file containing 100 reads, which can be downloaded using <code><a href="../reference/downloadUMI4CexampleData.html">downloadUMI4CexampleData(reduced = TRUE)</a></code>. The following sections addressing analysis and visualization of such data use the processed files from the 200K fastq files, which are also included inside the package and can be accessed using the <code><a href="https://rdrr.io/r/base/system.file.html">system.file()</a></code> function.</p>
</div>
</div>
<div id="quick-start" class="section level1">
<h1 class="hasAnchor">
<a href="#quick-start" class="anchor"></a>Quick start</h1>
<p>In this section we summarize a complete analysis using the examples provided in this package:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## 0) Download example data -------------------------------</span>
<span class="va">path</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/downloadUMI4CexampleData.html">downloadUMI4CexampleData</a></span><span class="op">(</span><span class="op">)</span>

<span class="co">## 1) Generate Digested genome ----------------------------</span>
<span class="co"># The selected RE in this case is DpnII (|GATC), so the cut_pos is 0, and the res_enz "GATC".</span>
<span class="va">hg19_dpnii</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/digestGenome.html">digestGenome</a></span><span class="op">(</span>
    cut_pos <span class="op">=</span> <span class="fl">0</span>,
    res_enz <span class="op">=</span> <span class="st">"GATC"</span>,
    name_RE <span class="op">=</span> <span class="st">"DpnII"</span>,
    ref_gen <span class="op">=</span> <span class="fu">BSgenome.Hsapiens.UCSC.hg19</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/BSgenome.Hsapiens.UCSC.hg19/man/package.html">BSgenome.Hsapiens.UCSC.hg19</a></span>,
    out_path <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/tempfile.html">tempdir</a></span><span class="op">(</span><span class="op">)</span>, <span class="st">"digested_genome/"</span><span class="op">)</span>
<span class="op">)</span>

<span class="co">## 2) Process UMI-4C fastq files --------------------------</span>
<span class="va">raw_dir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span><span class="op">(</span><span class="va">path</span>, <span class="st">"CIITA"</span>, <span class="st">"fastq"</span><span class="op">)</span>

<span class="fu"><a href="../reference/contactsUMI4C.html">contactsUMI4C</a></span><span class="op">(</span>
    fastq_dir <span class="op">=</span> <span class="va">raw_dir</span>,
    wk_dir <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span><span class="op">(</span><span class="va">path</span>, <span class="st">"CIITA"</span><span class="op">)</span>,
    bait_seq <span class="op">=</span> <span class="st">"GGACAAGCTCCCTGCAACTCA"</span>,
    bait_pad <span class="op">=</span> <span class="st">"GGACTTGCA"</span>,
    res_enz <span class="op">=</span> <span class="st">"GATC"</span>,
    cut_pos <span class="op">=</span> <span class="fl">0</span>,
    digested_genome <span class="op">=</span> <span class="va">hg19_dpnii</span>,
    bowtie_index <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span><span class="op">(</span><span class="va">path</span>, <span class="st">"ref_genome"</span>, <span class="st">"ucsc.hg19.chr16"</span><span class="op">)</span>,
    ref_gen <span class="op">=</span> <span class="fu">BSgenome.Hsapiens.UCSC.hg19</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/BSgenome.Hsapiens.UCSC.hg19/man/package.html">BSgenome.Hsapiens.UCSC.hg19</a></span>,
    threads <span class="op">=</span> <span class="fl">5</span>
<span class="op">)</span>

<span class="co">## 3) Get filtering and alignment stats -------------------</span>
<span class="fu"><a href="../reference/statsUMI4C.html">statsUMI4C</a></span><span class="op">(</span>wk_dir <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span><span class="op">(</span><span class="va">path</span>, <span class="st">"CIITA"</span><span class="op">)</span><span class="op">)</span>

<span class="co">## 4) Analyze the results ---------------------------------</span>
<span class="co"># Load sample processed file paths</span>
<span class="va">files</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html">list.files</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span><span class="op">(</span><span class="va">path</span>, <span class="st">"CIITA"</span>, <span class="st">"count"</span><span class="op">)</span>,
                    pattern <span class="op">=</span> <span class="st">"*_counts.tsv.gz"</span>,
                    full.names <span class="op">=</span> <span class="cn">TRUE</span>
<span class="op">)</span>

<span class="co"># Create colData including all relevant information</span>
<span class="va">colData</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span>
    sampleID <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html">gsub</a></span><span class="op">(</span><span class="st">"_counts.tsv.gz"</span>, <span class="st">""</span>, <span class="fu"><a href="https://rdrr.io/r/base/basename.html">basename</a></span><span class="op">(</span><span class="va">files</span><span class="op">)</span><span class="op">)</span>,
    file <span class="op">=</span> <span class="va">files</span>,
    stringsAsFactors <span class="op">=</span> <span class="cn">FALSE</span>
<span class="op">)</span>

<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyr.tidyverse.org">tidyr</a></span><span class="op">)</span>
<span class="va">colData</span> <span class="op">&lt;-</span> <span class="va">colData</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://tidyr.tidyverse.org/reference/separate.html">separate</a></span><span class="op">(</span><span class="va">sampleID</span>,
        into <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"condition"</span>, <span class="st">"replicate"</span>, <span class="st">"viewpoint"</span><span class="op">)</span>,
        remove <span class="op">=</span> <span class="cn">FALSE</span>
    <span class="op">)</span>

<span class="co"># Make UMI-4C object including grouping by condition</span>
<span class="va">umi</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/UMI4C.html">makeUMI4C</a></span><span class="op">(</span>
    colData <span class="op">=</span> <span class="va">colData</span>,
    viewpoint_name <span class="op">=</span> <span class="st">"CIITA"</span>,
    grouping <span class="op">=</span> <span class="st">"condition"</span>,
    bait_expansion <span class="op">=</span> <span class="fl">2e6</span>
<span class="op">)</span>

<span class="co"># Plot replicates</span>
<span class="fu"><a href="../reference/plotUMI4C.html">plotUMI4C</a></span><span class="op">(</span><span class="va">umi</span>, grouping<span class="op">=</span><span class="cn">NULL</span><span class="op">)</span>

<span class="co">## 5) Differential testing ----------------------</span>
<span class="co"># Fisher test</span>
<span class="va">umi_fisher</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/fisherUMI4C.html">fisherUMI4C</a></span><span class="op">(</span><span class="va">umi</span>, filter_low <span class="op">=</span> <span class="fl">30</span>, grouping<span class="op">=</span><span class="st">"condition"</span><span class="op">)</span>
<span class="fu"><a href="../reference/plotUMI4C.html">plotUMI4C</a></span><span class="op">(</span><span class="va">umi_fisher</span>, ylim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">10</span><span class="op">)</span>, grouping<span class="op">=</span><span class="st">"condition"</span><span class="op">)</span>

<span class="co"># DESeq2 Wald Test</span>
<span class="va">umi_wald</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/differentialNbinomWaldTestUMI4C.html">differentialNbinomWaldTestUMI4C</a></span><span class="op">(</span>umi4c<span class="op">=</span><span class="va">umi</span>,
                                            design<span class="op">=</span><span class="op">~</span><span class="va">condition</span>,
                                            alpha <span class="op">=</span> <span class="fl">20</span><span class="op">)</span>
<span class="fu"><a href="../reference/plotUMI4C.html">plotUMI4C</a></span><span class="op">(</span><span class="va">umi_wald</span>, ylim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">10</span><span class="op">)</span>, grouping<span class="op">=</span><span class="st">"condition"</span><span class="op">)</span></code></pre></div>
</div>
<div id="preparing-necessary-files" class="section level1">
<h1 class="hasAnchor">
<a href="#preparing-necessary-files" class="anchor"></a>Preparing necessary files</h1>
<div id="demultiplexing-fastq-files-containing-multiple-baits" class="section level2">
<h2 class="hasAnchor">
<a href="#demultiplexing-fastq-files-containing-multiple-baits" class="anchor"></a>Demultiplexing FastQ files containing multiple baits</h2>
<p>One of the many advantages of using the UMI-4C protocol is that it allows multiplexing of different baits using the same sample.</p>
<p>To facilitate the analysis, UMI4Cats provides a function for demultiplexing the paired-end FastQ files returned by the sequencing facility: <code>demultiplexFastq</code>.</p>
<p>This function requires the following inputs:</p>
<ul>
<li>Name of the R1 file as input – it will automatically detect the R2.</li>
<li>Barcode sequences.</li>
<li>Path and name for the output files.</li>
</ul>
<p>The barcode sequences and names to be used for each output file need to be provided as a <code>data.frame</code> with column names <code>sample</code> and <code>barcode</code>.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## Input files</span>
<span class="va">path</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/downloadUMI4CexampleData.html">downloadUMI4CexampleData</a></span><span class="op">(</span>reduced<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span>
<span class="va">fastq</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span><span class="op">(</span><span class="va">path</span>, <span class="st">"CIITA"</span>, <span class="st">"fastq"</span>, <span class="st">"ctrl_hi24_CIITA_R1.fastq.gz"</span><span class="op">)</span>

<span class="co">## Barcode info</span>
<span class="va">barcodes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span>
    sample <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"CIITA"</span><span class="op">)</span>,
    barcode <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"GGACAAGCTCCCTGCAACTCA"</span><span class="op">)</span>
<span class="op">)</span>

<span class="co">## Output path</span>
<span class="va">out_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html">tempdir</a></span><span class="op">(</span><span class="op">)</span>

<span class="co">## Demultiplex baits inside FastQ file</span>
<span class="fu"><a href="../reference/demultiplexFastq.html">demultiplexFastq</a></span><span class="op">(</span>
    fastq <span class="op">=</span> <span class="va">fastq</span>,
    barcodes <span class="op">=</span> <span class="va">barcodes</span>,
    out_path <span class="op">=</span> <span class="va">out_path</span>
<span class="op">)</span></code></pre></div>
</div>
<div id="reference-genome-digestion" class="section level2">
<h2 class="hasAnchor">
<a href="#reference-genome-digestion" class="anchor"></a>Reference genome digestion</h2>
<p>For the processing of the UMI-4C FastQ files it is necessary to construct a digested genome using the same restriction enzyme that was used in the UMI-4C experiments.</p>
<p>The <code>UMI4Cats</code> package includes the <code><a href="../reference/digestGenome.html">digestGenome()</a></code> function to make this process easier. The function uses a <code>BSgenome</code> object<a href="#fn1" class="footnote-ref" id="fnref1"><sup>1</sup></a> as the reference genome, which is digested <em>in silico</em> at a given restriction enzyme cutting sequence (<code>res_enz</code>). Besides the restriction sequence, it is also necessary to provide, as a zero-based numeric integer, the position at which the restriction enzyme cuts (<code>cut_pos</code>).</p>
<p>In the following table you can see three examples of the different cutting sequences for <em>DpnII</em>, <em>Csp6I</em> and <em>HindIII</em>.</p>
<table class="table">
<thead><tr class="header">
<th>Restriction enzyme</th>
<th>Restriction seq</th>
<th><code>res_enz</code></th>
<th><code>cut_pos</code></th>
</tr></thead>
<tbody>
<tr class="odd">
<td>DpnII</td>
<td>:<code>GATC</code>
</td>
<td>GATC</td>
<td>0</td>
</tr>
<tr class="even">
<td>Csp6I</td>
<td>
<code>G</code>:<code>TAC</code>
</td>
<td>GTAC</td>
<td>1</td>
</tr>
<tr class="odd">
<td>HindIII</td>
<td>
<code>A</code>:<code>AGCTT</code>
</td>
<td>AAGCTT</td>
<td>1</td>
</tr>
</tbody>
</table>
<p>For this example, we are using the hg19 <code>BSGenome</code> object and we are going to digest it using the <em>DpnII</em> enzyme.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">BSgenome.Hsapiens.UCSC.hg19</span><span class="op">)</span>
<span class="va">refgen</span> <span class="op">&lt;-</span> <span class="va">BSgenome.Hsapiens.UCSC.hg19</span>

<span class="va">hg19_dpnii</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/digestGenome.html">digestGenome</a></span><span class="op">(</span>
    res_enz <span class="op">=</span> <span class="st">"GATC"</span>,
    cut_pos <span class="op">=</span> <span class="fl">0</span>,
    name_RE <span class="op">=</span> <span class="st">"dpnII"</span>,
    ref_gen <span class="op">=</span> <span class="va">refgen</span>,
    sel_chr <span class="op">=</span> <span class="st">"chr16"</span>, <span class="co"># Select bait's chr (chr16) to make example faster</span>
    out_path <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/tempfile.html">tempdir</a></span><span class="op">(</span><span class="op">)</span>, <span class="st">"digested_genome/"</span><span class="op">)</span>
<span class="op">)</span>

<span class="va">hg19_dpnii</span>
<span class="co">#&gt; [1] "/tmp/RtmpADcNkU/digested_genome//BSgenome.Hsapiens.UCSC.hg19_dpnII"</span></code></pre></div>
<p>The digested genome will be saved in the <code>out_path</code> directory as RData objects, one for each chromosome. The path of the digested genome files is outputed by the function, so you can save it as a variable (in this case <code>hg19_dpnii</code>) and use it for downstream analyses.</p>
</div>
</div>
<div id="processing-umi-4c-fastq-files" class="section level1">
<h1 class="hasAnchor">
<a href="#processing-umi-4c-fastq-files" class="anchor"></a>Processing UMI-4C FASTQ files</h1>
<p>Before doing any analysis, we need to convert paired-end reads stored in the FastQ files to UMI counts in the fragments returned from the <em>in silico</em> digestion of the genome, which represent contact frequencies of the viewpoint with that specific fragment. This processing is implemented in the function <code><a href="../reference/contactsUMI4C.html">contactsUMI4C()</a></code>, which should be ran in samples generated with the same experimental design (same viewpoint and restriction enzyme).</p>
<p>The function will consider all FastQ files in the same folder <code>fastq_dir</code> to be part of the same experiment (viewpoint + restriction enzyme). However, if you want to specify a subset of samples to perform the analysis you can do so by using the <code>file_pattern</code> argument. This way, only the files matching the specified pattern will be used as input.</p>
<p>The R1 and R2 files for the each sample need to contain <code>_R1</code> or <code>_R2</code> and one of the following FastQ suffixes: <code>.fastq</code>, <code>.fq</code>, <code>.fq.gz</code> or <code>.fastq.gz</code>.</p>
<p>For each experiment, the user needs to define 3 different sequences:</p>
<ul>
<li>
<strong>Bait/viewpoint sequence</strong> (<code>bait_seq</code>). This is the downstream primer sequence (DS primer) that matches the sequence of the queried viewpoint.</li>
<li>
<strong>Padding sequence</strong> (<code>bait_pad</code>). The padding sequence corresponds to the nucleotides between the DS primer end and the next restriction enzyme site.</li>
<li>
<strong>Restriction enzyme sequence</strong> (<code>res_enz</code>). This sequence is the sequence recognized by the restriction enzyme used in the experiment.</li>
</ul>
<div class="figure" style="text-align: center">
<img src="figures/read_scheme.png" alt="Schematic of a UMI-4C read detailing the different elements that need to be used as input for processing the data." width="80%"><p class="caption">
Schematic of a UMI-4C read detailing the different elements that need to be used as input for processing the data.
</p>
</div>
<p>Additionally, it is necessary to define the restriction enzyme cutting position (<code>cut_pos</code>) as previously done for the digested genome generation, together with the path of such digested genome (<code>digested_genome</code>) returned by the <code><a href="../reference/digestGenome.html">digestGenome()</a></code> function.</p>
<p><code><a href="../reference/contactsUMI4C.html">contactsUMI4C()</a></code> performs the alignment using <a href="http://bowtie-bio.sourceforge.net/bowtie2/index.shtml">Bowtie2</a> and thus, you need to provide the reference genome indexes generated with bowtie2<a href="#fn2" class="footnote-ref" id="fnref2"><sup>2</sup></a>. Remember to make sure that both your Bowtie2 index and your reference and digested genomes correspond to the same build (in this example, hg19).</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## Use reduced example to make vignette faster</span>
<span class="co">## If you want to download the full dataset, set reduced = FALSE or remove</span>
<span class="co">## the reduce argument.</span>
<span class="co">## The reduced example is already downloaded in the demultiplexFastq chunk.</span>

<span class="co"># path &lt;- downloadUMI4CexampleData(reduced=TRUE)</span>
<span class="va">raw_dir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span><span class="op">(</span><span class="va">path</span>, <span class="st">"CIITA"</span>, <span class="st">"fastq"</span><span class="op">)</span>
<span class="va">index_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span><span class="op">(</span><span class="va">path</span>, <span class="st">"ref_genome"</span>, <span class="st">"ucsc.hg19.chr16"</span><span class="op">)</span>

<span class="co">## Run main function to process UMI-4C contacts</span>
<span class="fu"><a href="../reference/contactsUMI4C.html">contactsUMI4C</a></span><span class="op">(</span>
    fastq_dir <span class="op">=</span> <span class="va">raw_dir</span>,
    wk_dir <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span><span class="op">(</span><span class="va">path</span>, <span class="st">"CIITA"</span><span class="op">)</span>,
    file_pattern <span class="op">=</span> <span class="st">"ctrl_hi24_CIITA"</span>, <span class="co"># Select only one sample to reduce running time</span>
    bait_seq <span class="op">=</span> <span class="st">"GGACAAGCTCCCTGCAACTCA"</span>,
    bait_pad <span class="op">=</span> <span class="st">"GGACTTGCA"</span>,
    res_enz <span class="op">=</span> <span class="st">"GATC"</span>,
    cut_pos <span class="op">=</span> <span class="fl">0</span>,
    digested_genome <span class="op">=</span> <span class="va">hg19_dpnii</span>,
    bowtie_index <span class="op">=</span> <span class="va">index_path</span>,
    ref_gen <span class="op">=</span> <span class="fu">BSgenome.Hsapiens.UCSC.hg19</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/BSgenome.Hsapiens.UCSC.hg19/man/package.html">BSgenome.Hsapiens.UCSC.hg19</a></span>,
    sel_seqname <span class="op">=</span> <span class="st">"chr16"</span>, <span class="co"># Input bait chr to reduce running time</span>
    threads <span class="op">=</span> <span class="fl">2</span>
<span class="op">)</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; [2021-01-29 11:18:56] Starting prepUMI4C using:</span>
<span class="co">#&gt;  &gt; Fastq directory:</span>
<span class="co">#&gt;  /tmp/RtmpADcNkU/UMI4Cats_data_reduced/CIITA/fastq </span>
<span class="co">#&gt;  &gt; Work directory: /tmp/RtmpADcNkU/UMI4Cats_data_reduced/CIITA </span>
<span class="co">#&gt;  &gt; Bait sequence: GGACAAGCTCCCTGCAACTCA </span>
<span class="co">#&gt;  &gt; Bait pad: GGACTTGCA </span>
<span class="co">#&gt;  &gt; Restriction enzyme: GATC </span>
<span class="co">#&gt;  &gt; Number of reads loaded into memory: 1e+11</span>
<span class="co">#&gt; [2021-01-29 11:18:57] Finished sample ctrl_hi24_CIITA</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; [2021-01-29 11:18:57] Starting splitUMI4C using:</span>
<span class="co">#&gt;  &gt; Work directory: /tmp/RtmpADcNkU/UMI4Cats_data_reduced/CIITA </span>
<span class="co">#&gt;  &gt; Cut position: 0 </span>
<span class="co">#&gt;  &gt; Restriction enzyme: GATC </span>
<span class="co">#&gt;  &gt; Number of reads loaded into memory: 1e+11</span>
<span class="co">#&gt; [2021-01-29 11:18:57] Finished sample ctrl_hi24_CIITA_R1.fq.gz</span>
<span class="co">#&gt; [2021-01-29 11:18:58] Finished sample ctrl_hi24_CIITA_R2.fq.gz</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; [2021-01-29 11:18:58] Starting alignmentUMI4C using:</span>
<span class="co">#&gt;  &gt; Work directory: /tmp/RtmpADcNkU/UMI4Cats_data_reduced/CIITA </span>
<span class="co">#&gt;  &gt; Viewpoint position: chr16:10972515-10972548 </span>
<span class="co">#&gt;  &gt; Reference genome: /tmp/RtmpADcNkU/UMI4Cats_data_reduced/ref_genome/ucsc.hg19.chr16 </span>
<span class="co">#&gt;  &gt; Number of threads: 2</span>
<span class="co">#&gt; [2021-01-29 11:18:58] Finished sample ctrl_hi24_CIITA_R1.fq</span>
<span class="co">#&gt; [2021-01-29 11:18:58] Finished sample ctrl_hi24_CIITA_R2.fq</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; [2021-01-29 11:18:58] Starting counterUMI4C using:</span>
<span class="co">#&gt;  &gt; Work directory: /tmp/RtmpADcNkU/UMI4Cats_data_reduced/CIITA </span>
<span class="co">#&gt;  &gt; Viewpoint position: chr16:10972515-10972548 </span>
<span class="co">#&gt;  &gt; Restriction enzyme: GATC </span>
<span class="co">#&gt;  &gt; Digested genome: /tmp/RtmpADcNkU/digested_genome//BSgenome.Hsapiens.UCSC.hg19_dpnII</span>
<span class="co">#&gt; [2021-01-29 11:18:59] Finished sample ctrl_hi24_CIITA</span></code></pre></div>
<p>Internally, <code><a href="../reference/contactsUMI4C.html">contactsUMI4C()</a></code> runs the following processes sequentially:</p>
<ol style="list-style-type: decimal">
<li><p><strong>FastQ files preparation (<code>prepUMI4C</code>)</strong>. In this processing step, only reads containing the <code>bait_seq</code> + <code>bait_pad</code> + <code>res_enz</code> are selected. Reads with mean Phread quality scores &lt; 20 are filtered out.</p></li>
<li><p><strong>Split reads at restriction sites (<code>splitUMI4C</code>)</strong>. Using the <code>res_enz</code> sequence and the cutting position (<code>cut_pos</code>), all R1 and R2 reads are split to mimic the fragments generated experimentally.</p></li>
<li><p><strong>Align split reads to the reference genome (<code>alignmentUMI4C</code>)</strong>.</p></li>
<li><p><strong>Collapse reads using UMIs (<code>counterUMI4C</code>)</strong>. This step is done to count real molecular events and reduce artifacts due to PCR duplicates. The function returns contacts with restriction fragments &lt; 5Mb from the viewpoint.</p></li>
</ol>
<p><strong>Note on memory usage</strong>: For the preparation and splitting, the FastQ files are loaded into memory. If you are having problems with the memory size available in your computer, you can change the number of lines that are to be loaded using the <code>numb_reads</code> parameter. See <code><a href="../reference/contactsUMI4C.html">?contactsUMI4C</a></code> for more information.</p>
<p>The final output of this process is a compressed <code>tsv</code> file stored in <code>wk_dir/count</code>, which contains the coordinates for each contact (viewpoint + contact) and the number of UMIs that support that specific interaction. These files will be used as input for the analyses performed in the following section.</p>
<div id="quality-control-measures" class="section level2">
<h2 class="hasAnchor">
<a href="#quality-control-measures" class="anchor"></a>Quality control measures</h2>
<p>Once the processing step has been ran, the statistics of the UMI-4C filtering, alignment and final number of UMIs can be generated from the logs returned by the <code><a href="../reference/contactsUMI4C.html">contactsUMI4C()</a></code> function. By using these logs, the function <code><a href="../reference/statsUMI4C.html">statsUMI4C()</a></code> will produce a summary plot and a summary table with all statistics (<code>wk_dir/logs/stats_summary.txt</code>).</p>
<p>For demonstration purposes we used a reduced version of the fastq files to reduce comptutation time. Thus, now we will load the output of <code><a href="../reference/contactsUMI4C.html">contactsUMI4C()</a></code> with the full dataset, which has ben pre-computed and is saved within the UMI4Cats package.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Using the full dataset included in the package</span>
<span class="fu"><a href="../reference/statsUMI4C.html">statsUMI4C</a></span><span class="op">(</span>wk_dir <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"CIITA"</span>,
    package <span class="op">=</span> <span class="st">"UMI4Cats"</span>
<span class="op">)</span><span class="op">)</span></code></pre></div>
<p><img src="UMI4Cats_files/figure-html/stats-1.png" width="80%" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="co"># Read stats table</span>
<span class="va">stats</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.delim</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"CIITA"</span>, <span class="st">"logs"</span>, <span class="st">"stats_summary.txt"</span>,
    package <span class="op">=</span> <span class="st">"UMI4Cats"</span>
<span class="op">)</span><span class="op">)</span>

<span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html">kable</a></span><span class="op">(</span><span class="va">stats</span><span class="op">)</span></code></pre></div>
<table class="table">
<colgroup>
<col width="15%">
<col width="14%">
<col width="17%">
<col width="14%">
<col width="13%">
<col width="9%">
<col width="11%">
<col width="4%">
</colgroup>
<thead><tr class="header">
<th align="left">sample_id</th>
<th align="right">specific_reads</th>
<th align="right">nonspecific_reads</th>
<th align="right">filtered_reads</th>
<th align="right">filtout_reads</th>
<th align="right">al_mapped</th>
<th align="right">al_unmapped</th>
<th align="right">umi</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">ctrl_hi24_CIITA</td>
<td align="right">183831</td>
<td align="right">16169</td>
<td align="right">173342</td>
<td align="right">10489</td>
<td align="right">473730</td>
<td align="right">75809</td>
<td align="right">1115</td>
</tr>
<tr class="even">
<td align="left">ctrl_hi32_CIITA</td>
<td align="right">183759</td>
<td align="right">16241</td>
<td align="right">181789</td>
<td align="right">1970</td>
<td align="right">479286</td>
<td align="right">98065</td>
<td align="right">2340</td>
</tr>
<tr class="odd">
<td align="left">cyt_hi24_CIITA</td>
<td align="right">181278</td>
<td align="right">18722</td>
<td align="right">178473</td>
<td align="right">2805</td>
<td align="right">479134</td>
<td align="right">90485</td>
<td align="right">1333</td>
</tr>
<tr class="even">
<td align="left">cyt_hi32_CIITA</td>
<td align="right">183770</td>
<td align="right">16230</td>
<td align="right">182106</td>
<td align="right">1664</td>
<td align="right">466677</td>
<td align="right">111822</td>
<td align="right">2198</td>
</tr>
</tbody>
</table>
<p>The quality control measures summarized in the plot and the table are:</p>
<ul>
<li>
<strong>Specific reads</strong>. Corresponds to the number of reads that contain the full viewpoint sequence (<code>bait_seq</code> + <code>bait_pad</code> + <code>res_enz</code>).</li>
<li>
<strong>Filtered reads</strong>. Number of reads with mean Phred quality scores <code>&gt;= 20</code>.</li>
<li>
<strong>Mapping stats</strong>. Indicates how many split reads are mapped or unmapped to the reference genome.</li>
<li>
<strong>UMIs</strong>. Shows the final number of molecular contacts detected.</li>
</ul>
</div>
</div>
<div id="loading-umi-4c-data-into-r" class="section level1">
<h1 class="hasAnchor">
<a href="#loading-umi-4c-data-into-r" class="anchor"></a>Loading UMI-4C data into R</h1>
<p>After processing the FastQ reads and obtaining tables summarizing number of UMIs supporting each fragment interaction with the viewpoint, the next step is to analyze such data by detecting differential contacts and visualizing the genomic interactions.</p>
<div id="build-the-umi4c-object" class="section level2">
<h2 class="hasAnchor">
<a href="#build-the-umi4c-object" class="anchor"></a>Build the <code>UMI4C</code> object</h2>
<p>The first step of the UMI-4C data analysis consists on loading the tables generated by the function <code><a href="../reference/contactsUMI4C.html">contactsUMI4C()</a></code> and use them to construct a <code>UMI4C</code> object, which is based on the <code>SummarizedExperiment</code> class. All these steps are performed automatically by the <code><a href="../reference/UMI4C.html">makeUMI4C()</a></code> function.</p>
<p>The <code>makeUMI4C</code> will need as input a data frame (<code>colData</code>) containing all relevant experiment information that will be needed for analyzing the data later on. The mandatory columns are:</p>
<ol style="list-style-type: decimal">
<li>
<code>sampleID</code>: Unique identifier for the sample.</li>
<li>
<code>replicate</code>: Replicate character or number identifier.</li>
<li>
<code>condition</code>: Grouping variable for performing the differential analysis. For example: “control” and “treatment,” two different cell types, etc. The condition column should only have <strong>two</strong> different values. If more condition variables are provided, the differential analysis will fail.</li>
<li>
<code>file</code>: Complete path and filename of the tsv files generated by <code><a href="../reference/contactsUMI4C.html">contactsUMI4C()</a></code>.</li>
</ol>
<p>You can also include other additional columns to <code>colData</code>.</p>
<p>The <code>UMI4C</code> object will contain data from all the replicates. However, you might be interested in also including information of samples grouped using a specific variable, such as <strong>condition</strong>, to plot combined profiles or perform differential tests on merged replicates. The argument <code>grouping</code> controls this behavior. By default, the grouping argument is set to <code>grouping = "condition"</code>, which will group the samples according to the variables in the <code>condition</code> column. This grouped UMI4C can be accessed using <code>groupsUMI4C(umi4c)$condition</code>. You can also add additional groupings to a specific <code>UMI-4C</code> object using the <code><a href="../reference/addGrouping.html">addGrouping()</a></code> function or avoid the calculation of grouped sample setting <code>grouping = NULL</code>.</p>
<p>Additionally, the <code>makeUMI4C</code> function also contains other arguments that can be used if you want to tweak the default parameters of the analysis. See <code><a href="../reference/UMI4C.html">?makeUMI4C</a></code> to have a complete list and description of all the arguments.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Load sample processed file paths</span>
<span class="va">files</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html">list.files</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"CIITA"</span>, <span class="st">"count"</span>, package<span class="op">=</span><span class="st">"UMI4Cats"</span><span class="op">)</span>,
                    pattern <span class="op">=</span> <span class="st">"*_counts.tsv.gz"</span>,
                    full.names <span class="op">=</span> <span class="cn">TRUE</span>
<span class="op">)</span>

<span class="co"># Create colData including all relevant information</span>
<span class="va">colData</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span>
    sampleID <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html">gsub</a></span><span class="op">(</span><span class="st">"_counts.tsv.gz"</span>, <span class="st">""</span>, <span class="fu"><a href="https://rdrr.io/r/base/basename.html">basename</a></span><span class="op">(</span><span class="va">files</span><span class="op">)</span><span class="op">)</span>,
    file <span class="op">=</span> <span class="va">files</span>,
    stringsAsFactors <span class="op">=</span> <span class="cn">FALSE</span>
<span class="op">)</span>

<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyr.tidyverse.org">tidyr</a></span><span class="op">)</span>
<span class="va">colData</span> <span class="op">&lt;-</span> <span class="va">colData</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://tidyr.tidyverse.org/reference/separate.html">separate</a></span><span class="op">(</span><span class="va">sampleID</span>,
        into <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"condition"</span>, <span class="st">"replicate"</span>, <span class="st">"viewpoint"</span><span class="op">)</span>,
        remove <span class="op">=</span> <span class="cn">FALSE</span>
    <span class="op">)</span>

<span class="co"># Load UMI-4C data and generate UMI4C object</span>
<span class="va">umi</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/UMI4C.html">makeUMI4C</a></span><span class="op">(</span>
    colData <span class="op">=</span> <span class="va">colData</span>,
    viewpoint_name <span class="op">=</span> <span class="st">"CIITA"</span>,
    grouping <span class="op">=</span> <span class="st">"condition"</span>,
    ref_umi4c <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"condition"</span><span class="op">=</span><span class="st">"ctrl"</span><span class="op">)</span>,
    bait_expansion <span class="op">=</span> <span class="fl">2e6</span>
<span class="op">)</span>

<span class="va">umi</span>
<span class="co">#&gt; class: UMI4C </span>
<span class="co">#&gt; dim: 5853 4 </span>
<span class="co">#&gt; metadata(7): bait scales ... ref_umi4c region</span>
<span class="co">#&gt; assays(6): umi norm_mat ... scale sd</span>
<span class="co">#&gt; rownames(5853): frag_1 frag_2 ... frag_5859 frag_5860</span>
<span class="co">#&gt; rowData names(2): id_contact position</span>
<span class="co">#&gt; colnames(4): ctrl_hi24_CIITA ctrl_hi32_CIITA cyt_hi24_CIITA</span>
<span class="co">#&gt;   cyt_hi32_CIITA</span>
<span class="co">#&gt; colData names(5): sampleID condition replicate viewpoint file</span>
<span class="fu"><a href="../reference/dgram.html">groupsUMI4C</a></span><span class="op">(</span><span class="va">umi</span><span class="op">)</span>
<span class="co">#&gt; List of length 1</span>
<span class="co">#&gt; names(1): condition</span></code></pre></div>
<p>The <code>makeUMI4C</code> function will perform the following steps to generate the <code>UMI4C</code> object:</p>
<ol style="list-style-type: decimal">
<li>
<strong>Remove fragment ends around the bait</strong>, as they are generally biased because of their proximity to the viewpoint. The value of the region that will be excluded from the analysis can be specified using the <code>bait_exclusion</code> argument. The default is 3 kb.</li>
<li>
<strong>Focus the scope</strong> of the analysis in a specific genomic region around the bait, by default this is a 2Mb window centerd on the viewpoint. The default value can be changed using the <code>bait_expansion</code> argument.</li>
<li>Sum the UMIs of the different samples belonging to the same group (defined by the <code>grouping</code> variable).</li>
<li>
<strong>Obtain the normalization matrices</strong> that will be used to scale the groups to the reference, by default the group with less UMIs. If you want to avoid this normalization step, you can set <code>normalized</code> to <code>FALSE</code>.</li>
<li>Calculate the <strong>domainograms</strong> for each group.</li>
<li>Calculate the <strong>adaptive trend</strong> for each group.</li>
</ol>
</div>
<div id="accessing-umi4c-object-information" class="section level2">
<h2 class="hasAnchor">
<a href="#accessing-umi4c-object-information" class="anchor"></a>Accessing <code>UMI4C</code> object information</h2>
<p>The usual accessor functions from the <code>SummarizedExperiment-class</code><a href="#fn3" class="footnote-ref" id="fnref3"><sup>3</sup></a> also work with the UMI-4C class (for example: <code>assay</code>, <code>colData</code>, etc.). Other additional accessors have been created to retrieve different information:</p>
<ul>
<li>
<code><a href="../reference/dgram.html">dgram()</a></code>. Get a list of the domaingorams for each group.</li>
<li>
<code><a href="../reference/dgram.html">bait()</a></code>. Retrieve a GRanges object with the bait position.</li>
<li>
<code><a href="../reference/dgram.html">trend()</a></code>. Obtain a data.frame in long format with the adaptive smoothing trend.</li>
<li>
<code><a href="../reference/dgram.html">resultsUMI4C()</a></code>. Retrieve results from the differential analysis. This only works if a differential analysis has been performed on the UMI4C object.</li>
</ul>
<p>You can use this functions in the per-sample <code>UMI4C</code> object or in the grouped <code>UMI4C</code> object, which can be accessed using <code>groupsUMI4C(umi4c)$&lt;groupind-variable&gt;</code>. See an example below.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/dgram.html">groupsUMI4C</a></span><span class="op">(</span><span class="va">umi</span><span class="op">)</span> <span class="co"># Available grouped UMI-4C objects</span>
<span class="co">#&gt; List of length 1</span>
<span class="co">#&gt; names(1): condition</span>

<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="va">umi</span><span class="op">)</span><span class="op">)</span> <span class="co"># Retrieve raw UMIs</span>
<span class="co">#&gt;        ctrl_hi24_CIITA ctrl_hi32_CIITA cyt_hi24_CIITA cyt_hi32_CIITA</span>
<span class="co">#&gt; frag_1               0               0              0              0</span>
<span class="co">#&gt; frag_2               0               0              0              0</span>
<span class="co">#&gt; frag_3               0               2              0              0</span>
<span class="co">#&gt; frag_4               0               0              0              0</span>
<span class="co">#&gt; frag_5               0               0              0              0</span>
<span class="co">#&gt; frag_6               0               0              0              0</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="fu"><a href="../reference/dgram.html">groupsUMI4C</a></span><span class="op">(</span><span class="va">umi</span><span class="op">)</span><span class="op">$</span><span class="va">condition</span><span class="op">)</span><span class="op">)</span> <span class="co"># Retrieve UMIs grouped by condition</span>
<span class="co">#&gt;        ctrl cyt</span>
<span class="co">#&gt; frag_1    0   0</span>
<span class="co">#&gt; frag_2    0   0</span>
<span class="co">#&gt; frag_3    2   0</span>
<span class="co">#&gt; frag_4    0   0</span>
<span class="co">#&gt; frag_5    0   0</span>
<span class="co">#&gt; frag_6    0   0</span>

<span class="fu">colData</span><span class="op">(</span><span class="va">umi</span><span class="op">)</span> <span class="co"># Retrieve column information</span>
<span class="co">#&gt; DataFrame with 4 rows and 5 columns</span>
<span class="co">#&gt;                        sampleID   condition   replicate   viewpoint</span>
<span class="co">#&gt;                     &lt;character&gt; &lt;character&gt; &lt;character&gt; &lt;character&gt;</span>
<span class="co">#&gt; ctrl_hi24_CIITA ctrl_hi24_CIITA        ctrl        hi24       CIITA</span>
<span class="co">#&gt; ctrl_hi32_CIITA ctrl_hi32_CIITA        ctrl        hi32       CIITA</span>
<span class="co">#&gt; cyt_hi24_CIITA   cyt_hi24_CIITA         cyt        hi24       CIITA</span>
<span class="co">#&gt; cyt_hi32_CIITA   cyt_hi32_CIITA         cyt        hi32       CIITA</span>
<span class="co">#&gt;                                   file</span>
<span class="co">#&gt;                            &lt;character&gt;</span>
<span class="co">#&gt; ctrl_hi24_CIITA /usr/local/lib/R/hos..</span>
<span class="co">#&gt; ctrl_hi32_CIITA /usr/local/lib/R/hos..</span>
<span class="co">#&gt; cyt_hi24_CIITA  /usr/local/lib/R/hos..</span>
<span class="co">#&gt; cyt_hi32_CIITA  /usr/local/lib/R/hos..</span>

<span class="fu">rowRanges</span><span class="op">(</span><span class="va">umi</span><span class="op">)</span> <span class="co"># Retrieve fragment coordinates</span>
<span class="co">#&gt; GRanges object with 5853 ranges and 2 metadata columns:</span>
<span class="co">#&gt;             seqnames            ranges strand |  id_contact    position</span>
<span class="co">#&gt;                &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt; &lt;character&gt;</span>
<span class="co">#&gt;      frag_1    chr16   9972440-9972810      * |      frag_1    upstream</span>
<span class="co">#&gt;      frag_2    chr16   9972811-9973022      * |      frag_2    upstream</span>
<span class="co">#&gt;      frag_3    chr16   9973023-9973900      * |      frag_3    upstream</span>
<span class="co">#&gt;      frag_4    chr16   9973901-9974010      * |      frag_4    upstream</span>
<span class="co">#&gt;      frag_5    chr16   9974011-9974144      * |      frag_5    upstream</span>
<span class="co">#&gt;         ...      ...               ...    ... .         ...         ...</span>
<span class="co">#&gt;   frag_5856    chr16 11970551-11971015      * |   frag_5856  downstream</span>
<span class="co">#&gt;   frag_5857    chr16 11971016-11971304      * |   frag_5857  downstream</span>
<span class="co">#&gt;   frag_5858    chr16 11971305-11971499      * |   frag_5858  downstream</span>
<span class="co">#&gt;   frag_5859    chr16 11971500-11971588      * |   frag_5859  downstream</span>
<span class="co">#&gt;   frag_5860    chr16 11971589-11972035      * |   frag_5860  downstream</span>
<span class="co">#&gt;   -------</span>
<span class="co">#&gt;   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span>

<span class="fu"><a href="../reference/dgram.html">dgram</a></span><span class="op">(</span><span class="va">umi</span><span class="op">)</span> <span class="co"># Retrieve domainograms</span>
<span class="co">#&gt; List of length 4</span>
<span class="co">#&gt; names(4): ctrl_hi24_CIITA ctrl_hi32_CIITA cyt_hi24_CIITA cyt_hi32_CIITA</span>
<span class="fu"><a href="../reference/dgram.html">dgram</a></span><span class="op">(</span><span class="fu"><a href="../reference/dgram.html">groupsUMI4C</a></span><span class="op">(</span><span class="va">umi</span><span class="op">)</span><span class="op">$</span><span class="va">condition</span><span class="op">)</span> <span class="co"># Retrieve domainograms</span>
<span class="co">#&gt; List of length 2</span>
<span class="co">#&gt; names(2): ctrl cyt</span>

<span class="fu"><a href="../reference/dgram.html">bait</a></span><span class="op">(</span><span class="va">umi</span><span class="op">)</span> <span class="co"># Retrieve bait coordinates</span>
<span class="co">#&gt; GRanges object with 1 range and 1 metadata column:</span>
<span class="co">#&gt;       seqnames            ranges strand |        name</span>
<span class="co">#&gt;          &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;</span>
<span class="co">#&gt;   [1]    chr16 10970996-10972544      * |       CIITA</span>
<span class="co">#&gt;   -------</span>
<span class="co">#&gt;   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span>

<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="fu"><a href="../reference/dgram.html">trend</a></span><span class="op">(</span><span class="va">umi</span><span class="op">)</span><span class="op">)</span> <span class="co"># Retrieve adaptive smoothing trend</span>
<span class="co">#&gt;   geo_coord      trend         sd scale id_contact          sample sampleID</span>
<span class="co">#&gt; 1  10254166 0.08333333 0.03333333   150   frag_754 ctrl_hi24_CIITA     &lt;NA&gt;</span>
<span class="co">#&gt; 2  10254519 0.08389262 0.03355705   149   frag_755 ctrl_hi24_CIITA     &lt;NA&gt;</span>
<span class="co">#&gt; 3  10254871 0.08445946 0.03378378   148   frag_756 ctrl_hi24_CIITA     &lt;NA&gt;</span>
<span class="co">#&gt; 4  10255223 0.08503401 0.03401361   147   frag_757 ctrl_hi24_CIITA     &lt;NA&gt;</span>
<span class="co">#&gt; 5  10255577 0.08561644 0.03424658   146   frag_758 ctrl_hi24_CIITA     &lt;NA&gt;</span>
<span class="co">#&gt; 6  10255933 0.08620690 0.03448276   145   frag_759 ctrl_hi24_CIITA     &lt;NA&gt;</span>
<span class="co">#&gt;   replicate viewpoint file</span>
<span class="co">#&gt; 1      &lt;NA&gt;      &lt;NA&gt; &lt;NA&gt;</span>
<span class="co">#&gt; 2      &lt;NA&gt;      &lt;NA&gt; &lt;NA&gt;</span>
<span class="co">#&gt; 3      &lt;NA&gt;      &lt;NA&gt; &lt;NA&gt;</span>
<span class="co">#&gt; 4      &lt;NA&gt;      &lt;NA&gt; &lt;NA&gt;</span>
<span class="co">#&gt; 5      &lt;NA&gt;      &lt;NA&gt; &lt;NA&gt;</span>
<span class="co">#&gt; 6      &lt;NA&gt;      &lt;NA&gt; &lt;NA&gt;</span></code></pre></div>
</div>
</div>
<div id="differential-analysis" class="section level1">
<h1 class="hasAnchor">
<a href="#differential-analysis" class="anchor"></a>Differential analysis</h1>
<div id="differential-analysis-using-fishers-exact-test" class="section level2">
<h2 class="hasAnchor">
<a href="#differential-analysis-using-fishers-exact-test" class="anchor"></a>Differential analysis using Fisher’s Exact Test</h2>
<p>Once the <code>UMI4C</code> object is generated, you can perform a differential analysis using the <code><a href="../reference/fisherUMI4C.html">fisherUMI4C()</a></code> function.</p>
<p>First of all, queried regions will be filtered according to the median number of UMIs present in the <code>filter_low</code> parameter. You can reduce this number or disable filtering using <code>filter_low = FALSE</code>.</p>
<p>Then, a contingency table for each region where the differential test should be performed will be created, where the group stored in <code>metadata(umi)$ref_umi4c</code> will be used as references. The values on the contingency table correspond to the following:</p>
<table class="table">
<thead><tr class="header">
<th>Group</th>
<th>Query region</th>
<th>Region</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>Reference</td>
<td><span class="math inline">\(n1\)</span></td>
<td><span class="math inline">\(N1 - n1\)</span></td>
</tr>
<tr class="even">
<td>Condition</td>
<td><span class="math inline">\(n2\)</span></td>
<td><span class="math inline">\(N2 - n2\)</span></td>
</tr>
</tbody>
</table>
<p>Where <span class="math inline">\(N1\)</span> and <span class="math inline">\(N2\)</span> correspond to the total number of UMIs in the whole analyzed region (<code>metadata(umi)$region</code>) and <span class="math inline">\(n1\)</span> and <span class="math inline">\(n2\)</span> correspond to the total number of UMIs in the query region that is to be tested.</p>
<p>After all the Fisher’s Exact Tests are performed, p-values are adjusted using the FDR method. Query regions with adjusted p-values &gt; 0.05 will be considered significantly different. Check <code>?fisherUMI4()</code> for more information and additional arguments you can provide to the function.</p>
<div id="differential-analysis-in-a-specific-set-of-regions" class="section level3">
<h3 class="hasAnchor">
<a href="#differential-analysis-in-a-specific-set-of-regions" class="anchor"></a>Differential analysis in a specific set of regions</h3>
<p>Many times, UMI-4C is used to test whether a certain set of promoter-enhancer contacts differ between conditions or cell types. For this cases, the <code>fisherUMI4c()</code> with the argument <code>query_regions</code> set to those regions of interest can be used. Your query regions should be provided as a <code>GRanges</code> object. If your set of query regions is narrow (~1kb), you can resize them to the desired width using the <code>resize</code> argument.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/GenomicRanges">GenomicRanges</a></span><span class="op">)</span>

<span class="co"># Provide your own query regions as GRanges objects</span>
<span class="va">enhancers</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html">GRanges</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span>
    <span class="st">"chr16:10925006-10928900"</span>,
    <span class="st">"chr16:11102721-11103700"</span>
<span class="op">)</span><span class="op">)</span>

<span class="co"># Perform differential test</span>
<span class="va">umi_dif1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/fisherUMI4C.html">fisherUMI4C</a></span><span class="op">(</span><span class="va">umi</span>,
    grouping <span class="op">=</span> <span class="st">"condition"</span>,
    query_regions <span class="op">=</span> <span class="va">enhancers</span>,
    filter_low <span class="op">=</span> <span class="fl">20</span>,
    resize <span class="op">=</span> <span class="fl">5e3</span>
<span class="op">)</span></code></pre></div>
</div>
<div id="differential-analysis-in-the-binned-region" class="section level3">
<h3 class="hasAnchor">
<a href="#differential-analysis-in-the-binned-region" class="anchor"></a>Differential analysis in the binned region</h3>
<p>In case you do not have a candidate set of enhancer regions, the <code><a href="../reference/fisherUMI4C.html">fisherUMI4C()</a></code> function will bin the experiment region in bin of <code>window_size</code> size (default is 5kb). This will then be the query regions where the differential test will be performed, after filtering for number of UMIs with the <code>filter_low</code> argument.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Perform differential test</span>
<span class="va">umi_dif2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/fisherUMI4C.html">fisherUMI4C</a></span><span class="op">(</span><span class="va">umi</span>,
    grouping <span class="op">=</span> <span class="st">"condition"</span>,
    filter_low <span class="op">=</span> <span class="fl">20</span>,
    window_size <span class="op">=</span> <span class="fl">5e3</span>
<span class="op">)</span></code></pre></div>
</div>
</div>
<div id="differential-analysis-using-deseq2" class="section level2">
<h2 class="hasAnchor">
<a href="#differential-analysis-using-deseq2" class="anchor"></a>Differential Analysis using DESeq2</h2>
<p>It is also possible to perform a differential analysis based on <code>DESeq2</code>’s Wald test. For using this differential test it is necessary to have replicates for the grouping variable included in the design, as they are needed to calculate dispersions.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">umi_dif3</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/differentialNbinomWaldTestUMI4C.html">differentialNbinomWaldTestUMI4C</a></span><span class="op">(</span><span class="va">umi</span>,
                                           design <span class="op">=</span> <span class="op">~</span><span class="va">condition</span>,
                                           alpha<span class="op">=</span><span class="fl">100</span><span class="op">)</span> <span class="co"># Low alpha to make computation faster</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; [2021-01-29 11:19:05] Starting vstUMI4C</span>
<span class="co">#&gt;  &gt; Samples of DDS object:</span>
<span class="co">#&gt;  ctrl_hi24_CIITA, ctrl_hi32_CIITA, cyt_hi24_CIITA, cyt_hi32_CIITA</span>
<span class="co">#&gt; gene-wise dispersion estimates</span>
<span class="co">#&gt; mean-dispersion relationship</span>
<span class="co">#&gt; final dispersion estimates</span>
<span class="co">#&gt; [2021-01-29 11:19:07] Finished vstUMI4C</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; [2021-01-29 11:19:07] Starting smoothMonotoneUMI4C using:</span>
<span class="co">#&gt;  &gt; Samples of DDS object:</span>
<span class="co">#&gt;  ctrl_hi24_CIITA, ctrl_hi32_CIITA, cyt_hi24_CIITA, cyt_hi32_CIITA </span>
<span class="co">#&gt;  &gt; Alpha:</span>
<span class="co">#&gt;  100 </span>
<span class="co">#&gt;  &gt; Penalty:</span>
<span class="co">#&gt;  0.1</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;  Results </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Iter.   PENSSE   Grad Length Intercept   Slope</span>
<span class="co">#&gt; 0        0.1605      0.002      1.3693      -0.1402</span>
<span class="co">#&gt; 1        0.1554      0.003      2.4147      -0.2588</span>
<span class="co">#&gt; 2        0.155      2e-04      2.4607      -0.2651</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;  Results </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Iter.   PENSSE   Grad Length Intercept   Slope</span>
<span class="co">#&gt; 0        0.3019      0.0068      2.2821      -0.2848</span>
<span class="co">#&gt; 1        0.2944      0.006      2.9711      -0.3932</span>
<span class="co">#&gt; 2        0.2915      0.0028      3.3079      -0.4437</span>
<span class="co">#&gt; 3        0.2907      0.0016      3.5      -0.4759</span>
<span class="co">#&gt; 4        0.2905      8e-04      3.6235      -0.4962</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;  Results </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Iter.   PENSSE   Grad Length Intercept   Slope</span>
<span class="co">#&gt; 0        0.1911      0.0027      1.529      -0.1656</span>
<span class="co">#&gt; 1        0.1853      0.0039      2.5313      -0.2908</span>
<span class="co">#&gt; 2        0.1844      4e-04      2.6208      -0.3035</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;  Results </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Iter.   PENSSE   Grad Length Intercept   Slope</span>
<span class="co">#&gt; 0        0.2977      0.0051      2.1222      -0.2579</span>
<span class="co">#&gt; 1        0.2935      0.0043      2.5996      -0.3353</span>
<span class="co">#&gt; 2        0.2919      0.0019      2.8275      -0.3712</span>
<span class="co">#&gt; 3        0.2914      0.0012      2.9752      -0.3965</span>
<span class="co">#&gt; 4        0.2913      6e-04      3.0622      -0.4112</span>
<span class="co">#&gt; [2021-01-29 11:19:20] Finished smoothMonotoneUMI4C</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; [2021-01-29 11:19:20] Starting nbinomWaldTestUMI4C using:</span>
<span class="co">#&gt;  &gt; Samples of DDS object:</span>
<span class="co">#&gt;  ctrl_hi24_CIITA, ctrl_hi32_CIITA, cyt_hi24_CIITA, cyt_hi32_CIITA</span>
<span class="co">#&gt; found already estimated dispersions, replacing these</span>
<span class="co">#&gt; gene-wise dispersion estimates</span>
<span class="co">#&gt; mean-dispersion relationship</span>
<span class="co">#&gt; -- note: fitType='parametric', but the dispersion trend was not well captured by the</span>
<span class="co">#&gt;    function: y = a/x + b, and a local regression fit was automatically substituted.</span>
<span class="co">#&gt;    specify fitType='local' or 'mean' to avoid this message next time.</span>
<span class="co">#&gt; final dispersion estimates</span>
<span class="co">#&gt; [2021-01-29 11:19:22] Finished sample nbinomWaldTestUMI4C</span></code></pre></div>
<p>This differential analysis works as follow:</p>
<ol style="list-style-type: decimal">
<li>Variance stabilizing transformation (VST) of the raw UMI counts.</li>
<li>Smooth monotone fitting of the VST counts.</li>
<li>Differential interactions between groups.</li>
</ol>
<div class="figure" style="text-align: center">
<img src="figures/differentialNbinomWaldTestUMI4C.png" alt="DESeq2 Differential Analysis flowchart." width="80%"><p class="caption">
DESeq2 Differential Analysis flowchart.
</p>
</div>
<div id="variance-stabilizing-transformation-vst-of-the-raw-counts" class="section level3">
<h3 class="hasAnchor">
<a href="#variance-stabilizing-transformation-vst-of-the-raw-counts" class="anchor"></a>Variance stabilizing transformation (VST) of the raw counts</h3>
<p>In 4C experiments, the standard deviations across samples are large for high abundance fragments. Such heteroscedasticity may skew the differential analysis towards fragments either far or close to the viewpoint. Variance stabilizing transformation (VST) is used to remove the dependence of the variance on the mean, thus correcting the dependence of the standard deviations to the fragment abundance.</p>
<p>VST is calculated performing the following steps:</p>
<ol style="list-style-type: decimal">
<li>Count data normalization using a median of ratios method.</li>
<li>Estimation of dispersions.</li>
<li>Fitting a curve to the dispersion estimates for calculating the VST.</li>
<li>Transform the count data using the VST.</li>
</ol>
</div>
<div id="smooth-monotone-fitting" class="section level3">
<h3 class="hasAnchor">
<a href="#smooth-monotone-fitting" class="anchor"></a>Smooth monotone fitting</h3>
<p>The 4C signal decays with genomic distance from the viewpoint and converges towards a constant level of background. 4C data reflects a smooth strictly increasing or strictly decreasing function, so it is desirable to smooth the data with a strictly monotone function. Thus, the general decay of the 4C signal with genomic distance from the viewpoint is fitted using a symmetric monotone fit.</p>
<p>The monotone smoothing function is calculated using the transformed raw counts, using the <a href="https://cran.r-project.org/package=fda"><code>fda</code></a> package. Then, the fitted count values are generated using such function.</p>
</div>
<div id="differential-interactions-between-groups" class="section level3">
<h3 class="hasAnchor">
<a href="#differential-interactions-between-groups" class="anchor"></a>Differential interactions between groups</h3>
<p>The fitted values are VST-back-transformed to calculate the normalization factors, which will represent the fitted distance dependence on the scale of the raw counts. This will allow to take into account the signal’s distance dependence between samples for detecting differential interactions.</p>
<p>Differential interactions are then tested using <a href="https://bioconductor.org/packages/release/bioc/html/DESeq2.html"><code>DESeq2</code></a>’s Negative Binomial GLM test, taking advantage of the previously calculated normalization factors and dispersion estimates.</p>
</div>
</div>
<div id="retrieve-differential-analysis-results" class="section level2">
<h2 class="hasAnchor">
<a href="#retrieve-differential-analysis-results" class="anchor"></a>Retrieve differential analysis results</h2>
<p>Results from both Fisher’s Exact test and DESeq2 can be retrieved using the <code><a href="../reference/dgram.html">resultsUMI4C()</a></code> on the <code>UMI4C</code> object returned by both functions.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/dgram.html">resultsUMI4C</a></span><span class="op">(</span><span class="va">umi_dif1</span>, ordered <span class="op">=</span> <span class="cn">TRUE</span>, counts <span class="op">=</span> <span class="cn">TRUE</span>, format <span class="op">=</span> <span class="st">"data.frame"</span><span class="op">)</span>
<span class="co">#&gt;   seqnames    start      end       id umis_ref umis_cond       pvalue</span>
<span class="co">#&gt; 1    chr16 10924453 10929452 region_1       29        74 1.461849e-05</span>
<span class="co">#&gt; 2    chr16 11100711 11105710 region_2       15        11           NA</span>
<span class="co">#&gt;   odds_ratio log2_odds_ratio         padj sign</span>
<span class="co">#&gt; 1   2.525929        1.336814 1.461849e-05 TRUE</span>
<span class="co">#&gt; 2         NA              NA           NA   NA</span></code></pre></div>
<p>The parameter <code>counts</code> indicates whether raw counts used for the test should be outputted. In Fisher’s Exact Test, <code>umis_ref</code> corresponds to the number of raw UMIs from the sample/group used as reference (accessible through <code>metadata(umi_dif)$ref_umi4c</code>).</p>
</div>
</div>
<div id="visualizing-umi-4c-contact-data" class="section level1">
<h1 class="hasAnchor">
<a href="#visualizing-umi-4c-contact-data" class="anchor"></a>Visualizing UMI-4C contact data</h1>
<p>Once the <code>UMI4C</code> object is created, you can visualize detected chromatin interactions using the <code>plotUMI4C</code> function.</p>
<p>The gene annotations will be extracted from the <code>TxDb.Hsapiens.UCSC.hg19.knownGene</code> package by default. Make sure that the annotations coincide with your reference genome. You can check the package <a href="https://bioconductor.org/packages/release/bioc/html/GenomicFeatures.html"><code>GenomicFeatures</code></a> for more information on available <code>TxDb</code> objects. The domainogram plotting is controlled by the <code>dgram_plot</code> argument. If you set it to <code>FALSE</code>, the domainogram will not be plotted.</p>
<p>In case you are interested in plotting the profiles of the different samples contained in your experiment, you just need to set the <code>grouping</code> argument to <code>NULL</code>, which will disable sample grouping:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plotUMI4C.html">plotUMI4C</a></span><span class="op">(</span><span class="va">umi</span>,
    grouping <span class="op">=</span> <span class="cn">NULL</span>,
    TxDb <span class="op">=</span> <span class="fu">TxDb.Hsapiens.UCSC.hg19.knownGene</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/TxDb.Hsapiens.UCSC.hg19.knownGene/man/package.html">TxDb.Hsapiens.UCSC.hg19.knownGene</a></span>,
    dgram_plot <span class="op">=</span> <span class="cn">FALSE</span>
<span class="op">)</span></code></pre></div>
<p><img src="UMI4Cats_files/figure-html/plot-umi4c-1.png" width="80%" style="display: block; margin: auto;"></p>
<p>If the <code>UMI4C</code> object contains information on the differential contacts, this data will be shown in the plot as well. The grouping argument uses the grouped trends and domainograms stored in <code>groupsuMI4C()</code>. If you want to add a new grouping, you can use the <code>addGRouping()</code> function.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plotUMI4C.html">plotUMI4C</a></span><span class="op">(</span><span class="va">umi_dif1</span>, grouping <span class="op">=</span> <span class="st">"condition"</span><span class="op">)</span></code></pre></div>
<p><img src="UMI4Cats_files/figure-html/plot-dif-1.png" width="80%" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plotUMI4C.html">plotUMI4C</a></span><span class="op">(</span><span class="va">umi_dif3</span>, grouping <span class="op">=</span> <span class="st">"condition"</span><span class="op">)</span></code></pre></div>
<p><img src="UMI4Cats_files/figure-html/plot-dif-2.png" width="80%" style="display: block; margin: auto;"></p>
<p>There are several different arguments you can provide to <code>plotUMI4C</code> to modify the output plot. You can check them by typing <code><a href="../reference/plotUMI4C.html">?plotUMI4C</a></code> in your R console.</p>
<p>The <code>plotUMI4C</code> function is a wrapper for separate functions that plot the different elements included in the figure. You can use each of the functions separately if you are interesting in combining them differently or with other ggplot2 objects. Check each function documentation at <code><a href="../reference/plotTrend.html">?plotTrend</a></code>, <code><a href="../reference/plotGenes.html">?plotGenes</a></code>, <code><a href="../reference/plotDomainogram.html">?plotDomainogram</a></code> and <code><a href="../reference/plotDifferential.html">?plotDifferential</a></code>.</p>
</div>
<div id="references" class="section level1 unnumbered">
<h1 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h1>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-Ramos-Rodriguez2019" class="csl-entry">
Ramos-Rodríguez, Mireia, Helena Raurell-Vila, Maikel L. Colli, Maria Inês Alvelos, Marc Subirana-Granés, Jonàs Juan-Mateu, Richard Norris, et al. 2019. <span>“<span class="nocase">The impact of proinflammatory cytokines on the <span class="math inline">\(\beta\)</span>-cell regulatory landscape provides insights into the genetics of type 1 diabetes</span>.”</span> <em>Nature Genetics</em> 51 (11): 1588–95. <a href="https://doi.org/10.1038/s41588-019-0524-6">https://doi.org/10.1038/s41588-019-0524-6</a>.
</div>
<div id="ref-Schwartzman2016" class="csl-entry">
Schwartzman, Omer, Zohar Mukamel, Noa Oded-Elkayam, Pedro Olivares-Chauvet, Yaniv Lubling, Gilad Landan, Shai Izraeli, and Amos Tanay. 2016. <span>“<span class="nocase">UMI-4C for quantitative and targeted chromosomal contact profiling</span>.”</span> <em>Nature Methods</em> 13 (8): 685–91. <a href="https://doi.org/10.1038/nmeth.3922">https://doi.org/10.1038/nmeth.3922</a>.
</div>
</div>
</div>
<div class="footnotes">
<hr>
<ol>
<li id="fn1"><p>More information on <code>BSgenome</code> package and objects can be found <a href="https://bioconductor.org/packages/release/bioc/html/BSgenome.html">here</a><a href="#fnref1" class="footnote-back">↩︎</a></p></li>
<li id="fn2"><p>See the <a href="http://bowtie-bio.sourceforge.net/bowtie2/manual.shtml#getting-started-with-bowtie-2-lambda-phage-example">getting started section</a> on the Bowtie2 page for more information on how to generate the index for the reference genome.<a href="#fnref2" class="footnote-back">↩︎</a></p></li>
<li id="fn3"><p>See more about the SummarizedExperiment class <a href="https://bioconductor.org/packages/release/bioc/html/SummarizedExperiment.html">here</a><a href="#fnref3" class="footnote-back">↩︎</a></p></li>
</ol>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Mireia Ramos-Rodriguez, Marc Subirana-Granes.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.9000.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
